<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to slendr ‚Äì Simulation-Based Population Genomics in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./handouts.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-485d01fc63b59abcd3ee1bf1e8e2748d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./handouts.html">Handouts</a></li><li class="breadcrumb-item"><a href="./handouts_slendr.html"><span class="chapter-title">Introduction to <em>slendr</em></span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Simulation-Based Population Genomics in R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">R programming</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-language.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">R programming language</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-thinking.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Algorithmic thinking</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Modeling fundamentals</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-models.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Demographic models</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-stats.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Tree-sequence statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-nonspatial-pca.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Exploring PCA patterns</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-spatial-pca.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Spatial PCA</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-tracts-dating.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Admixture tracts and dating</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Natural selection</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Simulation-based inference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inference-afs.html" class="sidebar-item-text sidebar-link"><span class="chapter-title"><span class="math inline">\(N_e\)</span> inference with AFS</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./handouts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Handouts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./handouts_slendr.html" class="sidebar-item-text sidebar-link active"><span class="chapter-title">Introduction to <em>slendr</em></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Installation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Slides</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#why-use-simulations" id="toc-why-use-simulations" class="nav-link active" data-scroll-target="#why-use-simulations">Why use simulations?</a></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  <li><a href="#why-use-simulations-1" id="toc-why-use-simulations-1" class="nav-link" data-scroll-target="#why-use-simulations-1">Why use simulations?</a>
  <ul class="collapse">
  <li><a href="#developing-intuition-into-statistics" id="toc-developing-intuition-into-statistics" class="nav-link" data-scroll-target="#developing-intuition-into-statistics">Developing intuition into statistics</a></li>
  <li><a href="#developing-intuition-into-statistics-1" id="toc-developing-intuition-into-statistics-1" class="nav-link" data-scroll-target="#developing-intuition-into-statistics-1">Developing intuition into statistics</a></li>
  <li><a href="#estimating-model-parameters-i.e.-abc" id="toc-estimating-model-parameters-i.e.-abc" class="nav-link" data-scroll-target="#estimating-model-parameters-i.e.-abc">Estimating model parameters (i.e.&nbsp;ABC)</a></li>
  <li><a href="#ground-truth-for-method-development" id="toc-ground-truth-for-method-development" class="nav-link" data-scroll-target="#ground-truth-for-method-development">Ground truth for method development</a></li>
  <li><a href="#simulation-software" id="toc-simulation-software" class="nav-link" data-scroll-target="#simulation-software">Simulation software</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1"></a></li>
  <li><a href="#simple-simulation-using-msprime" id="toc-simple-simulation-using-msprime" class="nav-link" data-scroll-target="#simple-simulation-using-msprime">Simple simulation using <em>msprime</em></a></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2"></a></li>
  <li><a href="#simple-neutral-simulation-in-slim" id="toc-simple-neutral-simulation-in-slim" class="nav-link" data-scroll-target="#simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</a></li>
  </ul></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3"></a>
  <ul class="collapse">
  <li><a href="#so-why-slendr" id="toc-so-why-slendr" class="nav-link" data-scroll-target="#so-why-slendr">‚Ä¶ so why <em>slendr</em>?</a></li>
  <li><a href="#first-motivation-spatial-simulations" id="toc-first-motivation-spatial-simulations" class="nav-link" data-scroll-target="#first-motivation-spatial-simulations">First motivation: spatial simulations!</a></li>
  <li><a href="#a-broader-motivation-for-slendr" id="toc-a-broader-motivation-for-slendr" class="nav-link" data-scroll-target="#a-broader-motivation-for-slendr">A broader motivation for <em>slendr</em></a></li>
  </ul></li>
  <li><a href="#slendr-crash-course" id="toc-slendr-crash-course" class="nav-link" data-scroll-target="#slendr-crash-course"><em>slendr</em> crash course</a>
  <ul class="collapse">
  <li><a href="#typical-slendr-workflow" id="toc-typical-slendr-workflow" class="nav-link" data-scroll-target="#typical-slendr-workflow">Typical <em>slendr</em> workflow</a></li>
  <li><a href="#creating-populations" id="toc-creating-populations" class="nav-link" data-scroll-target="#creating-populations">Creating populations</a></li>
  <li><a href="#programming-population-splits" id="toc-programming-population-splits" class="nav-link" data-scroll-target="#programming-population-splits">Programming population splits</a></li>
  <li><a href="#scheduling-resize-events" id="toc-scheduling-resize-events" class="nav-link" data-scroll-target="#scheduling-resize-events">Scheduling resize events</a></li>
  <li><a href="#tidyverse-style-pipe-interface" id="toc-tidyverse-style-pipe-interface" class="nav-link" data-scroll-target="#tidyverse-style-pipe-interface">Tidyverse-style pipe <code>%&gt;%</code> interface</a></li>
  <li><a href="#a-more-complex-model" id="toc-a-more-complex-model" class="nav-link" data-scroll-target="#a-more-complex-model">A more complex model</a></li>
  <li><a href="#again-each-object-carries-its-history" id="toc-again-each-object-carries-its-history" class="nav-link" data-scroll-target="#again-each-object-carries-its-history">Again, each object carries its history!</a></li>
  <li><a href="#gene-flow-admixture" id="toc-gene-flow-admixture" class="nav-link" data-scroll-target="#gene-flow-admixture">Gene flow / admixture</a></li>
  <li><a href="#model-compilation" id="toc-model-compilation" class="nav-link" data-scroll-target="#model-compilation">Model compilation</a></li>
  <li><a href="#model-compilation-1" id="toc-model-compilation-1" class="nav-link" data-scroll-target="#model-compilation-1">Model compilation</a></li>
  <li><a href="#model-summary" id="toc-model-summary" class="nav-link" data-scroll-target="#model-summary">Model summary</a></li>
  <li><a href="#model-visualization" id="toc-model-visualization" class="nav-link" data-scroll-target="#model-visualization">Model visualization</a></li>
  </ul></li>
  <li><a href="#a-note-on-units-of-time-and-its-direction" id="toc-a-note-on-units-of-time-and-its-direction" class="nav-link" data-scroll-target="#a-note-on-units-of-time-and-its-direction">A note on units of time (and its direction)</a>
  <ul class="collapse">
  <li><a href="#forward-time-units" id="toc-forward-time-units" class="nav-link" data-scroll-target="#forward-time-units">‚ÄúForward time units‚Äù</a></li>
  <li><a href="#forward-time-units-1" id="toc-forward-time-units-1" class="nav-link" data-scroll-target="#forward-time-units-1">‚ÄúForward time units‚Äù</a></li>
  <li><a href="#backward-time-units" id="toc-backward-time-units" class="nav-link" data-scroll-target="#backward-time-units">‚ÄúBackward time units‚Äù</a></li>
  <li><a href="#backward-time-units-1" id="toc-backward-time-units-1" class="nav-link" data-scroll-target="#backward-time-units-1">‚ÄúBackward time units‚Äù</a></li>
  </ul></li>
  <li><a href="#so-we-built-a-model" id="toc-so-we-built-a-model" class="nav-link" data-scroll-target="#so-we-built-a-model">So we built a model‚Ä¶</a>
  <ul class="collapse">
  <li><a href="#built-in-simulation-engines" id="toc-built-in-simulation-engines" class="nav-link" data-scroll-target="#built-in-simulation-engines">Built-in simulation ‚Äúengines‚Äù</a></li>
  </ul></li>
  <li><a href="#the-result-of-a-simulation-is-a-tree-sequence-ts" id="toc-the-result-of-a-simulation-is-a-tree-sequence-ts" class="nav-link" data-scroll-target="#the-result-of-a-simulation-is-a-tree-sequence-ts">The result of a simulation is a tree sequence (<code>ts</code>)</a>
  <ul class="collapse">
  <li><a href="#what-is-tree-sequence" id="toc-what-is-tree-sequence" class="nav-link" data-scroll-target="#what-is-tree-sequence">What is tree sequence?</a></li>
  </ul></li>
  <li><a href="#why-tree-sequence" id="toc-why-tree-sequence" class="nav-link" data-scroll-target="#why-tree-sequence">Why tree sequence?</a>
  <ul class="collapse">
  <li><a href="#what-we-usually-have" id="toc-what-we-usually-have" class="nav-link" data-scroll-target="#what-we-usually-have">What we usually have</a></li>
  <li><a href="#what-we-usually-want" id="toc-what-we-usually-want" class="nav-link" data-scroll-target="#what-we-usually-want">What we usually <em>want</em></a></li>
  <li><a href="#the-magic-of-tree-sequences" id="toc-the-magic-of-tree-sequences" class="nav-link" data-scroll-target="#the-magic-of-tree-sequences">The magic of tree sequences</a></li>
  <li><a href="#what-if-we-need-mutations-though" id="toc-what-if-we-need-mutations-though" class="nav-link" data-scroll-target="#what-if-we-need-mutations-though">What if we need mutations though?</a></li>
  <li><a href="#what-if-we-need-mutations-though-1" id="toc-what-if-we-need-mutations-though-1" class="nav-link" data-scroll-target="#what-if-we-need-mutations-though-1">What if we need mutations though?</a></li>
  <li><a href="#lets-go-back-to-our-example-model" id="toc-lets-go-back-to-our-example-model" class="nav-link" data-scroll-target="#lets-go-back-to-our-example-model">Let‚Äôs go back to our example <code>model</code>‚Ä¶</a></li>
  <li><a href="#simulate-a-tree-sequence" id="toc-simulate-a-tree-sequence" class="nav-link" data-scroll-target="#simulate-a-tree-sequence">‚Ä¶ simulate a tree sequence‚Ä¶</a></li>
  <li><a href="#and-overlay-mutations-on-it" id="toc-and-overlay-mutations-on-it" class="nav-link" data-scroll-target="#and-overlay-mutations-on-it">‚Ä¶ and overlay mutations on it</a></li>
  </ul></li>
  <li><a href="#so-we-can-simulate-data" id="toc-so-we-can-simulate-data" class="nav-link" data-scroll-target="#so-we-can-simulate-data">So we can simulate data</a>
  <ul class="collapse">
  <li><a href="#slendrs-r-interface-to-tskit-statistics" id="toc-slendrs-r-interface-to-tskit-statistics" class="nav-link" data-scroll-target="#slendrs-r-interface-to-tskit-statistics"><em>slendr</em>‚Äôs R interface to <em>tskit</em> statistics</a></li>
  <li><a href="#extracting-names-of-recorded-samples" id="toc-extracting-names-of-recorded-samples" class="nav-link" data-scroll-target="#extracting-names-of-recorded-samples">Extracting names of recorded samples</a></li>
  </ul></li>
  <li><a href="#section-4" id="toc-section-4" class="nav-link" data-scroll-target="#section-4"></a>
  <ul class="collapse">
  <li><a href="#tskit-computation-option-1" id="toc-tskit-computation-option-1" class="nav-link" data-scroll-target="#tskit-computation-option-1"><em>tskit</em> computation ‚Äì option #1</a></li>
  <li><a href="#tskit-computation-option-2" id="toc-tskit-computation-option-2" class="nav-link" data-scroll-target="#tskit-computation-option-2"><em>tskit</em> computation ‚Äì option #2</a></li>
  <li><a href="#tskit-computation-option-3" id="toc-tskit-computation-option-3" class="nav-link" data-scroll-target="#tskit-computation-option-3"><em>tskit</em> computation ‚Äì option #3</a></li>
  </ul></li>
  <li><a href="#section-5" id="toc-section-5" class="nav-link" data-scroll-target="#section-5"></a>
  <ul class="collapse">
  <li><a href="#example-nucleotide-diversity" id="toc-example-nucleotide-diversity" class="nav-link" data-scroll-target="#example-nucleotide-diversity">Example: nucleotide diversity</a></li>
  <li><a href="#example-allele-frequency-spectrum" id="toc-example-allele-frequency-spectrum" class="nav-link" data-scroll-target="#example-allele-frequency-spectrum">Example: allele frequency spectrum</a></li>
  </ul></li>
  <li><a href="#section-6" id="toc-section-6" class="nav-link" data-scroll-target="#section-6"></a>
  <ul class="collapse">
  <li><a href="#tree-sequence-tables" id="toc-tree-sequence-tables" class="nav-link" data-scroll-target="#tree-sequence-tables">Tree sequence tables</a></li>
  <li><a href="#how-is-this-even-possible" id="toc-how-is-this-even-possible" class="nav-link" data-scroll-target="#how-is-this-even-possible">How is this even possible?!</a></li>
  <li><a href="#tree-sequence-tables-1" id="toc-tree-sequence-tables-1" class="nav-link" data-scroll-target="#tree-sequence-tables-1">Tree-sequence tables</a></li>
  <li><a href="#tree-sequence-tables-in-practice" id="toc-tree-sequence-tables-in-practice" class="nav-link" data-scroll-target="#tree-sequence-tables-in-practice">Tree-sequence tables in practice</a></li>
  </ul></li>
  <li><a href="#other-data-formats" id="toc-other-data-formats" class="nav-link" data-scroll-target="#other-data-formats">Other data formats</a>
  <ul class="collapse">
  <li><a href="#standard-genotype-formats" id="toc-standard-genotype-formats" class="nav-link" data-scroll-target="#standard-genotype-formats">Standard genotype formats</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./handouts.html">Handouts</a></li><li class="breadcrumb-item"><a href="./handouts_slendr.html"><span class="chapter-title">Introduction to <em>slendr</em></span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-title">Introduction to <em>slendr</em></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="why-use-simulations" class="level1">
<h1>Why use simulations?</h1>
</section>
<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It [‚Ä¶] permits people with limited mathematical knowledge to work on important problems [‚Ä¶].</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="files/slendr/images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> ‚Äì <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-use-simulations-1" class="level1 page-columns page-full">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Developing intuition into statistics</li>
<li>Estimating model parameters</li>
<li>Ground truth for method development</li>
</ol>
<section id="developing-intuition-into-statistics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="developing-intuition-into-statistics">Developing intuition into statistics</h2>
<center>
<img src="files/slendr/images/fstats_sims.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><small>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></small></p>
</div></div></section>
<section id="developing-intuition-into-statistics-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="developing-intuition-into-statistics-1">Developing intuition into statistics</h2>
<center>
<img src="files/slendr/images/admixture.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><small>Image from <a href="https://www.nature.com/articles/s41467-018-05257-7">Lawson <em>et al.</em> (2018)</a></small></p>
</div></div></section>
<section id="estimating-model-parameters-i.e.-abc" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="estimating-model-parameters-i.e.-abc">Estimating model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="files/slendr/images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><small>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></small></p>
</div></div></section>
<section id="ground-truth-for-method-development" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="ground-truth-for-method-development">Ground truth for method development</h2>
<center>
<img src="files/slendr/images/mcmc.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><small>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></small></p>
</div></div></section>
<section id="simulation-software" class="level2">
<h2 class="anchored" data-anchor-id="simulation-software">Simulation software</h2>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<div class="fragment">
<p>They are <u>very</u> powerful and (nearly) infinitely flexible.</p>
</div>
<div class="fragment">
<p>However, they both require:</p>
<ul>
<li>quite a bit of code for complex simulations (‚Äúcomplex‚Äù is relative, of course)</li>
<li>relatively high confidence in programming</li>
</ul>
<!-- (üêõü™≤üêú). -->
</div>
<div class="fragment">
<center>
<h4 class="anchored">
Our exercises will focus on the <a href="http://www.slendr.net"><em>slendr</em></a><br> simulation toolkit for population genetics in R.
</h4>
</center>
</div>
<div class="fragment">
<center>
<br>But, as a recap, let‚Äôs look at <em>msprime</em> and SLiM a little bit‚Ä¶
</center>
</div>
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <em>msprime</em>?
</h2>
<ul>
<li><p>A Python module for writing <strong>coalescent simulations</strong></p></li>
<li><p>Extremely fast (genome-scale, population-scale data!)</p></li>
<li><p>You should know Python fairly well to build complex models</p></li>
</ul>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="files/slendr/images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><small>Image modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></small></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="level2">
<h2 class="anchored" data-anchor-id="simple-simulation-using-msprime">Simple simulation using <em>msprime</em></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
</section>
<section id="section-2" class="level2">
<h2 class="anchored" data-anchor-id="section-2"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is SLiM?
</h2>
<ul>
<li><p><strong>A forward-time simulator</strong></p></li>
<li><p>Has its own programming language</p></li>
<li><p>Massive library of functions for:</p>
<ul>
<li>demographic events</li>
<li>various mating systems</li>
<li>natural selection</li>
</ul></li>
<li><p>More than 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v4.3/SLiM_Manual.pdf">manual</a>!</p></li>
</ul>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="files/slendr/images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><small>Image modified from <a href="http://alexeidrummond.org/bayesian_phylo_lectures/lecture10/">Alexei Drummond</a></small></figcaption>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-neutral-simulation-in-slim" class="level2">
<h2 class="anchored" data-anchor-id="simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);

    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);

    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");

    sim.simulationFinished();
}</code></pre>
</div>
</div>
</section>
</section>
<section id="section-3" class="level1">
<h1></h1>
<h3 class="anchored">
SLiM and <em>msprime</em> are both incredible pieces of software‚Ä¶
</h3>
<section id="so-why-slendr" class="level2">
<h2 class="anchored" data-anchor-id="so-why-slendr">‚Ä¶ so why <em>slendr</em>?</h2>
<center>
<p><img src="files/slendr/images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
<h4 style="font-family: 'Courier New'" class="anchored">
<a href="https://www.slendr.net">www.slendr.net</a>
</h4>
</center>
<p><small><br></small></p>
</section>
<section id="first-motivation-spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="first-motivation-spatial-simulations">First motivation: spatial simulations!</h2>
<center>
<img src="files/slendr/images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</section>
<section id="a-broader-motivation-for-slendr" class="level2">
<h2 class="anchored" data-anchor-id="a-broader-motivation-for-slendr">A broader motivation for <em>slendr</em></h2>
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
<div class="fragment">
<ul>
<li><p>Yet, 90% <sup><font color="blue">[citation needed]</font></sup> of simulations are basically the same!</p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify admixture rates and admixture times</p></li>
</ul></li>
<li><p>‚Ä¶ all this means duplication of code across many projects</p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Computing statistics presents even more hurdles</li>
</ul>
</div>
<div class="fragment">
<center>
<h4 class="anchored">
<em>slendr</em> makes this <u>very easy</u>, even for ‚Äúcomplex models‚Äù
</h4>
</center>
</div>
</section>
</section>
<section id="slendr-crash-course" class="level1 page-columns page-full">
<h1><em>slendr</em> crash course</h1>
<p><br> <!-- ### remove for slides  --> You can use this document as a cheat sheet as you work on the exercises. <!-- ### remove for slides  --> It contains everything you need to know about <em>slendr</em>, in a compressed form. <!-- ### remove for slides  --></p>
<section id="typical-slendr-workflow" class="level2">
<h2 class="anchored" data-anchor-id="typical-slendr-workflow">Typical <em>slendr</em> workflow</h2>
<p>We will always start our R scripts with this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr) <span class="co"># You can safely ignore any potential warnings!</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()      <span class="co"># This activates the internal Python environmet</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br>Followed by some combination of the following:</p>
<ol type="1">
<li>creating populations</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulating genomic data</li>
<li>computing popgen statistics</li>
</ol>
</section>
<section id="creating-populations" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="creating-populations">Creating populations</h2>
<p>At minimum, we need its name, size and ‚Äútime of appearance‚Äù:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p>This creates a normal R object! Typing it out gives a summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Because <em>slendr</em> uses either <em>msprime</em> or SLiM internally for simulation of genomic data, all individuals are assumed to be diploid.</p>
</div></div></section>
<section id="programming-population-splits" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are defined by providing a <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Here <code>pop1</code> is an R object created above, not a string <code>"pop1"</code>!</p>
</div></div><div class="fragment">
<p>The split is again reported in the ‚Äúhistorical summary‚Äù:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events">Scheduling resize events</h2>
<ul>
<li>Step size decrease:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<ul>
<li>Exponential increase:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb11-2"><a href="#cb11-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> <code>%&gt;%</code> interface</h2>
<p>The following leads to a more concise (and ‚Äúelegant‚Äù) code.</p>
<ul>
<li>Step size decrease:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Exponential increase:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> You can read (and understand) <code>a() %&gt;% b() %&gt;% c()</code> as ‚Äútake the result of the function <code>a</code>, pipe it into function <code>b</code>, and then pipe <em>that</em> to function <code>c</code>‚Äù.</p>
</div></div></section>
<section id="a-more-complex-model" class="level2">
<h2 class="anchored" data-anchor-id="a-more-complex-model">A more complex model</h2>
<p>Using just the two functions introduced so far:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb14-5"><a href="#cb14-5"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb14-8"><a href="#cb14-8"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb14-9"><a href="#cb14-9"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb14-10"><a href="#cb14-10"></a></span>
<span id="cb14-11"><a href="#cb14-11"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb14-12"><a href="#cb14-12"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb14-13"><a href="#cb14-13"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb14-14"><a href="#cb14-14"></a></span>
<span id="cb14-15"><a href="#cb14-15"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb14-16"><a href="#cb14-16"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb14-17"><a href="#cb14-17"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-18"><a href="#cb14-18"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="again-each-object-carries-its-history" class="level2">
<h2 class="anchored" data-anchor-id="again-each-object-carries-its-history">Again, each object carries its history!</h2>
<p>For instance, this is the summary you will get from the last population from the previous code chunk:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4 (N = 100)
  - time 900: resize from 100 to 50 individuals
  - time 1600-2200: exponential resize from 50 to 1000 individuals</code></pre>
</div>
</div>
<div class="fragment">
<p><strong>This way, you can build up complex models step by step, checking things as you go by interacting with the R console.</strong></p>
</div>
</section>
<section id="gene-flow-admixture" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="gene-flow-admixture">Gene flow / admixture</h2>
<p>We can schedule gene flow from <code>pop1</code> into <code>pop2</code> with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">2000</span>, <span class="at">end =</span> <span class="dv">2200</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Here <code>rate = 0.13</code> means 13% migrants over the given time window will come from ‚Äúpop1‚Äù into ‚Äúpop2‚Äù.</p>
</div></div><div class="fragment">
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="at">start =</span> ..., <span class="at">end =</span> ..., <span class="at">rate =</span> ...),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  ...</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="model-compilation" class="level2">
<h2 class="anchored" data-anchor-id="model-compilation">Model compilation</h2>
<p><br></p>
<p>This is the final step before we can simulate data.</p>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="at">populations =</span> <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb19-3"><a href="#cb19-3"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,       <span class="co"># (converts the all times into generations)</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span>,  <span class="co"># (number of generations to run the simulation for)</span></span>
<span id="cb19-5"><a href="#cb19-5"></a>  <span class="at">direction =</span> <span class="st">"forward"</span>      <span class="co"># (not totally necessary but good practice)</span></span>
<span id="cb19-6"><a href="#cb19-6"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong><code>compile_model()</code> takes a list of components, performs some consistency checks, and returns a single R object</strong>
</center>
</section>
<section id="model-compilation-1" class="level2">
<h2 class="anchored" data-anchor-id="model-compilation-1">Model compilation</h2>
<p><br></p>
<p>This is the final step before we can simulate data.</p>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb20-2"><a href="#cb20-2"></a>  <span class="at">populations =</span> <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb20-3"><a href="#cb20-3"></a>  <span class="at">gene_flow =</span> gf,      <span class="co"># &lt;----- in case our model includes gene flow(s)</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb20-5"><a href="#cb20-5"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span>,</span>
<span id="cb20-6"><a href="#cb20-6"></a>  <span class="at">direction =</span> <span class="st">"forward"</span></span>
<span id="cb20-7"><a href="#cb20-7"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Gene flow(s) can be included via the <code>gene_flow</code> argument.</strong>
</center>
</section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Typing the compiled <code>model</code> into R prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: 1 
generation time: 1 
time direction: forward 
time units: generations 
total running length: 3000 time units
model type: non-spatial

configuration files in: /private/var/folders/h2/qs0z_44x2vn2sskqc0cct7540000gn/T/Rtmpxr43r9/fileaa7d3ed70579_slendr_model </code></pre>
</div>
</div>
<p>This can be useful as a quick overview of the model we are working with. However, <strong>a better way to check a model is‚Ä¶</strong></p>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="a-note-on-units-of-time-and-its-direction" class="level1 page-columns page-full">
<h1>A note on units of time (and its direction)</h1>
<p><br></p>
<p>Sometimes you want to work with time units such as ‚Äúyears ago‚Äù (aDNA radio-carbon dated samples, etc.), but you have to convert those to ‚Äúgenerations forward‚Äù for some software.</p>
<p><em>slendr</em> helps by making it possible to use whatever time units or time directions are more natural for a particular project.</p>
<section id="forward-time-units" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="forward-time-units">‚ÄúForward time units‚Äù</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb24-2"><a href="#cb24-2"></a></span>
<span id="cb24-3"><a href="#cb24-3"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb24-4"><a href="#cb24-4"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb24-6"><a href="#cb24-6"></a></span>
<span id="cb24-7"><a href="#cb24-7"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb24-8"><a href="#cb24-8"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb24-9"><a href="#cb24-9"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb24-10"><a href="#cb24-10"></a></span>
<span id="cb24-11"><a href="#cb24-11"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb24-12"><a href="#cb24-12"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb24-13"><a href="#cb24-13"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb24-14"><a href="#cb24-14"></a></span>
<span id="cb24-15"><a href="#cb24-15"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb24-16"><a href="#cb24-16"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb24-17"><a href="#cb24-17"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-18"><a href="#cb24-18"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>)</span>
<span id="cb24-19"><a href="#cb24-19"></a></span>
<span id="cb24-20"><a href="#cb24-20"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb24-21"><a href="#cb24-21"></a>  <span class="at">populations =</span> <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb24-22"><a href="#cb24-22"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb24-23"><a href="#cb24-23"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span>, <span class="co"># forward-time sims need an explicit end</span></span>
<span id="cb24-24"><a href="#cb24-24"></a>  <span class="at">direction =</span> <span class="st">"forward"</span></span>
<span id="cb24-25"><a href="#cb24-25"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>We started with <code>pop1</code> in generation 1, with later events at an increasing time value.</strong></p>
</div></div></section>
<section id="forward-time-units-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="forward-time-units-1">‚ÄúForward time units‚Äù</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># see time progressing from generation 1 forwards</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>We started with <code>pop1</code> in generation 1, with later events at an increasing time value.</strong></p>
</div></div></section>
<section id="backward-time-units" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="backward-time-units">‚ÄúBackward time units‚Äù</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">30000</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb26-4"><a href="#cb26-4"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">27000</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb26-5"><a href="#cb26-5"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">20000</span>)</span>
<span id="cb26-6"><a href="#cb26-6"></a></span>
<span id="cb26-7"><a href="#cb26-7"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb26-8"><a href="#cb26-8"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">26000</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb26-9"><a href="#cb26-9"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">22000</span>)</span>
<span id="cb26-10"><a href="#cb26-10"></a></span>
<span id="cb26-11"><a href="#cb26-11"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb26-12"><a href="#cb26-12"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">25000</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb26-13"><a href="#cb26-13"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">18000</span>, <span class="at">end =</span> <span class="dv">10000</span>)</span>
<span id="cb26-14"><a href="#cb26-14"></a></span>
<span id="cb26-15"><a href="#cb26-15"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb26-16"><a href="#cb26-16"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">24000</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb26-17"><a href="#cb26-17"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">21000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-18"><a href="#cb26-18"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">14000</span>, <span class="at">end =</span> <span class="dv">8000</span>)</span>
<span id="cb26-19"><a href="#cb26-19"></a></span>
<span id="cb26-20"><a href="#cb26-20"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb26-21"><a href="#cb26-21"></a>  <span class="at">populations =</span> <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb26-22"><a href="#cb26-22"></a>  <span class="at">generation_time =</span> <span class="dv">10</span> <span class="co"># (10 time units for each generation)</span></span>
<span id="cb26-23"><a href="#cb26-23"></a>  <span class="co"># (we don't need to provide `simulation_length =` because</span></span>
<span id="cb26-24"><a href="#cb26-24"></a>  <span class="co"># "backwards" models end at time 0 by default, i.e. "present-day")</span></span>
<span id="cb26-25"><a href="#cb26-25"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Same model as before, except now expressed in units of ‚Äúyears before present‚Äù.</strong></p>
</div></div></section>
<section id="backward-time-units-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="backward-time-units-1">‚ÄúBackward time units‚Äù</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># see time progressing from "year" 30000 backwards</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Same model as before, except now expressed in units of ‚Äúyears before present‚Äù.</strong></p>
</div></div></section>
</section>
<section id="so-we-built-a-model" class="level1">
<h1>So we built a model‚Ä¶</h1>
<center>
<h3 class="anchored">
‚Ä¶ but how do we simulate data from it?
</h3>
</center>
<section id="built-in-simulation-engines" class="level2">
<h2 class="anchored" data-anchor-id="built-in-simulation-engines">Built-in simulation ‚Äúengines‚Äù</h2>
<p><em>slendr</em> has two simulation ‚Äúengine scripts‚Äù built-in:</p>
<ul>
<li><em>msprime</em> engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py"><em>slendr</em> source</a>) ‚Äì R function <code>msprime()</code></li>
<li>SLiM engine (<a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim"><em>slendr</em> source</a>) ‚Äì R function <code>slim()</code></li>
</ul>
<div class="fragment">
<p>They are designed to ‚Äúunderstand‚Äù <em>slendr</em> <code>model</code>s, meaning that you can simulate data just with this command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p><br></p>
<center>
<h3 class="anchored">
<strong>No need to write any <em>msprime</em> or SLiM code!</strong>
</h3>
</center>
</div>
</section>
</section>
<section id="the-result-of-a-simulation-is-a-tree-sequence-ts" class="level1">
<h1>The result of a simulation is a tree sequence (<code>ts</code>)</h1>
<section id="what-is-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="what-is-tree-sequence">What is tree sequence?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="files/slendr/images/tree_sequence_diagram_no_muts.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
</figure>
</div>
<ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section>
</section>
<section id="why-tree-sequence" class="level1 page-columns page-full">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3 class="anchored">
Why not VCF or a normal genotype table?
</h3>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="files/slendr/images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>An understanding of our samples‚Äô evolutionary history:</p>
<center>
<img src="files/slendr/images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<h3 class="anchored">
<strong>This is exactly what a tree sequence <em>is</em>!</strong>
</h3>
</center>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><small>Image from the <a href="https://tskit.dev/tutorials/what_is.html"><em>tskit</em> documentation</a></small></p>
</div></div></section>
<section id="the-magic-of-tree-sequences" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="the-magic-of-tree-sequences">The magic of tree sequences</h2>
<p>They allow us to compute statistics <em>without genotypes</em>!</p>
<center>
<img src="files/slendr/images/tree_sequence_diagram_no_muts.png" class="img-fluid">
</center>
<p>There is a ‚Äúduality‚Äù between mutations and branch lengths.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> See an amazing paper by <a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph <em>et al.</em> (2020)</a> for more detail.</p>
</div></div></section>
<section id="what-if-we-need-mutations-though" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though">What if we need mutations though?</h2>
<p>&nbsp;</p>
<center>
<p><img src="files/slendr/images/tree_sequence_diagram_no_muts.png" class="img-fluid"></p>
</center></section>
<section id="what-if-we-need-mutations-though-1" class="level2">
<h2 class="anchored" data-anchor-id="what-if-we-need-mutations-though-1">What if we need mutations though?</h2>
<p>Coalescent and mutation processes can be decoupled!</p>
<center>
<p><img src="files/slendr/images/tree_sequence_diagram.png" class="img-fluid"></p>
<div class="fragment">
<h3 class="anchored">
This means we can add mutations to <code>ts</code> <em>after</em> the simulation using <code>ts_mutate()</code>.
</h3>
</div>
</center>
</section>
<section id="lets-go-back-to-our-example-model" class="level2">
<h2 class="anchored" data-anchor-id="lets-go-back-to-our-example-model">Let‚Äôs go back to our example <code>model</code>‚Ä¶</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="simulate-a-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="simulate-a-tree-sequence">‚Ä¶ simulate a tree sequence‚Ä¶</h2>
<p><br></p>
<p>In our script we‚Äôll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb30-2"><a href="#cb30-2"></a><span class="fu">init_env</span>()</span>
<span id="cb30-3"><a href="#cb30-3"></a></span>
<span id="cb30-4"><a href="#cb30-4"></a><span class="co"># &lt;... population() definitions ...&gt;</span></span>
<span id="cb30-5"><a href="#cb30-5"></a></span>
<span id="cb30-6"><a href="#cb30-6"></a><span class="co"># &lt;... gene_flow() definition ...&gt;</span></span>
<span id="cb30-7"><a href="#cb30-7"></a></span>
<span id="cb30-8"><a href="#cb30-8"></a><span class="co"># &lt;... compile_model(...) ...&gt;</span></span>
<span id="cb30-9"><a href="#cb30-9"></a>  </span>
<span id="cb30-10"><a href="#cb30-10"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb30-11"><a href="#cb30-11"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="and-overlay-mutations-on-it" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="and-overlay-mutations-on-it">‚Ä¶ and overlay mutations on it</h2>
<p><br></p>
<p>In our script we‚Äôll have something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="fu">init_env</span>()</span>
<span id="cb31-3"><a href="#cb31-3"></a></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="co"># &lt;... population() definitions ...&gt;</span></span>
<span id="cb31-5"><a href="#cb31-5"></a></span>
<span id="cb31-6"><a href="#cb31-6"></a><span class="co"># &lt;... gene_flow() definition ...&gt;</span></span>
<span id="cb31-7"><a href="#cb31-7"></a></span>
<span id="cb31-8"><a href="#cb31-8"></a><span class="co"># &lt;... compile_model(...) ...&gt;</span></span>
<span id="cb31-9"><a href="#cb31-9"></a>  </span>
<span id="cb31-10"><a href="#cb31-10"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb31-11"><a href="#cb31-11"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb31-12"><a href="#cb31-12"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<strong>Note:</strong> In some exercises, mutations won‚Äôt be necessary. Where we will need them, you can use <code>ts_mutate()</code> using the pattern shown here.

</div></div></section>
</section>
<section id="so-we-can-simulate-data" class="level1">
<h1>So we can simulate data</h1>
<center>
<h3 class="anchored">
How do we work with this <code>ts</code> thing?
</h3>
</center>
<section id="slendrs-r-interface-to-tskit-statistics" class="level2">
<h2 class="anchored" data-anchor-id="slendrs-r-interface-to-tskit-statistics"><em>slendr</em>‚Äôs R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a> statistics</h2>
<center>
<img src="files/slendr/images/slendr_tskit.png" class="img-fluid">
</center>
<p>Allele-frequecy spectrum, diversity <span class="math inline">\(\pi\)</span>, <span class="math inline">\(F_{ST}\)</span>, Tajima‚Äôs D, etc.</p>
<p><strong>Find help at <a href="https://slendr.net/reference">slendr.net/reference</a> or in R under <code>?ts_fst</code> etc.</strong></p>
</section>
<section id="extracting-names-of-recorded-samples" class="level2">
<h2 class="anchored" data-anchor-id="extracting-names-of-recorded-samples">Extracting names of recorded samples</h2>
<ol type="1">
<li>We can get individuals recorded in <code>ts</code> with <code>ts_samples()</code>:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">1</span>) <span class="co"># returns a data frame (one row here, for brevity)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>    name time  pop
1 pop1_1 3001 pop1</code></pre>
</div>
</div>
<div class="fragment">
<ol start="2" type="1">
<li>A shortcut <code>ts_names()</code> can also be useful:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>) <span class="co"># returns a vector of individuals' names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3" "pop1_4" "pop1_5"</code></pre>
</div>
</div>
</div>
<div class="fragment">
<ol start="3" type="1">
<li>We can get a per-population list of individuals like this:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>) <span class="co"># returns a named list of such vectors</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$pop1
[1] "pop1_319" "pop1_298" "pop1_984" "pop1_456" "pop1_747"</code></pre>
</div>
</div>
</div>
</section>
</section>
<section id="section-4" class="level1 page-columns page-full">
<h1></h1>
<center>
<h2 class="anchored" data-anchor-id="section-4">
All <em>slendr</em> statistics take individuals‚Äô names as their function arguments.
</h2>
</center>
<p><br></p>
<center>
This is modelled after the <code>sample_sets=</code> argument of the respective <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em> Python methods</a> (except you use names of individuals directly, not tree-sequence node numbers).
</center>
<br>
<center>
Let‚Äôs take a look at how it works in general‚Ä¶
</center>
<section id="tskit-computation-option-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="tskit-computation-option-1"><em>tskit</em> computation ‚Äì option #1</h2>
<p><strong>For a function which operates on one set of individuals</strong>, we can first get a vector of names to compute on like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a random selection of names of three individuals in a tree sequence</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"popX_1"</span>, <span class="st">"popX_2"</span>, <span class="st">"popY_42"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>Then we can calculate the statistic of interest like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this computes nucleotide diversity in our set of individuals</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>df_result <span class="ot">&lt;-</span> <span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> <span class="fu">list</span>(samples))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Wherever you see <code>list(&lt;vector of names&gt;)</code>, you can think of it as <em>‚Äúcompute a statistic for the entire group of individuals‚Äù</em> (you get a single number). Without the <code>list()</code>, it would mean <em>‚Äúcompute the statistic for each individual separately‚Äù</em> (and get a value for each of them individually).</p>
</div></div></section>
<section id="tskit-computation-option-2" class="level2">
<h2 class="anchored" data-anchor-id="tskit-computation-option-2"><em>tskit</em> computation ‚Äì option #2</h2>
<p>For a function operating on multiple sets of individuals, we want a list of vectors of names (one such vector per group):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># when we compute on multiple groups, it's a good idea to name them</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">popX =</span> <span class="fu">c</span>(<span class="st">"popX_1"</span>, <span class="st">"popX_2"</span>, <span class="st">"popX_3"</span>),</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">popY =</span> <span class="fu">c</span>(<span class="st">"popY_1"</span>, <span class="st">"popY_2"</span>, <span class="st">"popY_3"</span>),</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">popZ =</span> <span class="fu">c</span>(<span class="st">"popZ_1"</span>, <span class="st">"popZ_2"</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Then we use this list of vectors in the same way as before:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this computes a pairwise divergence between all three groups</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>df_result <span class="ot">&lt;-</span> <span class="fu">ts_divergence</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tskit-computation-option-3" class="level2">
<h2 class="anchored" data-anchor-id="tskit-computation-option-3"><em>tskit</em> computation ‚Äì option #3</h2>
<p>For something like <span class="math inline">\(f\)</span> statistics, the function arguments must be more precisely specified (here <code>A</code>, <code>B</code>, <code>C</code>, not <code>sample_sets</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>df_result <span class="ot">&lt;-</span> <span class="fu">ts_f3</span>(</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  ts,</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">A =</span> <span class="fu">c</span>(<span class="st">"popX_1"</span>, <span class="st">"popX_2"</span>, <span class="st">"popX_3"</span>),</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">B =</span> <span class="fu">c</span>(<span class="st">"popY_1"</span>, <span class="st">"popY_2"</span>, <span class="st">"popY_3"</span>),</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">C =</span> <span class="fu">c</span>(<span class="st">"popZ_1"</span>, <span class="st">"popZ_2"</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p>Doing this manually can be annoying ‚Äî <code>ts_names()</code> helps by preparing the list of names in the correct format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get names of individuals in each population as a named list of vectors</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># use this list directly by specifying which vectors to take out</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">A =</span> samples<span class="sc">$</span>popX, <span class="at">B =</span> samples<span class="sc">$</span>popY, <span class="at">C =</span> samples<span class="sc">$</span>popZ)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
<section id="section-5" class="level1 page-columns page-full">
<h1></h1>
<center>
<h1>
Some examples on simulated data
</h1>
<br>
<h3 class="anchored">
(A tree sequence <code>ts</code> we got earlier.)
</h3>
</center>
<section id="example-nucleotide-diversity" class="level2">
<h2 class="anchored" data-anchor-id="example-nucleotide-diversity">Example: <a href="https://en.wikipedia.org/wiki/Nucleotide_diversity">nucleotide diversity</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Get a list of individuals in each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1" "pop2"</code></pre>
</div>
</div>
<div class="fragment">
<p><br></p>
<p>We can index into the list via population name:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop2_1" "pop2_2" "pop2_3"</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<p>Compute nucleotide diversity (note the list <code>samples</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 pop1  0.000391 
2 pop2  0.0000632</code></pre>
</div>
</div>
<p><br></p>
<p>Our tree sequence had two populations, <code>pop1</code> and <code>pop2</code>, which is why we get a data frame with diversity in each of them.</p>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum">Example: <a href="https://en.wikipedia.org/wiki/Allele_frequency_spectrum">allele frequency spectrum</a></h2>
<div class="columns">
<div class="column" style="width:45%;">
<p>Get names of individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop_1" "pop_2" "pop_3" "pop_4" "pop_5"</code></pre>
</div>
</div>
<div class="fragment">
<p>Compute the AFS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(ts, <span class="at">sample_sets =</span> <span class="fu">list</span>(samples))</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we skip the 1st item because it has a special meaning in tskit</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>afs[<span class="sc">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 3917 2151 1432  941  740  624  607  587  416  385</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:53%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs[<span class="sc">-</span><span class="dv">1</span>], <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> One of the rare examples when a <em>slendr</em> / <em>tskit</em> statistical function does not return a data frame (<code>ts_afs()</code> returns a numerical vector, not a data frame).</p>
</div></div></section>
</section>
<section id="section-6" class="level1 page-columns page-full">
<h1></h1>
<h2 class="anchored" data-anchor-id="section-6">
More details on tree-sequences
</h2>
<section id="tree-sequence-tables" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables">Tree sequence tables</h2>
<p><br></p>
<p>This simulates 2 <span class="math inline">\(\times\)</span> 10000 chromosomes of 100 Mb:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>(<span class="at">quiet =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">time =</span> <span class="fl">100e6</span>, <span class="at">N =</span> <span class="dv">10000</span>)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">30</span>, <span class="at">direction =</span> <span class="st">"backward"</span>)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
<p><br></p>
<p><strong>Runs in less than 30 seconds on my laptop!</strong></p>
<p><strong>Takes only about 66 Mb of memory!</strong></p>
</section>
<section id="how-is-this-even-possible" class="level2">
<h2 class="anchored" data-anchor-id="how-is-this-even-possible">How is this even possible?!</h2>
<center>
<p><img src="files/slendr/images/tables.jpeg" class="img-fluid"></p>
<center>
</center></center></section>
<section id="tree-sequence-tables-1" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-1">Tree-sequence tables</h2>
<div class="row">
<div class="columns">
<div class="column" style="width:60%;">
<p>A tree can be represented by</p>
<div class="incremental">
<ul class="incremental">
<li><p>a table of <font color="orange">n</font><font color="green">o</font><font color="darkblue">d</font><font color="green">e</font><font color="darkblue">s</font>,</p></li>
<li><p>a table of <u>edges</u> between nodes,</p></li>
<li><p>a table of <font color="red">mutations</font> on edges</p></li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<p><br></p>
<center>
<img src="files/slendr/images/tree_diagram.png" class="img-fluid">
</center>
</div>
</div>
</div>
<h3 class="anchored">
<strong>A collection of such tables is a tree sequence.</strong>
</h3>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> This is a huge oversimplification. Find more information in <a href="https://tskit.dev/tutorials/tables_and_editing.html">tskit docs</a>.</p>
</div></div></section>
<section id="tree-sequence-tables-in-practice" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-in-practice">Tree-sequence tables in practice</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="handouts_slendr_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="columns">
<div class="column" style="width:30%;">
<div class="fragment">
<p><strong>Nodes:</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  node     time
1  360 871895.1
2  256 475982.3
3  255 471179.5</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:30%;">
<div class="fragment">
<p><strong>Edges:</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  child parent
1   256    360
2   255    256
3    69    256</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:30%;">
<div class="fragment">
<p><strong>Mutations:</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>   id node     time
1  69   74 125539.4
2 272   22 242337.9
3 277   22 129474.1</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="other-data-formats" class="level1">
<h1>Other data formats</h1>
<p><br></p>
<p>Tree sequence is a useful, cutting-edge data structure, but there are many well-established bioinformatics tools out there.</p>
<p><br></p>
<p><em>tskit</em> and <em>slendr</em> offer a couple of functions to help integrate their simulation results into third-party software.</p>
<section id="standard-genotype-formats" class="level2">
<h2 class="anchored" data-anchor-id="standard-genotype-formats">Standard genotype formats</h2>
<p>If a tree sequence doesn‚Äôt cut it, you can always‚Ä¶</p>
<div class="fragment">
<ul>
<li>export genotypes to a VCF file:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<ul>
<li>export genotypes in the EIGENSTRAT format:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<ul>
<li>access genotypes as a data frame:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>    pos pop_1_chr1 pop_1_chr2 pop_2_chr1 pop_2_chr2 pop_3_chr1 pop_3_chr2
1 12977          0          1          0          0          0          0
2 15467          1          0          1          1          1          1</code></pre>
</div>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./handouts.html" class="pagination-link" aria-label="Handouts">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Handouts</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>