<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Simulations with slendr – Simulation-Based Population Genomics and Inference in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./intro.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-626149efe8f5d16e1d391ba177679bf0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./slendr.html">Simulations with <em>slendr</em></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Simulation-Based Population Genomics and Inference in R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Simulations with <em>slendr</em></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#installation-and-setup" id="toc-installation-and-setup" class="nav-link active" data-scroll-target="#installation-and-setup">Installation and setup</a></li>
  <li><a href="#organization-of-the-exercises" id="toc-organization-of-the-exercises" class="nav-link" data-scroll-target="#organization-of-the-exercises">Organization of the exercises</a></li>
  <li><a href="#programming-demographic-models-with-slendr" id="toc-programming-demographic-models-with-slendr" class="nav-link" data-scroll-target="#programming-demographic-models-with-slendr">Programming demographic models with <em>slendr</em></a>
  <ul class="collapse">
  <li><a href="#part-1-building-a-demographic-model-in-slendr" id="toc-part-1-building-a-demographic-model-in-slendr" class="nav-link" data-scroll-target="#part-1-building-a-demographic-model-in-slendr">Part 1: Building a demographic model in <em>slendr</em></a></li>
  <li><a href="#part-2-inspecting-the-model-visually" id="toc-part-2-inspecting-the-model-visually" class="nav-link" data-scroll-target="#part-2-inspecting-the-model-visually">Part 2: Inspecting the model visually</a></li>
  <li><a href="#part-3-simulating-genomic-data" id="toc-part-3-simulating-genomic-data" class="nav-link" data-scroll-target="#part-3-simulating-genomic-data">Part 3: Simulating genomic data</a></li>
  <li><a href="#part-4-inspecting-the-tree-sequence-object" id="toc-part-4-inspecting-the-tree-sequence-object" class="nav-link" data-scroll-target="#part-4-inspecting-the-tree-sequence-object">Part 4: Inspecting the tree-sequence object</a></li>
  <li><a href="#part-5-scheduling-sampling-events" id="toc-part-5-scheduling-sampling-events" class="nav-link" data-scroll-target="#part-5-scheduling-sampling-events">Part 5: Scheduling sampling events</a></li>
  <li><a href="#part-6-simulating-a-defined-set-of-individuals" id="toc-part-6-simulating-a-defined-set-of-individuals" class="nav-link" data-scroll-target="#part-6-simulating-a-defined-set-of-individuals">Part 6: Simulating a defined set of individuals</a></li>
  </ul></li>
  <li><a href="#computing-popgen-statistics-on-tree-sequences-from-slendr" id="toc-computing-popgen-statistics-on-tree-sequences-from-slendr" class="nav-link" data-scroll-target="#computing-popgen-statistics-on-tree-sequences-from-slendr">Computing popgen statistics on tree sequences from <em>slendr</em></a>
  <ul class="collapse">
  <li><a href="#part-1-computing-nucleotide-diversity" id="toc-part-1-computing-nucleotide-diversity" class="nav-link" data-scroll-target="#part-1-computing-nucleotide-diversity">Part 1: Computing nucleotide diversity</a></li>
  <li><a href="#part-2-computing-pairwise-divergence" id="toc-part-2-computing-pairwise-divergence" class="nav-link" data-scroll-target="#part-2-computing-pairwise-divergence">Part 2: Computing pairwise divergence</a></li>
  <li><a href="#part-3-detecting-neanderthal-admixture-in-europeans" id="toc-part-3-detecting-neanderthal-admixture-in-europeans" class="nav-link" data-scroll-target="#part-3-detecting-neanderthal-admixture-in-europeans">Part 3: Detecting Neanderthal admixture in Europeans</a></li>
  <li><a href="#part-4-detecting-neanderthal-admixture-in-europeans-v2.0" id="toc-part-4-detecting-neanderthal-admixture-in-europeans-v2.0" class="nav-link" data-scroll-target="#part-4-detecting-neanderthal-admixture-in-europeans-v2.0">Part 4: Detecting Neanderthal admixture in Europeans v2.0</a></li>
  </ul></li>
  <li><a href="#simulation-based-inference-of-n_e" id="toc-simulation-based-inference-of-n_e" class="nav-link" data-scroll-target="#simulation-based-inference-of-n_e">Simulation-based inference of <span class="math inline">\(N_e\)</span></a>
  <ul class="collapse">
  <li><a href="#part-1-a-self-contained-slendr-function-of-n_e-rightarrow-textrmafs" id="toc-part-1-a-self-contained-slendr-function-of-n_e-rightarrow-textrmafs" class="nav-link" data-scroll-target="#part-1-a-self-contained-slendr-function-of-n_e-rightarrow-textrmafs">Part 1: A self-contained <em>slendr</em> function of <span class="math inline">\(N_e \rightarrow \textrm{AFS}\)</span></a></li>
  <li><a href="#part-2-estimating-unknown-n_e-from-empirical-afs" id="toc-part-2-estimating-unknown-n_e-from-empirical-afs" class="nav-link" data-scroll-target="#part-2-estimating-unknown-n_e-from-empirical-afs">Part 2: Estimating unknown <span class="math inline">\(N_e\)</span> from empirical AFS</a></li>
  </ul></li>
  <li><a href="#simulating-and-detecting-natural-selection" id="toc-simulating-and-detecting-natural-selection" class="nav-link" data-scroll-target="#simulating-and-detecting-natural-selection">Simulating (and detecting) natural selection</a>
  <ul class="collapse">
  <li><a href="#part-1-simulating-a-tree-sequence-and-computing-tajimas-d" id="toc-part-1-simulating-a-tree-sequence-and-computing-tajimas-d" class="nav-link" data-scroll-target="#part-1-simulating-a-tree-sequence-and-computing-tajimas-d">Part 1: Simulating a tree sequence and computing Tajima’s D</a></li>
  <li><a href="#part-2-computing-tajimas-d-in-windows" id="toc-part-2-computing-tajimas-d-in-windows" class="nav-link" data-scroll-target="#part-2-computing-tajimas-d-in-windows">Part 2: Computing Tajima’s D in windows</a></li>
  <li><a href="#part-3-adding-positive-selection-to-the-base-demographic-model" id="toc-part-3-adding-positive-selection-to-the-base-demographic-model" class="nav-link" data-scroll-target="#part-3-adding-positive-selection-to-the-base-demographic-model">Part 3: Adding positive selection to the base demographic model</a></li>
  <li><a href="#part-4-running-a-selection-simulation-using-slim" id="toc-part-4-running-a-selection-simulation-using-slim" class="nav-link" data-scroll-target="#part-4-running-a-selection-simulation-using-slim">Part 4: Running a selection simulation using <code>slim()</code></a></li>
  <li><a href="#part-5-investigating-allele-frequency-trajectories" id="toc-part-5-investigating-allele-frequency-trajectories" class="nav-link" data-scroll-target="#part-5-investigating-allele-frequency-trajectories">Part 5: Investigating allele frequency trajectories</a></li>
  <li><a href="#part-6-tajimas-d-genome-wide-and-window-based-from-the-selection-model" id="toc-part-6-tajimas-d-genome-wide-and-window-based-from-the-selection-model" class="nav-link" data-scroll-target="#part-6-tajimas-d-genome-wide-and-window-based-from-the-selection-model">Part 6: Tajima’s D (genome-wide and window-based) from the selection model</a></li>
  </ul></li>
  <li><a href="#playing-around-with-pca-patterns" id="toc-playing-around-with-pca-patterns" class="nav-link" data-scroll-target="#playing-around-with-pca-patterns">Playing around with PCA patterns</a>
  <ul class="collapse">
  <li><a href="#part-1-create-a-model" id="toc-part-1-create-a-model" class="nav-link" data-scroll-target="#part-1-create-a-model">Part 1: Create a model</a></li>
  <li><a href="#part-2-simulate-a-mutated-tree-sequence" id="toc-part-2-simulate-a-mutated-tree-sequence" class="nav-link" data-scroll-target="#part-2-simulate-a-mutated-tree-sequence">Part 2: Simulate a (mutated!) tree sequence</a></li>
  <li><a href="#part-3-converting-a-tree-sequence-into-eigenstrat" id="toc-part-3-converting-a-tree-sequence-into-eigenstrat" class="nav-link" data-scroll-target="#part-3-converting-a-tree-sequence-into-eigenstrat">Part 3: Converting a tree sequence into EIGENSTRAT</a></li>
  <li><a href="#part-4-inspect-the-eigenstrat-data-produced-by-slendr" id="toc-part-4-inspect-the-eigenstrat-data-produced-by-slendr" class="nav-link" data-scroll-target="#part-4-inspect-the-eigenstrat-data-produced-by-slendr">Part 4: Inspect the EIGENSTRAT data produced by <em>slendr</em></a></li>
  <li><a href="#part-5-principal-component-analysis-on-the-entire-simulated-data-set" id="toc-part-5-principal-component-analysis-on-the-entire-simulated-data-set" class="nav-link" data-scroll-target="#part-5-principal-component-analysis-on-the-entire-simulated-data-set">Part 5: Principal Component Analysis on the entire simulated data set</a></li>
  </ul></li>
  <li><a href="#spatial-pca" id="toc-spatial-pca" class="nav-link" data-scroll-target="#spatial-pca">Spatial PCA</a>
  <ul class="collapse">
  <li><a href="#per-population-ne-values" id="toc-per-population-ne-values" class="nav-link" data-scroll-target="#per-population-ne-values">Per-population Ne values</a></li>
  </ul></li>
  <li><a href="#ancestry-tracts-and-chromosome-painintg" id="toc-ancestry-tracts-and-chromosome-painintg" class="nav-link" data-scroll-target="#ancestry-tracts-and-chromosome-painintg">Ancestry tracts and chromosome painintg</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Simulations with <em>slendr</em></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- TODO: Looks at these links to define a custom "Show the solution" callout -->
<!-- https://www.youtube.com/watch?v=DDQO_3R-q74 -->
<!-- https://examples.quarto.pub/collapse-callout/ -->
<section id="installation-and-setup" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="installation-and-setup">Installation and setup</h2>
<p>The easiest way to set up everything on your computer is to download all of the book materials (text, code, data, everything) to your machine by following these steps:</p>
<ol type="1">
<li><p><strong>Clone the repository with the course materials</strong>. In a shell terminal on Linux or macOS, in your home directory (or anywhere else, really) you can run:</p>
<pre><code>$ git clone https://github.com/bodkan/simgen ~/simgen</code></pre></li>
</ol>
<p>(For Windows users only: You should be able to do all of this on Windows too, but that would require you having some kind of Linux environment (terminal, R, etc.) already set up. Unfortunately I don’t know much about that so my recommendation is that, if you do use Windows, just do the exercises on the cloud instance provided by the course organizers.)</p>
<ol start="2" type="1">
<li><strong>Install all R package dependencies</strong> by going into the activity repository directory you just cloned and installing the necessary R packages by following these steps:</li>
</ol>
<ul>
<li><p>First <strong>go into the project directory you just cloned</strong>:</p>
<pre><code>$ cd ~/simgen</code></pre></li>
<li><p><strong>Open the R terminal in that directory</strong>. You should get a note that the <em>renv</em> package is being automatically setup, like this:</p>
<pre><code>$ R

[... R startup information stuff ...]

# Bootstrapping renv 1.0.11 --------------------------------------------------
- Downloading renv ... OK
- Installing renv  ... OK

- Project '~/simgen' loaded. [renv 1.0.11]
- One or more packages recorded in the lockfile are not installed.
- Use `renv::status()` for more details.</code></pre></li>
<li><p><strong>Install R package dependencies</strong> (still in the R console!):</p>
<pre><code>&gt; renv::restore(prompt = FALSE)</code></pre></li>
<li><p><strong>Set up the Python environment</strong> used by the <em>slendr</em> R package for simulation and tree-sequence analysis (still in the R console!):</p>
<pre><code>&gt; slendr::setup_env(agree = TRUE)</code></pre>
<p>If everything worked, you should get an optimistic message saying:</p>
<pre><code>======================================================================
Python environment for slendr has been successfuly created, and the R
interface to msprime, tskit, and pyslim modules has been activated.

In future sessions, activate this environment by calling init_env().
=======================================================================</code></pre></li>
</ul>
<ol start="4" type="1">
<li><strong>Open the repository directory in RStudio</strong> (on your laptop if you cloned the repository there, or in your browser if you’re using the provided cloud instance) and navigate to the book repository directory via <code>File -&gt; Open Project...</code> in the RStudio menu, opening the <code>simgen.Rproj</code> file.</li>
</ol>
<p><strong>If the <code>setup_env()</code> installation procedure fails, try the following:</strong></p>
<ol type="1">
<li>Delete the failed installation attempt:</li>
</ol>
<pre><code>slendr::clear_env(force = TRUE, all = TRUE)</code></pre>
<ol start="2" type="1">
<li>Try installing it again, this time using <code>pip</code> as a Python installation method (the default is <code>conda</code> which unfortunately fails fairly often):</li>
</ol>
<pre><code>slendr::setup_env(agree = TRUE, pip = TRUE)</code></pre>
</section>
<section id="organization-of-the-exercises" class="level1 unnumbered">
<h1 class="unnumbered">Organization of the exercises</h1>
<p>For each exercise, you will get a brief explanation of the problem at hand and some information about functions that could be useful to solve the exercise. <strong>The concrete, specific task will be always written like this in bold. As you work on each part of each exercises, look for these bold lines.</strong></p>
<p><strong>Your goal for each exercise is to write a complete R script script (in RStudio <code>File -&gt; New file -&gt; R script</code>). I suggest you save the script for each exercise as <code>exercise1.R</code>, <code>exercise2.R</code>, etc., just to keep things tidy and easy to troubleshoot if needed.</strong></p>
<p><strong>Each exercise is composed of individual <em>parts</em>, which are designed to build one upon the other in the order they are specified. As you progress through sequential parts of each exercise, you will be adding code to your script for that exercise.</strong></p>
<section id="note-on-the-programming-aspect-of-the-exercises" class="level4">
<h4 class="anchored" data-anchor-id="note-on-the-programming-aspect-of-the-exercises">Note on the programming aspect of the exercises</h4>
<p>All the exercises will involve “real coding”! If you’ve never really programmed entire scripts before, this could feel a little intimidating. Don’t worry. If you’re ever lost, just take a peek at the solutions which are (by default hidden) under each exercise part. Always try to work on a solution on your own, but never let this be a barrier to your progress. Feel free to copy-paste bits of my solutions into your own scripts.</p>
<p>If you find yourself totally lost, don’t hesitate to read my solutions from the get go, copy-pasting them into your own script in the RStudio, executing them line by line, and trying to understand what’s going on. <strong>It’s the understanding that we’re after here. Whether or not you can implement everything yourself from scratch does not actually matter at all.</strong> (Nobody in history has learned to program from scratch. Everyone started by copy-pasting code examples written by someone else. So feel free to do the same!)</p>
</section>
</section>
<section id="programming-demographic-models-with-slendr" class="level1 page-columns page-full">
<h1>Programming demographic models with <em>slendr</em></h1>
<section id="part-1-building-a-demographic-model-in-slendr" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-1-building-a-demographic-model-in-slendr">Part 1: Building a demographic model in <em>slendr</em></h2>
<p><strong>Use functions such as <code>population()</code>, <code>gene_flow()</code>, and <code>compile_model()</code>, which we discussed in the “crash course” at the start of this session, to program the following toy model of human demographic history in <em>slendr</em>.</strong> (Apologies for my bad handwriting and poor artistic ability.)</p>
<p><img src="images/intro_model1.png" class="img-fluid" style="width:50.0%"></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> You could easily program the model so that different ancestral populations are represented by separate <code>population()</code> commands (i.e., your model would start with a population called “human_chimp_ancestor” from which a “CHIMP” and “hominin_ancestor” populations would split at 6 Mya, etc.) but generally this is too annoying to do and requires too much code.</p>
<p>Feel free to write the model so that “CHIMP” is the first population, then “AFR” population splits from it at 6 Mya, etc… Although it probably isn’t the most accurate way to describe the real evolutionary history, it simplifies the coding significantly.</p>
<p><br> [Mya = million years ago; kya = thousand years ago]</p>
</div></div><p><strong>Hint:</strong> Create a new script <code>exercise1.R</code> in your RStudio session using the following “template”. Then add a sequence of appropriate <code>population()</code> calls using the syntax from the introductory slides (using the <code>parent = &lt;pop&gt;</code> argument for programming splits of daughter populations – which will be all except the CHIMP lineage in our example), etc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;replace this with `population()` definitions like in the slides&gt;</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;replace this with your gene-flow definition in variable `gf`&gt;</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(...), <span class="co"># &lt;put your list of populations here&gt;</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> With <em>slendr</em> you can specify time in whatever format is more convenient or readable for your model. For instance here, because we’re dealing with historical events which are commonly expressed in times given as”years ago”, we can write them in a decreasing order – i.e.&nbsp;7Mya → 6Mya → …, as shown above – or, in terms of R code, 7e6 (or 7000000), 6e6 (6000000), etc.</p>
<p>In a later example, you will see that you can also encode the events in the time direction going “forward” (i.e., the first event starting in generation 1, a following event in generation 42, and so on).</p>
</div></div><p><strong>Hint:</strong> Remember that <em>slendr</em> is designed with interactivity in mind! When you write a chunk of code (such as a command to create a population through a population split, or model compilation to create a <code>model</code> object), execute that bit of code in the R console and inspect the summary information printed by evaluating the respective R object you just created. You can either copy-pasted stuff from your script to the R console, or use a convenient RStudio shortcut like Ctrl+Enter (Linux and Windows), or Cmd+Enter (Mac).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="do">## The interface to all required Python modules has been activated.</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Chimpanzee outgroup</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"CHIMP"</span>, <span class="at">time =</span> <span class="fl">7e6</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Two populations of anatomically modern humans: Africans and Europeans</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>afr <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"AFR"</span>, <span class="at">parent =</span> chimp, <span class="at">time =</span> <span class="fl">6e6</span>, <span class="at">N =</span> <span class="dv">15000</span>)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>eur <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"EUR"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">70e3</span>, <span class="at">N =</span> <span class="dv">3000</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Neanderthal population splitting at 600 ky ago from modern humans</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (becomes extinct by 40 ky ago)</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>nea <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"NEA"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">600e3</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">remove =</span> <span class="fl">40e3</span>)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Neanderthal introgression event (3% admixture between 55-50 kya)</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> nea, <span class="at">to =</span> eur, <span class="at">rate =</span> <span class="fl">0.03</span>, <span class="at">start =</span> <span class="dv">55000</span>, <span class="at">end =</span> <span class="dv">50000</span>)</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Compile the entire model into a single slendr R object</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, nea, afr, eur),</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf,</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span>,</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/introgression"</span>),  <span class="co"># &lt;--- don't worry about these two</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite =</span> <span class="cn">TRUE</span>, <span class="at">force =</span> <span class="cn">TRUE</span>            <span class="co"># &lt;--- lines of code (ask me if interested)</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-2-inspecting-the-model-visually" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-2-inspecting-the-model-visually">Part 2: Inspecting the model visually</h2>
<p>To visualize a <em>slendr</em> model, you use the function <code>plot_model()</code>. <strong>Plot your compiled <code>model</code> to make sure you programmed it correctly!</strong> Your figure should roughly correspond to my doodle above.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Plotting of models in <em>slendr</em> can be sometimes a little wonky, especially if many things are happening at once. When plotting your model, experiment with arguments <code>log = TRUE</code>, <code>proportions = TRUE</code>, <code>gene_flow = TRUE</code>. Check <code>?plot_model</code> for more information on these.</p>
</div></div><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">sizes =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">sizes =</span> <span class="cn">FALSE</span>, <span class="at">log =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">log =</span> <span class="cn">TRUE</span>, <span class="at">proportions =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-3-simulating-genomic-data" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-3-simulating-genomic-data">Part 3: Simulating genomic data</h2>
<p>Once you have a compiled <em>slendr</em> model stored in an R variable (from now on, <code>model</code> will always mean a variable containing a compiled <em>slendr</em> model object relevant for the given exercise, for simplicity), we can simulate data from it. By default, <em>slendr</em> models always produce a <a href="https://tskit.dev/tutorials/what_is.html">tree sequence</a>.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Tree sequence provides an extremely efficient means to store and work with genomic data at a massive scale. However, you can always get simulated data even in <a href="https://www.slendr.net/reference/index.html#tree-sequence-format-conversion">traditional file formats</a>, such as VCF, EIGENSTRAT, or a plain old table of ancestral/derived genotypes.</p>
<p>In this activity we will be only working with tree sequences, because it’s much easier and faster to get interesting statistics from it directly in R.</p>
</div></div><p>There are two simulation engines built into <em>slendr</em> implemented by functions <code>msprime()</code> and <code>slim()</code>. For traditional, non-spatial, neutral demographic models, the engine provided by the <code>msprime()</code> function is much more efficient, so we’ll be using that for the time being. However, from a popgen theoretical perspective, both simulation functions will give you the same results for any given compiled <em>slendr</em> model (up to some level of stochastic noise, of course).</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Yes, this means you don’t have to write any <em>msprime</em> (or SLiM) code to simulate data from a <em>slendr</em> model!</p>
</div></div><p>Here’s how you can use the function to simulate a tree sequence from the model you’ve just created using <code>compile_model()</code> in your script:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequence_length =</span> <span class="sc">&lt;</span>length of sequence to simulate [as bp]<span class="sc">&gt;</span>,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recombination_rate =</span> <span class="sc">&lt;</span>uniform recombination rate [per bp per generation]<span class="sc">&gt;</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will be seeing this kind of pattern over and over again in this exercise, so it’s a good idea to keep it in mind.</p>
<p><strong>Hint:</strong> The <code>msprime()</code> function has also arguments <code>debug</code> and <code>run</code> which can be extremely useful for debugging.</p>
<p><strong>Simulate a tree sequence from your compiled <code>model</code> using the <code>msprime()</code> engine, storing it to a variable <code>ts</code> as shown right above. Use <code>sequence_length = 1e6</code> (so 1 Mb of sequence) and <code>recombination_rate = 1e-8</code> (crossover events per base pair per generation). Then experiment with setting <code>debug = TRUE</code> (this prints out <em>msprime</em>’s own debugging summary which you might already be familiar with from your previous activity?) and then <code>run = FALSE</code> (this prints out a raw command-line which can run a <em>slendr</em> simulation in the shell).</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This simulates a tskit tree sequence from a slendr model. Note that you didn't have</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to write any msprime or tskit Python code!</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Setting `debug = TRUE` instructs slendr's built-in msprime script to print</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co"># out msprime's own debugger information. This can be very useful for debugging,</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co"># in addition to the visualization of the model as shown above.</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">debug =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="do">## DemographyDebugger</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠═════════════════════════════════════╗</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[0]: [0, 1.67e+03) generations ║</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠═════════════════════════════════════╝</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=4)</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌─────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │     start│       end│growth_rate  │ CHIMP │ AFR │ NEA │ EUR │</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├─────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│    5000.0│    5000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    AFR│   15000.0│   15000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    NEA│    1000.0│    1000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    EUR│    3000.0│    3000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └─────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Events @ generation 1.67e+03</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌─────────────────────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  time│type            │parameters              │effect                                  │</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├─────────────────────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  1666│Migration rate  │source=EUR, dest=NEA,   │Backwards-time migration rate from EUR  │</span></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │change          │rate=0.000179640718562  │to NEA → 0.00017964071856287425         │</span></span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │                │87425                   │                                        │</span></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └─────────────────────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════════════╗</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[1]: [1.67e+03, 1.83e+03) generations ║</span></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════════════╝</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=4)</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌───────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │     start│       end│growth_rate  │ CHIMP │ AFR │    NEA    │ EUR │</span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├───────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│    5000.0│    5000.0│ 0           │   0   │  0  │     0     │  0  │</span></span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    AFR│   15000.0│   15000.0│ 0           │   0   │  0  │     0     │  0  │</span></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    NEA│    1000.0│    1000.0│ 0           │   0   │  0  │     0     │  0  │</span></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    EUR│    3000.0│    3000.0│ 0           │   0   │  0  │ 0.0001796 │  0  │</span></span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └───────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Events @ generation 1.83e+03</span></span>
<span id="cb16-43"><a href="#cb16-43" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌────────────────────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-44"><a href="#cb16-44" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  time│type            │parameters             │effect                                  │</span></span>
<span id="cb16-45"><a href="#cb16-45" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├────────────────────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-46"><a href="#cb16-46" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  1833│Migration rate  │source=EUR, dest=NEA,  │Backwards-time migration rate from EUR  │</span></span>
<span id="cb16-47"><a href="#cb16-47" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │change          │rate=0                 │to NEA → 0                              │</span></span>
<span id="cb16-48"><a href="#cb16-48" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈│</span></span>
<span id="cb16-49"><a href="#cb16-49" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  1833│Census          │                       │Insert census nodes to record the       │</span></span>
<span id="cb16-50"><a href="#cb16-50" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │                │                       │location of all lineages                │</span></span>
<span id="cb16-51"><a href="#cb16-51" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └────────────────────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-52"><a href="#cb16-52" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════════════╗</span></span>
<span id="cb16-53"><a href="#cb16-53" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[2]: [1.83e+03, 2.33e+03) generations ║</span></span>
<span id="cb16-54"><a href="#cb16-54" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════════════╝</span></span>
<span id="cb16-55"><a href="#cb16-55" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=4)</span></span>
<span id="cb16-56"><a href="#cb16-56" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌─────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-57"><a href="#cb16-57" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │     start│       end│growth_rate  │ CHIMP │ AFR │ NEA │ EUR │</span></span>
<span id="cb16-58"><a href="#cb16-58" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├─────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-59"><a href="#cb16-59" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│    5000.0│    5000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-60"><a href="#cb16-60" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    AFR│   15000.0│   15000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-61"><a href="#cb16-61" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    NEA│    1000.0│    1000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-62"><a href="#cb16-62" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    EUR│    3000.0│    3000.0│ 0           │   0   │  0  │  0  │  0  │</span></span>
<span id="cb16-63"><a href="#cb16-63" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └─────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-64"><a href="#cb16-64" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Events @ generation 2.33e+03</span></span>
<span id="cb16-65"><a href="#cb16-65" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌───────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-66"><a href="#cb16-66" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  time│type        │parameters      │effect                                │</span></span>
<span id="cb16-67"><a href="#cb16-67" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├───────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-68"><a href="#cb16-68" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  2333│Population  │derived=[EUR],  │Moves all lineages from the 'EUR'     │</span></span>
<span id="cb16-69"><a href="#cb16-69" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │Split       │ancestral=AFR   │derived population to the ancestral   │</span></span>
<span id="cb16-70"><a href="#cb16-70" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │            │                │'AFR' population. Also set 'EUR' to   │</span></span>
<span id="cb16-71"><a href="#cb16-71" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │            │                │inactive, and all migration rates to  │</span></span>
<span id="cb16-72"><a href="#cb16-72" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │            │                │and from the derived population to    │</span></span>
<span id="cb16-73"><a href="#cb16-73" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │      │            │                │zero.                                 │</span></span>
<span id="cb16-74"><a href="#cb16-74" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └───────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-75"><a href="#cb16-75" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠═════════════════════════════════════════╗</span></span>
<span id="cb16-76"><a href="#cb16-76" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[3]: [2.33e+03, 2e+04) generations ║</span></span>
<span id="cb16-77"><a href="#cb16-77" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠═════════════════════════════════════════╝</span></span>
<span id="cb16-78"><a href="#cb16-78" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=3)</span></span>
<span id="cb16-79"><a href="#cb16-79" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌───────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-80"><a href="#cb16-80" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │     start│       end│growth_rate  │ CHIMP │ AFR │ NEA │</span></span>
<span id="cb16-81"><a href="#cb16-81" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├───────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-82"><a href="#cb16-82" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│    5000.0│    5000.0│ 0           │   0   │  0  │  0  │</span></span>
<span id="cb16-83"><a href="#cb16-83" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    AFR│   15000.0│   15000.0│ 0           │   0   │  0  │  0  │</span></span>
<span id="cb16-84"><a href="#cb16-84" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    NEA│    1000.0│    1000.0│ 0           │   0   │  0  │  0  │</span></span>
<span id="cb16-85"><a href="#cb16-85" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └───────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-86"><a href="#cb16-86" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Events @ generation 2e+04</span></span>
<span id="cb16-87"><a href="#cb16-87" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌────────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-88"><a href="#cb16-88" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │   time│type        │parameters      │effect                                │</span></span>
<span id="cb16-89"><a href="#cb16-89" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├────────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-90"><a href="#cb16-90" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  2e+04│Population  │derived=[NEA],  │Moves all lineages from the 'NEA'     │</span></span>
<span id="cb16-91"><a href="#cb16-91" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │Split       │ancestral=AFR   │derived population to the ancestral   │</span></span>
<span id="cb16-92"><a href="#cb16-92" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                │'AFR' population. Also set 'NEA' to   │</span></span>
<span id="cb16-93"><a href="#cb16-93" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                │inactive, and all migration rates to  │</span></span>
<span id="cb16-94"><a href="#cb16-94" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                │and from the derived population to    │</span></span>
<span id="cb16-95"><a href="#cb16-95" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                │zero.                                 │</span></span>
<span id="cb16-96"><a href="#cb16-96" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └────────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-97"><a href="#cb16-97" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠══════════════════════════════════════╗</span></span>
<span id="cb16-98"><a href="#cb16-98" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[4]: [2e+04, 2e+05) generations ║</span></span>
<span id="cb16-99"><a href="#cb16-99" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠══════════════════════════════════════╝</span></span>
<span id="cb16-100"><a href="#cb16-100" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=2)</span></span>
<span id="cb16-101"><a href="#cb16-101" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌─────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-102"><a href="#cb16-102" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │     start│       end│growth_rate  │ CHIMP │ AFR │</span></span>
<span id="cb16-103"><a href="#cb16-103" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├─────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-104"><a href="#cb16-104" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│    5000.0│    5000.0│ 0           │   0   │  0  │</span></span>
<span id="cb16-105"><a href="#cb16-105" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │    AFR│   15000.0│   15000.0│ 0           │   0   │  0  │</span></span>
<span id="cb16-106"><a href="#cb16-106" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └─────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-107"><a href="#cb16-107" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Events @ generation 2e+05</span></span>
<span id="cb16-108"><a href="#cb16-108" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌──────────────────────────────────────────────────────────────────────────────┐</span></span>
<span id="cb16-109"><a href="#cb16-109" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │   time│type        │parameters       │effect                                 │</span></span>
<span id="cb16-110"><a href="#cb16-110" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├──────────────────────────────────────────────────────────────────────────────┤</span></span>
<span id="cb16-111"><a href="#cb16-111" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  2e+05│Population  │derived=[AFR],   │Moves all lineages from the 'AFR'      │</span></span>
<span id="cb16-112"><a href="#cb16-112" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │Split       │ancestral=CHIMP  │derived population to the ancestral    │</span></span>
<span id="cb16-113"><a href="#cb16-113" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                 │'CHIMP' population. Also set 'AFR' to  │</span></span>
<span id="cb16-114"><a href="#cb16-114" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                 │inactive, and all migration rates to   │</span></span>
<span id="cb16-115"><a href="#cb16-115" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                 │and from the derived population to     │</span></span>
<span id="cb16-116"><a href="#cb16-116" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │            │                 │zero.                                  │</span></span>
<span id="cb16-117"><a href="#cb16-117" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └──────────────────────────────────────────────────────────────────────────────┘</span></span>
<span id="cb16-118"><a href="#cb16-118" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════╗</span></span>
<span id="cb16-119"><a href="#cb16-119" aria-hidden="true" tabindex="-1"></a><span class="do">## ║ Epoch[5]: [2e+05, inf) generations ║</span></span>
<span id="cb16-120"><a href="#cb16-120" aria-hidden="true" tabindex="-1"></a><span class="do">## ╠════════════════════════════════════╝</span></span>
<span id="cb16-121"><a href="#cb16-121" aria-hidden="true" tabindex="-1"></a><span class="do">## ╟    Populations (total=4 active=1)</span></span>
<span id="cb16-122"><a href="#cb16-122" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ┌─────────────────────────────────────────┐</span></span>
<span id="cb16-123"><a href="#cb16-123" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │       │    start│      end│growth_rate  │</span></span>
<span id="cb16-124"><a href="#cb16-124" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    ├─────────────────────────────────────────┤</span></span>
<span id="cb16-125"><a href="#cb16-125" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    │  CHIMP│   5000.0│   5000.0│ 0           │</span></span>
<span id="cb16-126"><a href="#cb16-126" aria-hidden="true" tabindex="-1"></a><span class="do">## ║    └─────────────────────────────────────────┘</span></span>
<span id="cb16-127"><a href="#cb16-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-128"><a href="#cb16-128" aria-hidden="true" tabindex="-1"></a><span class="co"># For debugging of technical issues (with msprime, with slendr, or both), it is</span></span>
<span id="cb16-129"><a href="#cb16-129" aria-hidden="true" tabindex="-1"></a><span class="co"># very useful to have the `msprime` function dump the "raw" command-line to</span></span>
<span id="cb16-130"><a href="#cb16-130" aria-hidden="true" tabindex="-1"></a><span class="co"># run the simulation on the terminal using plain Python interpreter</span></span>
<span id="cb16-131"><a href="#cb16-131" aria-hidden="true" tabindex="-1"></a><span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">run =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-132"><a href="#cb16-132" aria-hidden="true" tabindex="-1"></a><span class="do">## /Users/mp/Library/r-miniconda-arm64/envs/Python-3.12_msprime-1.3.3_tskit-0.5.8_pyslim-1.0.4_tspop-0.0.2/bin/python /Users/mp/Projects/simgen/data/introgression/script.py --seed 1429643904 --model /Users/mp/Projects/simgen/data/introgression --sequence-length 1000000 --recombination-rate 1e-08    --path &lt;path to a .trees file&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-4-inspecting-the-tree-sequence-object" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-4-inspecting-the-tree-sequence-object">Part 4: Inspecting the tree-sequence object</h2>
<p>As we will see later, <em>slendr</em> provides an R-friendly interface to accessing a <a href="https://tskit.dev/tskit/docs/stable/python-api.html">subset of <em>tskit</em>’s functionality</a> for working with tree sequences and for computing various popgen statistics.</p>
<p>For now, <strong>type out the <code>ts</code> object in the terminal – what do you see?</strong> You should get a summary of a tree-sequence object that you’re familiar with from your <em>msprime</em> and <em>tskit</em> activity earlier in the day.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> This is a very important feature of <em>slendr</em> – when a simulation is concluded (doesn’t matter if it was a <code>slim()</code> or <code>msprime()</code> simulation), you will get a normal <em>tskit</em> object. In fact, the fact that <em>slendr</em> supports (so far, and likely always) only a “subset” of all of <em>tskit</em>’s functionality isn’t stopping you to write custom Python/<em>tskit</em> processing code of a tree sequence generated from a <em>slendr</em> model. Under the hood, a <em>slendr</em> simulation <em>really is</em> just an <em>msprime</em> (or SLiM) simulation! It’s just executed through a simplified interface.</p>
</div></div><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Typing out the object with the result shows that it's a good old tskit</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># tree-sequence object</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │       9909║
╟───────────────┼───────────╢
║Sequence Length│    1000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      48000║
╟───────────────┼───────────╢
║Total Size     │    8.9 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤══════╤═════════╤════════════╗
║Table      │Rows  │Size     │Has Metadata║
╠═══════════╪══════╪═════════╪════════════╣
║Edges      │135698│  4.1 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Individuals│ 24000│656.3 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Migrations │     0│  8 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Mutations  │     0│ 16 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Nodes      │105245│  2.8 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Populations│     4│341 Bytes│         Yes║
╟───────────┼──────┼─────────┼────────────╢
║Provenances│     1│  2.7 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Sites      │     0│ 16 Bytes│          No║
╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The brilliance of the tree-sequence data structure rests on its elegant table-based implementation (much more information on that is <a href="https://tskit.dev/tskit/docs/stable/data-model.html">here</a>). <em>slendr</em> isn’t really designed to run very complex low-level manipulations of tree-sequence data (its strength lies in the convenient interface to popgen statistical functions implemented by <em>tskit</em>), but it does contain a couple of functions which can be useful for inspecting the lower-level nature of a tree sequence. Let’s look at a couple of them now.</p>
<p><strong>Use the <code>ts_table</code> function to inspect the low-level table-based representation of a tree sequence.</strong> For instance, you can get the table of nodes with <code>ts_table(ts, "nodes")</code>, edges with <code>ts_table(ts, "edges")</code>, and do the same thing for “individuals”, “mutations”, and “sites”. <strong>Does your tree sequence contain any mutations? If not, why, and how can we even do any popgen with data without any mutations? As you’re doing this, take a look at at the <a href="https://tskit.dev/tutorials/_images/0327585c23c21d289094cc6394cc71ecc7e43f14197d7961b9d759c5abcc0e29.svg">following figure</a> (this was made from a different tree sequence than you have, but that’s OK) to help you relate the information in the tables to a tree sequence which those tables (particularly tables of nodes and edges) implicitly encode.</strong></p>
<p>This should convince you that the final product of a <em>slendr</em> simulation really is the same kind of tree-sequence object that you learned about in the previous activities today. You don’t have to study these tables in detail!</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># slendr provides a helper function which allows access to all the low-level</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># components of every tree-sequence object</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_table</span>(ts, <span class="st">"nodes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 105,245 × 5
   node_id    ind_id pop_id  time time_tskit
     &lt;int&gt; &lt;int[1d]&gt;  &lt;int&gt; &lt;dbl&gt;      &lt;dbl&gt;
 1       0         0      0     0          0
 2       1         0      0     0          0
 3       2         1      0     0          0
 4       3         1      0     0          0
 5       4         2      0     0          0
 6       5         2      0     0          0
 7       6         3      0     0          0
 8       7         3      0     0          0
 9       8         4      0     0          0
10       9         4      0     0          0
# ℹ 105,235 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_table</span>(ts, <span class="st">"edges"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 135,698 × 5
      id child parent  left   right
   &lt;dbl&gt; &lt;int&gt;  &lt;int&gt; &lt;dbl&gt;   &lt;dbl&gt;
 1     0  6465  48000     0 1000000
 2     1  8962  48000     0 1000000
 3     2 40506  48001     0 1000000
 4     3 41949  48001     0 1000000
 5     4 17805  48002     0 1000000
 6     5 18225  48002     0 1000000
 7     6 10642  48003     0 1000000
 8     7 26605  48003     0 1000000
 9     8 30461  48004     0 1000000
10     9 39735  48004     0 1000000
# ℹ 135,688 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_table</span>(ts, <span class="st">"individuals"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,000 × 5
   ind_id  time    pop_id time_tskit sampled
    &lt;dbl&gt; &lt;dbl&gt; &lt;int[1d]&gt;  &lt;dbl[1d]&gt; &lt;lgl&gt;  
 1      0     0         0          0 TRUE   
 2      1     0         0          0 TRUE   
 3      2     0         0          0 TRUE   
 4      3     0         0          0 TRUE   
 5      4     0         0          0 TRUE   
 6      5     0         0          0 TRUE   
 7      6     0         0          0 TRUE   
 8      7     0         0          0 TRUE   
 9      8     0         0          0 TRUE   
10      9     0         0          0 TRUE   
# ℹ 23,990 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We didn't simulate any mutations, so we only have genealogies for now.</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_table</span>(ts, <span class="st">"mutations"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 × 5
# ℹ 5 variables: id &lt;dbl&gt;, site &lt;int&gt;, node &lt;int&gt;, time &lt;dbl&gt;, time_tskit &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_table</span>(ts, <span class="st">"sites"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 0 × 2
# ℹ 2 variables: id &lt;dbl&gt;, position &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>There are also two <em>slendr</em>-specific functions called <code>ts_samples()</code> (which retrieves the “symbolic names” and dates of all recorded individuals at the end of a simulation) and <code>ts_nodes()</code>. <strong>You can run them simply as <code>ts_samples(ts)</code> and <code>ts_nodes(ts)</code>. How many individuals (samples) are in your tree sequence as you simulated it? How is the result of <code>ts_nodes()</code> different from <code>ts_samples()</code>?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># slendr provides a convenient function `ts_samples()` which allows us to</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect the contents of a simulated tree sequence in a more human-readable,</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co"># simplified way. We can see that our tree sequence contains a massive number</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of individuals. Too many, in fact -- we recorded every single individual alive</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co"># at the end of our simulation, which is something we're unlikely to be ever lucky</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co"># enough to have, regardless of which species we study.</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,000 × 3
   name    time pop  
   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 AFR_1      0 AFR  
 2 AFR_2      0 AFR  
 3 AFR_3      0 AFR  
 4 AFR_4      0 AFR  
 5 AFR_5      0 AFR  
 6 AFR_6      0 AFR  
 7 AFR_7      0 AFR  
 8 AFR_8      0 AFR  
 9 AFR_9      0 AFR  
10 AFR_10     0 AFR  
# ℹ 23,990 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop) <span class="sc">%&gt;%</span> tally</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  pop       n
  &lt;chr&gt; &lt;int&gt;
1 AFR   15000
2 CHIMP  5000
3 EUR    3000
4 NEA    1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This function returns a table similar to the one produced by `ts_table(ts, "nodes")`</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co"># above, except that it contains additional slendr metadata (names of individuals</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># belonging to each node, spatial coordinates of nodes for spatial models, etc.).</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It's a bit more useful for analyzing tree-sequence data than the "low-level" functions.</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_nodes</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 8
  name  pop   ind_id node_id  time time_tskit sampled pop_id
  &lt;chr&gt; &lt;fct&gt;  &lt;dbl&gt;   &lt;int&gt; &lt;dbl&gt;      &lt;dbl&gt; &lt;lgl&gt;    &lt;int&gt;
1 AFR_1 AFR     5000   10000     0          0 TRUE         1
2 AFR_1 AFR     5000   10001     0          0 TRUE         1
3 AFR_2 AFR     5001   10002     0          0 TRUE         1
4 AFR_2 AFR     5001   10003     0          0 TRUE         1
5 AFR_3 AFR     5002   10004     0          0 TRUE         1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_nodes</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">tail</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 8
  name  pop   ind_id node_id     time time_tskit sampled pop_id
  &lt;chr&gt; &lt;fct&gt;  &lt;dbl&gt;   &lt;int&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;lgl&gt;    &lt;int&gt;
1 &lt;NA&gt;  CHIMP     NA  105240 7437613.    247920. FALSE        0
2 &lt;NA&gt;  CHIMP     NA  105241 7503971.    250132. FALSE        0
3 &lt;NA&gt;  CHIMP     NA  105242 7559759.    251992. FALSE        0
4 &lt;NA&gt;  CHIMP     NA  105243 7600706.    253357. FALSE        0
5 &lt;NA&gt;  CHIMP     NA  105244 7863052.    262102. FALSE        0</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-5-scheduling-sampling-events" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-5-scheduling-sampling-events">Part 5: Scheduling sampling events</h2>
<p>In the table produced by the <code>ts_samples()</code> function you saw that the tree sequence we simulated recorded <em>everyone</em>. It’s very unlikely, unless we’re extremely lucky, that we’ll ever have a sequence of every single individual in a population that we study. To get a little closer to the scale of the genomic data that we usually work with on a day-to-day basis, we can restrict our simulation to only record a subset of individuals.</p>
<p>We can precisely define which individuals (from which populations, and at which times) should be recorded in a tree sequence using the <em>slendr</em> function <code>schedule_sampling()</code>. For instance, if we have a <code>model</code> with some <em>slendr</em> populations in variables <code>eur</code> and <code>afr</code>, we can schedule the recording of 5 individuals from each at times 10000 (years ago) and 0 (present-day) (using the “years before present” direction of time in our current model of Neanderthal introgression) with the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>pop_schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">5</span>), <span class="fu">list</span>(afr, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function simply returns a data frame. As such, we can create multiple of such schedules (of arbitrary complexity and granularity), and then bind them together into a single sampling schedule with a single line of code, like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that the `times =` argument of the `schedule_sampling()` function can be</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a vector of times like here...</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>ancient_times <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> ancient_times, <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ... but also just a single number like here</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>nea_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">time =</span> <span class="dv">60000</span>, <span class="fu">list</span>(nea, <span class="dv">1</span>))</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="co"># But whatever the means you create the individual sampling schedules with,</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="co"># you can always bind them all to a single table with the `rbind()` function</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(eur_samples, afr_samples, nea_samples)</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>schedule</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Using the function <code>schedule_sampling</code> (and with the help of <code>rbind</code> as shown in the previous code chunk), program the sampling of the following sample sets at given times, saving it to variable called <code>schedule</code>:</strong></p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>time</th>
<th style="text-align: left;">population</th>
<th># individuals</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>70000</td>
<td style="text-align: left;">nea</td>
<td>1</td>
</tr>
<tr class="even">
<td>40000</td>
<td style="text-align: left;">nea</td>
<td>1</td>
</tr>
<tr class="odd">
<td>0</td>
<td style="text-align: left;">chimp</td>
<td>1</td>
</tr>
<tr class="even">
<td>0</td>
<td style="text-align: left;">afr</td>
<td>5</td>
</tr>
<tr class="odd">
<td>0</td>
<td style="text-align: left;">eur</td>
<td>10</td>
</tr>
</tbody>
</table>
<p><strong>Additionally, schedule the sampling of a single <code>eur</code> individual at the following times:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">40000</span>, <span class="dv">2000</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">2000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> You can provide a vector variable (such as <code>t</code> in this example) as the <code>times =</code> argument of <code>schedule_sampling()</code>.</p>
</div></div><p><strong>In total, you should schedule the recording of 38 individuals.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we scheduled the sampling of two Neanderthals at 70kya and 40kya</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>nea_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">70000</span>, <span class="dv">40000</span>), <span class="fu">list</span>(nea, <span class="dv">1</span>))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>nea_samples <span class="co"># (this function produces a plain old data frame!)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 7
   time pop       n y_orig x_orig y     x    
  &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
1 40000 NEA       1 NA     NA     NA    NA   
2 70000 NEA       1 NA     NA     NA    NA   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we schedule one Chimpanzee sample, 5 African samples, and 10 European samples</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>present_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(chimp, <span class="dv">1</span>), <span class="fu">list</span>(afr, <span class="dv">5</span>), <span class="fu">list</span>(eur, <span class="dv">10</span>))</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We also schedule the recording of one European sample between 50kya and 2kya,</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co"># every 2000 years</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>times <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">40000</span>, <span class="dv">2000</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">2000</span>)</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>emh_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, times, <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Because those functions produce nothing but a data frame, we can bind</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a><span class="co"># individual sampling schedules together</span></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(nea_samples, present_samples, emh_samples)</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>schedule</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 25 × 7
    time pop       n y_orig x_orig y     x    
   &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
 1 40000 NEA       1 NA     NA     NA    NA   
 2 70000 NEA       1 NA     NA     NA    NA   
 3     0 CHIMP     1 NA     NA     NA    NA   
 4     0 AFR       5 NA     NA     NA    NA   
 5     0 EUR      10 NA     NA     NA    NA   
 6  2000 EUR       1 NA     NA     NA    NA   
 7  4000 EUR       1 NA     NA     NA    NA   
 8  6000 EUR       1 NA     NA     NA    NA   
 9  8000 EUR       1 NA     NA     NA    NA   
10 10000 EUR       1 NA     NA     NA    NA   
# ℹ 15 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><strong>Then, verify the correctness of your overall sampling <code>schedule</code> by visualizing it together with your <code>model</code> like this:</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> As you’ve seen above, the visualization is often a bit wonky and convoluted with overlapping elements and it can be even worse with samples added, but try to experiment with arguments to <code>plot_model</code> described above to make the plot a bit more helpful for sanity checking.</p>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">sizes =</span> <span class="cn">FALSE</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">sizes =</span> <span class="cn">FALSE</span>, <span class="at">log =</span> <span class="cn">TRUE</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-6-simulating-a-defined-set-of-individuals" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-6-simulating-a-defined-set-of-individuals">Part 6: Simulating a defined set of individuals</h2>
<p>You have now both a compiled <em>slendr</em> <code>model</code> and a well-defined sampling <code>schedule</code>.</p>
<p><strong>Use your combined sampling schedule stored in the <code>schedule</code> variable to run a new tree-sequence simulation from your model (again using the <code>msprime()</code> function), this time restricted to just those individuals scheduled for recording. You can do this by providing the combined sampling <code>schedule</code> as the <code>samples = schedule</code> argument of the function <code>msprime</code> you used above.</strong> Just replace the line(s) with your first <code>msprime()</code> from the previous part of this exercise with the new one, which uses the <code>schedule</code> for customized sampling.</p>
<p><strong>Also, while you’re doing this, use the <code>ts_mutate()</code> function to overlay neutral mutations on the simulated tree sequence right after the <code>msprime()</code> call.</strong> (Take a look at the handounts for a reminder of the <code>%&gt;%</code> pipe patterns I showed you.)</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The command below will likely take a few minutes to run, so feel free to go</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co"># down from 100 Mb sequence_length to even 10Mb (it doesn't matter much).</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="co"># (The `random_seed =` argument is there for reproducibility purposes.)</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">samples =</span> schedule, <span class="at">random_seed =</span> <span class="dv">1269258439</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>, <span class="at">random_seed =</span> <span class="dv">1269258439</span>)</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Time difference of 2.141642 mins</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If you're bothered by ho long this takes, feel free to call these two lines</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="co"># to 100% reproduce my results without any expensive computation:</span></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">read_model</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/introgression"</span>))</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">ts_read</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="at">file =</span> <span class="st">"data/introgression.trees"</span>), <span class="at">model =</span> model)</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a><span class="co"># We can save a tree sequence object using a slendr function `ts_write` (this</span></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a><span class="co"># can be useful if we want to save the results of a simulation for later use).</span></span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_write</span>(ts, <span class="st">"data/introgression.trees"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p><strong>Inspect the tree-sequence object saved in the <code>ts</code> variable by typing it into the R console again</strong> (this interactivity really helps with catching nasty bugs early during the programming of your script). <strong>You can also do a similar thing via the table produced by the <code>ts_samples()</code> function. You should see a much smaller number of individuals being recorded, indicating that the simulation was much more efficient and produced genomic data for only the individuals of interest.</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> When you think about it, it’s actually quite astonishing how fast <em>msprime</em> and <em>tskit</em> are when dealing with such a huge amount of sequence data from tens of thousands of individuals on a simple laptop!</p>
</div></div><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the (tskit/Python-based) summary of the new tree sequence</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (note the much smaller number of "sample nodes"!)</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     295199║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │         76║
╟───────────────┼───────────╢
║Total Size     │   85.1 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1051064│ 32.1 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│     38│  1.1 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │ 663287│ 23.4 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 218703│  5.8 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      4│341 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      2│  5.7 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │ 660993│ 15.8 MiB│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the table of all recorded samples in the tree sequence</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 38 × 3
   name   time pop  
   &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
 1 NEA_1 70000 NEA  
 2 EUR_1 40000 EUR  
 3 NEA_2 40000 NEA  
 4 EUR_2 38000 EUR  
 5 EUR_3 36000 EUR  
 6 EUR_4 34000 EUR  
 7 EUR_5 32000 EUR  
 8 EUR_6 30000 EUR  
 9 EUR_7 28000 EUR  
10 EUR_8 26000 EUR  
# ℹ 28 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the count of individuals in different time points</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop, <span class="at">present_day =</span> time <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> tally <span class="sc">%&gt;%</span> <span class="fu">select</span>(present_day, pop, n)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
# Groups:   pop [4]
  present_day pop       n
  &lt;lgl&gt;       &lt;chr&gt; &lt;int&gt;
1 TRUE        AFR       5
2 TRUE        CHIMP     1
3 FALSE       EUR      20
4 TRUE        EUR      10
5 FALSE       NEA       2</code></pre>
</div>
</div>
</div>
</div>
</div>
<!-- End of Bonus exercises -->
</section>
</section>
<section id="computing-popgen-statistics-on-tree-sequences-from-slendr" class="level1 page-columns page-full">
<h1>Computing popgen statistics on tree sequences from <em>slendr</em></h1>
<p>In this exercise, you will build on top of the results from Exercise 1. Specifically, we will learn how to compute popgen statistics on <em>slendr</em>-simulated tree sequences using <a href="https://www.slendr.net/reference/index.html#tree-sequence-statistics"><em>slendr</em>’s interface</a> to the <em>tskit</em> Python module.</p>
<p><strong>First, create a new R script <code>exercise2.R</code> and paste in the following code.</strong> This is one of the possible solutions to the Exercise 1, and it’s easier if we all use it to be on the same page from now on, starting from the same model and the same simulated tree sequence:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="do">## The interface to all required Python modules has been activated.</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"CHIMP"</span>, <span class="at">time =</span> <span class="fl">7e6</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>afr <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"AFR"</span>, <span class="at">parent =</span> chimp, <span class="at">time =</span> <span class="fl">6e6</span>, <span class="at">N =</span> <span class="dv">15000</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>eur <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"EUR"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">70e3</span>, <span class="at">N =</span> <span class="dv">3000</span>)</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>nea <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"NEA"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">600e3</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">remove =</span> <span class="fl">40e3</span>)</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> nea, <span class="at">to =</span> eur, <span class="at">rate =</span> <span class="fl">0.03</span>, <span class="at">start =</span> <span class="dv">55000</span>, <span class="at">end =</span> <span class="dv">50000</span>)</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, nea, afr, eur),</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf,</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a><span class="co"># We will read a cached version of a tree sequence I simulated myself</span></span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a><span class="co"># to make sure we're all on the same page. That said, if you managed to</span></span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a><span class="co"># do Exercise 1 on your own, feel free to stick with your own tree sequence!</span></span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">ts_read</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/introgression.trees"</span>), <span class="at">model =</span> model)</span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>),</span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>, <span class="at">log =</span> <span class="cn">TRUE</span>),</span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">1</span></span>
<span id="cb60-27"><a href="#cb60-27" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As a sanity check, let’s use a couple of tidyverse table-munging tricks to make sure the tree sequence does contain a set of sample which matches our intended sampling schedule (particularly the time series of European individuals and the two Neanderthals):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co"># total number of recorded individuals in the tree sequence</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> nrow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 38</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># times of sampling of each recorded European individual</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"EUR"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 40000 38000 36000 34000 32000 30000 28000 26000 24000 22000 20000 18000
[13] 16000 14000 12000 10000  8000  6000  4000  2000     0     0     0     0
[25]     0     0     0     0     0     0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># times of sampling of each recorded Neanderthal</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"NEA"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 70000 40000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># count of individuals in each population</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pop, <span class="at">present_day =</span> time <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  tally <span class="sc">%&gt;%</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(pop, present_day, n) <span class="sc">%&gt;%</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(present_day)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
# Groups:   pop [4]
  pop   present_day     n
  &lt;chr&gt; &lt;lgl&gt;       &lt;int&gt;
1 EUR   FALSE          20
2 NEA   FALSE           2
3 AFR   TRUE            5
4 CHIMP TRUE            1
5 EUR   TRUE           10</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> These bits of tidyverse code are extremely helpful when you’re working with large tree sequences with many individuals as sanity checks that your sampling worked as intended. I’m listing them here in case you’ve never worked with the tidyverse family of R packages before (such as the <em>dplyr</em> package where <code>filter()</code>, <code>group_by()</code>, <code>tally()</code>, and <code>pull()</code> come from).</p>
</div></div><p>Everything looks good! Having made sure that the <code>ts</code> object contains the individuals we want, let’s move to the exercise.</p>
<section id="part-1-computing-nucleotide-diversity" class="level2">
<h2 class="anchored" data-anchor-id="part-1-computing-nucleotide-diversity">Part 1: Computing nucleotide diversity</h2>
<p>The toy model of ancient human history plotted above makes a fairly clear prediction of what would be the nucleotide diversity expected in the simulated populations. <strong>Compute the nucleotide diversity in all populations using the <em>slendr</em> function <a href="https://www.slendr.net/reference/ts_diversity.html#ref-examples"><code>ts_diversity()</code></a> in your tree sequence <code>ts</code>. Do you get numbers that (relatively between all populations) match what would expect from the model given the <span class="math inline">\(N_e\)</span> that you programmed for each?</strong></p>
<p><strong>Hint:</strong> Nearly every <em>slendr</em> statistic function interfacing with <em>tskit</em> accepts a <code>ts</code> tree-sequence object as its first argument, with further arguments being either a vector of individual names representing a group of samples to compute a statistic on, or a (named) list of such vectors (each element of that list for a group of samples) – these lists are intended to be equivalent to the <code>sample_sets =</code> argument of many <em>tskit</em> Python methods (which you’ve learned about in your activity on <em>tskit</em>), except that they allow symbolic names of individuals, rather then integer indices of nodes in a tree sequence.</p>
<p>Although you can get all the above information by processing the table produced by the <code>ts_samples()</code> function, <em>slendr</em> provides a useful helper function <code>ts_names()</code> which only returns the names of individuals as a vector (or a named list of such vectors, one vector per population as shown below).</p>
<p>When you call it directly, you get a plain vector of individual names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NEA_1"   "EUR_1"   "NEA_2"   "EUR_2"   "EUR_3"   "EUR_4"   "EUR_5"  
 [8] "EUR_6"   "EUR_7"   "EUR_8"   "EUR_9"   "EUR_10"  "EUR_11"  "EUR_12" 
[15] "EUR_13"  "EUR_14"  "EUR_15"  "EUR_16"  "EUR_17"  "EUR_18"  "EUR_19" 
[22] "EUR_20"  "AFR_1"   "AFR_2"   "AFR_3"   "AFR_4"   "AFR_5"   "CHIMP_1"
[29] "EUR_21"  "EUR_22"  "EUR_23"  "EUR_24"  "EUR_25"  "EUR_26"  "EUR_27" 
[36] "EUR_28"  "EUR_29"  "EUR_30" </code></pre>
</div>
</div>
<p>This is not super helpful, unless we want to compute some statistic for <em>everyone</em> in the tree sequence, regardless of their population assignment. Perhaps a bit more useful is to call the function like this, because it will produce a result which can be immediately used as the <code>sample_sets =</code> argument mentioned in the <strong>Hint</strong> above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"

$CHIMP
[1] "CHIMP_1"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
</div>
<p>As you can see, this gave us a normal R list, with each element containing a vector of individual names in a population. Note that we can use standard R list indexing to get subsets of individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>names[<span class="st">"NEA"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>names[<span class="fu">c</span>(<span class="st">"EUR"</span>, <span class="st">"NEA"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
</div>
<p>etc.</p>
<p>Many of the following exercises will use these kinds of tricks to instruct various <em>slendr</em> / <em>tskit</em> functions to compute statistics on subsets of all individuals sub-sampled in this way.</p>
<p><strong>After you computed nucleotide diversity per-population, compute it for each individual separately using the same function <code>ts_diversity()</code></strong> (which, in this setting, gives you effectively the heterozygosity for each individual). <strong>If you are familiar with plotting in R, visualize the individual-based heterozygosities across all populations.</strong></p>
<p><strong>Hint:</strong> You can do this by giving a vector of names as <code>sample_sets =</code> (so not an R list of vectors). You could also use the data frame produced by <code>ts_samples(ts)</code> to get the names, just adding the heterozygosities to that data frame as a new column.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Population-based nucleotide diversity:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's first get a named list of individuals in each group we want to be</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># working with (slendr tree-sequence statistic functions generally operate</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with this kind of structure)</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>sample_sets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"

$CHIMP
[1] "CHIMP_1"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use such `sample_sets` object to compute nucleotide diversity (pi)\</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="co"># in each population, in a bit of a similar manner to how we would do it</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with the standard tskit in Python</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>pi_pop <span class="ot">&lt;-</span> <span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> sample_sets)</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(pi_pop, diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 NEA   0.0000483
2 CHIMP 0.000196 
3 EUR   0.000510 
4 AFR   0.000600 </code></pre>
</div>
</div>
<p>You can see that this simple computation fits the extreme differences in long-term <span class="math inline">\(N_e\)</span> encoded by your <em>slendr</em> demografr model.</p>
<p><strong>Per-individual heterozygosity:</strong></p>
<p>We can do this by passing the vector of individual names directory as the <code>sample_sets =</code> argument, rather than in a list of groups as we did above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For convenience, we first get a table of all individuals (which of course</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co"># contains also their names) and in the next step, we'll just add their</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co"># heterozygosities as a new column.</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>pi_df <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts)</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>pi_df<span class="sc">$</span>name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NEA_1"   "EUR_1"   "NEA_2"   "EUR_2"   "EUR_3"   "EUR_4"   "EUR_5"  
 [8] "EUR_6"   "EUR_7"   "EUR_8"   "EUR_9"   "EUR_10"  "EUR_11"  "EUR_12" 
[15] "EUR_13"  "EUR_14"  "EUR_15"  "EUR_16"  "EUR_17"  "EUR_18"  "EUR_19" 
[22] "EUR_20"  "AFR_1"   "AFR_2"   "AFR_3"   "AFR_4"   "AFR_5"   "CHIMP_1"
[29] "EUR_21"  "EUR_22"  "EUR_23"  "EUR_24"  "EUR_25"  "EUR_26"  "EUR_27" 
[36] "EUR_28"  "EUR_29"  "EUR_30" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>pi_df<span class="sc">$</span>diversity <span class="ot">&lt;-</span> <span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> pi_df<span class="sc">$</span>name)<span class="sc">$</span>diversity</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>pi_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 38 × 4
   name   time pop   diversity
   &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 NEA_1 70000 NEA   0.0000431
 2 EUR_1 40000 EUR   0.000516 
 3 NEA_2 40000 NEA   0.0000410
 4 EUR_2 38000 EUR   0.000520 
 5 EUR_3 36000 EUR   0.000539 
 6 EUR_4 34000 EUR   0.000534 
 7 EUR_5 32000 EUR   0.000533 
 8 EUR_6 30000 EUR   0.000553 
 9 EUR_7 28000 EUR   0.000499 
10 EUR_8 26000 EUR   0.000506 
# ℹ 28 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's plot the results using the ggplot2 package</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (because a picture is worth a thousand numbers!)</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pi_df, <span class="fu">aes</span>(pop, diversity, <span class="at">color =</span> pop, <span class="at">group =</span> pop)) <span class="sc">+</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span></span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-2-computing-pairwise-divergence" class="level2">
<h2 class="anchored" data-anchor-id="part-2-computing-pairwise-divergence">Part 2: Computing pairwise divergence</h2>
<p><strong>Use the function <a href="https://www.slendr.net/reference/ts_divergence.html#ref-examples"><code>ts_divergence()</code></a> to compute genetic divergence between all pairs of populations. Again, do you get results compatible with our demographic model in terms of expectation given the split times between populations as you programmed them for your model?</strong></p>
<p><strong>Hint:</strong> Again, you can use the same concept of <code>sample_sets =</code> we discussed in the previous part. In this case, the function computes <em>pairwise</em> divergence between each element of the list given as <code>sample_sets =</code> (i.e., for each vector of individual names).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>div_df <span class="ot">&lt;-</span> <span class="fu">ts_divergence</span>(ts, sample_sets)</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(div_df, divergence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  x     y     divergence
  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
1 AFR   EUR     0.000649
2 EUR   NEA     0.000955
3 AFR   NEA     0.000983
4 CHIMP NEA     0.00416 
5 CHIMP EUR     0.00417 
6 AFR   CHIMP   0.00418 </code></pre>
</div>
</div>
<p>We can see that the pairwise nucleotide divergences between populations recapitulate the known population/species relationships we would expect from our model.</p>
</div>
</div>
</div>
</section>
<section id="part-3-detecting-neanderthal-admixture-in-europeans" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-3-detecting-neanderthal-admixture-in-europeans">Part 3: Detecting Neanderthal admixture in Europeans</h2>
<p>Let’s now pretend its about 2008, we’ve sequenced the first Neanderthal genome, and we are working on a project that will <a href="https://www.science.org/doi/10.1126/science.1188021">change human evolution research forever</a>. We also have the genomes of a couple of people from Africa and Europe, which we want to use to answer the most burning question of all evolutionary anthropology: <em>“Do some people living today carry Neanderthal ancestry?”</em></p>
<p>Earlier you’ve learned about <span class="math inline">\(f\)</span>-statistics of various kinds. You have also heard that an <span class="math inline">\(f_4\)</span> statistic (or its equivalent <span class="math inline">\(D\)</span> statistic) can be used as a test of “treeness”. Simply speaking, for some “quartet” of individuals or population samples, they can be used as a hypothesis test of whether the history of those samples is compatible with there not having been an introgression.</p>
<p><strong>Compute the <span class="math inline">\(f_4\)</span> test of Neanderthal introgression in EUR individuals using the <em>slendr</em> function <code>ts_f4()</code>.</strong> When you’re running it, you will have to provide individuals to compute the statistic using a slightly different format. Take a look at the help page available as <code>?ts_f4</code> for more information. <strong>When you’re computing the <span class="math inline">\(f_4\)</span>, make sure to set <code>mode = "branch"</code> argument of the <code>ts_f4()</code> function (we will get to why a bit later).</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> By default, each <em>slendr</em> / <em>tskit</em> statistic function operates on mutations, and this will switch them to use branch length (as you might know, <span class="math inline">\(f\)</span>-statistics are mathematically defined using branch lengths in trees and <code>mode = "branch"</code> does exactly that).</p>
</div></div><p><strong>Hint:</strong> If you haven’t learned this in your <span class="math inline">\(f\)</span>-statistics lecture, you want to compute (and compare) the values of these two statistics using the <em>slendr</em> function <code>ts_f4()</code>:</p>
<ol type="1">
<li><p><span class="math inline">\(f_4\)</span>(&lt;some African&gt;, &lt;another African&gt;; &lt;Neanderthal&gt;, &lt;Chimp&gt;)</p></li>
<li><p><span class="math inline">\(f_4\)</span>(&lt;some African&gt;, &lt;a test European&gt;; &lt;Neanderthal&gt;, &lt;Chimp&gt;),</p></li>
</ol>
<p>here <code>&lt;individual&gt;</code> can be the name of any individual recorded in your tree sequence, such as names you saw as <code>name</code> column in the table returned by <code>ts_samples(ts)</code> (i.e.&nbsp;<code>"NEA_1"</code> could be used as a “representative” &lt;Neanderthal&gt; in those equations, similarly for <code>"CHIMP_1"</code> as the fourth sample in the <span class="math inline">\(f_4\)</span> quarted representing the outgroup).</p>
<p>To simplify things a lot, we can understand the above equations as comparing the counts of so-called BABA and ABBA allele patterns between the quarted of samples specified in the statistics:</p>
<p><span class="math display">\[
f_4(AFR, X; NEA, CHIMP) = \frac{\#BABA - \#ABBA}{\#SNPs}
\]</span></p>
<p>The first <span class="math inline">\(f_4\)</span> statistic above is not expected to give values “too different” from 0 (even in case of Neanderthal introgression into Europeans) because we don’t expect two African individuals to differ “significantly” in terms of how much alleles they share with a Neanderthal (because their ancestors never met Neanderthals!). The other should – if there was a Neanderthal introgression into Europeans some time in their history – be “significantly negative”.</p>
<p><strong>Is the second of those two statistics “much more negative” than the first, as expected assuming introgression from Neanderthals into Europeans?</strong></p>
<p><strong>Why am I putting “significantly” and “much more negative” in quotes in the previous sentence? What are we missing here for this being a true hypothesis test as you might be accustomed to from computing <span class="math inline">\(f\)</span>-statistics using a tool such as ADMIXTOOLS?</strong> (We will get to this again in the following part of this exercise.)</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the difference in the amount of allele sharing between two African</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co"># individuals and a Neanderthal</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>f4_null <span class="ot">&lt;-</span> <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> <span class="st">"AFR_2"</span>, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>f4_null</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z          f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 AFR_1 AFR_2 NEA_1 CHIMP_1  43.4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the difference in the amount of allele sharing between an African</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co"># individual vs European individual and a Neanderthal</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>f4_alt <span class="ot">&lt;-</span> <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> <span class="st">"EUR_1"</span>, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>f4_alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z          f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 AFR_1 EUR_1 NEA_1 CHIMP_1 -853.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can see that the second test resulted in an f4 value about ~20 times more</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="co"># negative than the first test, indicating that a European in our test carries</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a><span class="co"># "significantly more" Neanderthal alleles compared to the baseline expectation</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of no introgression established by the comparison to an African ...</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abs</span>(f4_alt<span class="sc">$</span>f4 <span class="sc">/</span> f4_null<span class="sc">$</span>f4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19.65719</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ... although this is not a real test of significance (we have no Z-score or</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="co"># standard error which would give us something like a p-value for the hypothesis</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="co"># test, as we get by jackknife procedure in ADMIXTOOLS)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-4-detecting-neanderthal-admixture-in-europeans-v2.0" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-4-detecting-neanderthal-admixture-in-europeans-v2.0">Part 4: Detecting Neanderthal admixture in Europeans v2.0</h2>
<p>The fact that we don’t get something equivalent to a p-value in these kinds of simulations is generally not a problem, because we’re often interested in establishing a trend of a statistic under various conditions, and understanding when and how its <em>expected value</em> behaves in a certain way. If statistical noise is a problem, we work around this by computing a statistic on multiple simulation replicates or even increasing the sample sizes.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> To see this in practice, you can check out a <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig02">paper</a> in which I used this approach quite successfully on a related problem.</p>
</div></div><p>On top of that, p-value of something like an <span class="math inline">\(f\)</span>-statistic (whether it’s significantly different from zero) is also strongly affected by quality of the data, sequencing errors, coverage, etc. (which can certainly be examined using simulations!). However, these are aspects of modeling which are quite orthogonal to the problem of investigating the expectations and trends of statistics given some underlying evolutionary model, which is what we’re after in these exercises.</p>
<p>That said, even in perfect simulated data, what exactly does “significantly different from zero compared to some baseline expectation” mean can be blurred by noise with simple single-individual comparisons that we did above. Let’s increase the sample size a bit to see if a statistical pattern expected in <span class="math inline">\(f_4\)</span> statistic from our Neanderthal introgression model becomes more apparent.</p>
<p><strong>Compute the first <span class="math inline">\(f_4\)</span> statistic (the baseline expectation between a pair of Africans) and the second <span class="math inline">\(f_4\)</span> statistic (comparing an African and a European), but this time on <em>all recorded Africans</em> and <em>all recorded Europeans</em>, respectively. Plot the <em>distributions</em> of those two sets of statistics. This should remove lots of the uncertainty and make a statistical trend stand out much more clearly.</strong></p>
<p><strong>Hint:</strong> Whenever you need to compute something for many things in sequence, looping is very useful. One way to do compute, say, an <span class="math inline">\(f_4\)</span> statistic over many individuals is by using this kind of pattern using R’s looping function <code>lapply()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over vector of individual names (variable x) and apply a given ts_f4()</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="co"># expression on each individual (note the ts_f4(..., X = x, ...) in the code)</span></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>list_f4 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"ind_1"</span>, <span class="st">"ind_2"</span>, ...),</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a><span class="co"># The above gives us a list of data frames, so we need to bind them all into a</span></span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a><span class="co"># single table for easier interpretation and visualization</span></span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a>df_f4 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, list_f4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This gives us list of vectors of the names of all individuals in each</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co"># population...</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ... which we can then access like this</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>sample_sets<span class="sc">$</span>AFR <span class="co"># all Africans</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>sample_sets<span class="sc">$</span>EUR <span class="co"># all Europeans</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's compute the f4 statistic for all Africans... </span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>f4_afr_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  sample_sets<span class="sc">$</span>AFR,</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ... and Europeans</span></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>f4_eur_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>  sample_sets<span class="sc">$</span>EUR,</span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind each list of data frames into a single data frame</span></span>
<span id="cb100-13"><a href="#cb100-13" aria-hidden="true" tabindex="-1"></a>f4_afr <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, f4_afr_list)</span>
<span id="cb100-14"><a href="#cb100-14" aria-hidden="true" tabindex="-1"></a>f4_eur <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, f4_eur_list)</span>
<span id="cb100-15"><a href="#cb100-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-16"><a href="#cb100-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's add population columns to each of the two results for easier plotting</span></span>
<span id="cb100-17"><a href="#cb100-17" aria-hidden="true" tabindex="-1"></a>f4_afr<span class="sc">$</span>pop <span class="ot">&lt;-</span> <span class="st">"AFR"</span></span>
<span id="cb100-18"><a href="#cb100-18" aria-hidden="true" tabindex="-1"></a>f4_eur<span class="sc">$</span>pop <span class="ot">&lt;-</span> <span class="st">"EUR"</span></span>
<span id="cb100-19"><a href="#cb100-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-20"><a href="#cb100-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind both tables together</span></span>
<span id="cb100-21"><a href="#cb100-21" aria-hidden="true" tabindex="-1"></a>f4_results <span class="ot">&lt;-</span> <span class="fu">rbind</span>(f4_afr, f4_eur)</span>
<span id="cb100-22"><a href="#cb100-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-23"><a href="#cb100-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the results</span></span>
<span id="cb100-24"><a href="#cb100-24" aria-hidden="true" tabindex="-1"></a>f4_results <span class="sc">%&gt;%</span></span>
<span id="cb100-25"><a href="#cb100-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(pop, f4, <span class="at">color =</span> pop)) <span class="sc">+</span></span>
<span id="cb100-26"><a href="#cb100-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb100-27"><a href="#cb100-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span></span>
<span id="cb100-28"><a href="#cb100-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb100-29"><a href="#cb100-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"f4(AFR, EUR; NEA, CHIMP)"</span>) <span class="sc">+</span></span>
<span id="cb100-30"><a href="#cb100-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can see that the <span class="math inline">\(f_4\)</span> statistic test of Neanderthal introgression in Europeans indeed does give a much more negative distribution of values compared to the baseline expectation which compares two Africans to each other.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus exercises
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="bonus-1-mode-branch-vs-mode-site" class="level3">
<h3 class="anchored" data-anchor-id="bonus-1-mode-branch-vs-mode-site">Bonus 1: <code>mode = "branch"</code> vs <code>mode = "site"</code></h3>
<p><strong>Repeat the previous part of the exercise by setting <code>mode = "site"</code> in the <code>ts_f4()</code> function calls</strong> (this is actually the default behavior of all <em>slendr</em> tree-sequence based <em>tskit</em> functions). This will switch the <em>tskit</em> computation to using mutation counts along each branch of the tree sequence, rather than using branch length themselves. <strong>Why might the branch-based computation be a bit better if what we’re after is investigating the expected values of statistics under some model?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See <a href="https://tskit.dev/tutorials/no_mutations.html#genealogy-based-measures-are-less-noisy">this tutorial</a> (and particularly the directly linked section) for explanation.</p>
</div>
</div>
</div>
</section>
<section id="bonus-2-outgroup-f_3-statistic" class="level3">
<h3 class="anchored" data-anchor-id="bonus-2-outgroup-f_3-statistic">Bonus 2: Outgroup <span class="math inline">\(f_3\)</span> statistic</h3>
<p><strong>Use the function <code>ts_f3()</code> to compute the outgroup</strong> <span class="math inline">\(f_3\)</span> statistic between pairs of African-European, African-Neanderthal, and European-Neanderthal and a Chimpanzee outgroup.</p>
<p><strong>Hint:</strong> The <span class="math inline">\(f_3\)</span> statistic is traditionally expressed as <span class="math inline">\(f_3(A, B; C)\)</span>, where C represents the outgroup. Unfortunately, in <em>tskit</em> the outgroup is named A, with B and C being the pair of samples from which we trace the length of branches towards the outgroup, so the statistic is interpreted as <span class="math inline">\(f_3(B, C; A)\)</span>.</p>
<p><strong>How do the outgroup f3 results compare to your expectation based on simple population relationships (and to the divergence computation above)?</strong></p>
<p><strong>Do you see any impact of introgression on the <span class="math inline">\(f_3\)</span> value when a Neanderthal is included in the computation?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># f3(A, B; C) = E[ (A - C) * (B - C) ]</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This means that in tskit, C is the outgroup (different from ADMIXTOOLS!)</span></span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We can compute f3 for individuals...</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> <span class="st">"AFR_1"</span>, <span class="at">C =</span> <span class="st">"EUR_1"</span>, <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR_1 EUR_1 0.00375</code></pre>
</div>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ... but also whole populations (or population samples)</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"AFR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"EUR"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR   EUR   0.00375</code></pre>
</div>
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"AFR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"NEA"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR   NEA   0.00358</code></pre>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"EUR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"NEA"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 EUR   NEA   0.00359</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-3-outgroup-f_3-statistic-as-a-linear-combination-of-f_2-statistics" class="level3">
<h3 class="anchored" data-anchor-id="bonus-3-outgroup-f_3-statistic-as-a-linear-combination-of-f_2-statistics">Bonus 3: Outgroup <span class="math inline">\(f_3\)</span> statistic as a linear combination of <span class="math inline">\(f_2\)</span> statistics</h3>
<p>You might have learned that any complex <span class="math inline">\(f\)</span>-statistic can be expressed as a linear combination of multiple <span class="math inline">\(f_2\)</span> statistics (which represent simple branch length separating two lineages). <strong>Verify that this is the case by looking up equation <em>(20b)</em> in <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">this amazing paper</a> and compute an</strong> <span class="math inline">\(f_3\)</span> statistic for any arbitrary trio of individuals of your choosing using this linear combination of <span class="math inline">\(f_2\)</span> statistics.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># standard f3</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> <span class="st">"AFR_1"</span>, <span class="at">C =</span> <span class="st">"AFR_2"</span>, <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR_1 AFR_2 0.00378</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a "homemade" f3 statistic as a linear combination of f2 statistics</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="co"># f3(A, B; C) = f2(A, C) + f2(B, C) - f2(A, B) / 2</span></span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>homemade_f3 <span class="ot">&lt;-</span> (</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_1"</span>, <span class="at">B =</span> <span class="st">"CHIMP_1"</span>)<span class="sc">$</span>f2 <span class="sc">+</span></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_2"</span>, <span class="at">B =</span> <span class="st">"CHIMP_1"</span>)<span class="sc">$</span>f2 <span class="sc">-</span></span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_1"</span>, <span class="at">B =</span> <span class="st">"AFR_2"</span>)<span class="sc">$</span>f2</span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>homemade_f3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.003778796</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-4-trajectory-of-neanderthal-ancestry-in-europe-over-time" class="level3">
<h3 class="anchored" data-anchor-id="bonus-4-trajectory-of-neanderthal-ancestry-in-europe-over-time">Bonus 4: Trajectory of Neanderthal ancestry in Europe over time</h3>
<p>There used to be a lot of controversy about the question of whether or not did Neanderthal ancestry proportion in Europeans decline or not over the past 40 thousand years (see figure 1 in <a href="https://www.pnas.org/doi/full/10.1073/pnas.1814338116">this paper</a> figure 2 in <a href="https://www.nature.com/articles/nature17993">this paper</a>).</p>
<p>Your simulated tree sequence contains a time-series of European individuals over time. Use the <em>slendr</em> function <code>ts_f4ratio()</code> to compute (and then plot) the proportion (commonly designated as <code>alpha</code>) of Neanderthal ancestry in Europe over time. Use <span class="math inline">\(f_4\)</span>-ratio statistic of the following form:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> <span class="sc">&lt;</span>vector of ind. names<span class="sc">&gt;</span>, <span class="at">A =</span> <span class="st">"NEA_1"</span>, <span class="at">B =</span> <span class="st">"NEA_2"</span>, <span class="at">C =</span> <span class="st">"AFR_1"</span>, <span class="at">O =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract table with names and times of sampled Europeans (ancient and present day)</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>eur_inds <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"EUR"</span>)</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>eur_inds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 3
   name    time pop  
   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 EUR_1  40000 EUR  
 2 EUR_2  38000 EUR  
 3 EUR_3  36000 EUR  
 4 EUR_4  34000 EUR  
 5 EUR_5  32000 EUR  
 6 EUR_6  30000 EUR  
 7 EUR_7  28000 EUR  
 8 EUR_8  26000 EUR  
 9 EUR_9  24000 EUR  
10 EUR_10 22000 EUR  
# ℹ 20 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute f4-ration statistic (this will take ~30s) -- note that we can provide</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a vector of names for the X sample set to the `ts_f4ratio()` function</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>nea_ancestry <span class="ot">&lt;-</span> <span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> eur_inds<span class="sc">$</span>name, <span class="at">A =</span> <span class="st">"NEA_1"</span>, <span class="at">B =</span> <span class="st">"NEA_2"</span>, <span class="at">C =</span> <span class="st">"AFR_1"</span>, <span class="at">O =</span> <span class="st">"CHIMP_1"</span>)</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the age of each sample to the table of proportions</span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>nea_ancestry<span class="sc">$</span>time <span class="ot">&lt;-</span> eur_inds<span class="sc">$</span>time</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>nea_ancestry</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 7
   X      A     B     C     O        alpha  time
   &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;
 1 EUR_1  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0153 40000
 2 EUR_2  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0170 38000
 3 EUR_3  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0117 36000
 4 EUR_4  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0201 34000
 5 EUR_5  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0261 32000
 6 EUR_6  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0132 30000
 7 EUR_7  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0151 28000
 8 EUR_8  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0243 26000
 9 EUR_9  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0195 24000
10 EUR_10 NEA_1 NEA_2 AFR_1 CHIMP_1 0.0243 22000
# ℹ 20 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>nea_ancestry <span class="sc">%&gt;%</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(time, alpha)) <span class="sc">+</span></span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">40000</span>, <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>)) <span class="sc">+</span></span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"time [years ago]"</span>, <span class="at">y =</span> <span class="st">"Neanderthal ancestry proportion"</span>) <span class="sc">+</span></span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Neanderthal ancestry proportion in Europeans over time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For good measure, let's test the significance of the decline using a linear model</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(alpha <span class="sc">~</span> time, <span class="at">data =</span> nea_ancestry))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = alpha ~ time, data = nea_ancestry)

Residuals:
       Min         1Q     Median         3Q        Max 
-0.0102725 -0.0029417  0.0001505  0.0040630  0.0073792 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  2.205e-02  1.214e-03  18.173   &lt;2e-16 ***
time        -1.031e-07  6.204e-08  -1.662    0.108    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.004642 on 28 degrees of freedom
Multiple R-squared:  0.08979,   Adjusted R-squared:  0.05728 
F-statistic: 2.762 on 1 and 28 DF,  p-value: 0.1077</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-5-how-many-unique-f4-quartets-are-there" class="level3">
<h3 class="anchored" data-anchor-id="bonus-5-how-many-unique-f4-quartets-are-there">Bonus 5: How many unique f4 quartets are there?</h3>
<p>In your lecture about <span class="math inline">\(f\)</span>-statistics, you’ve probably learned about various symmetries in <span class="math inline">\(f_4\)</span> (but also other <span class="math inline">\(f\)</span>-statistics) depending on the arrangement of the “quartet”. As a trivial example, an <span class="math inline">\(f_3(A; B, C)\)</span> and <span class="math inline">\(f_3(A; C, B)\)</span> will give you exactly the same value, and the same thing applies even to more complex <span class="math inline">\(f\)</span>-statistics like <span class="math inline">\(f_4\)</span>.</p>
<p><strong>Use simulations to compute how manu unique</strong> <span class="math inline">\(f_4\)</span> values involving a single quartet are there.</p>
<p><strong>Hint:</strong> Draw some trees to figure out why could that be true. Also, when computing <code>ts_f4()</code>, set <code>mode = "branch"</code> to avoid the effect of statistical noise due to mutations.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # install a combinatorics R package</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("combinat")</span></span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(combinat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'combinat'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    combn</code></pre>
</div>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># These are the four samples we can create quartet combinations from</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>quartet <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"AFR_1"</span>, <span class="st">"EUR_1"</span>, <span class="st">"NEA_1"</span>, <span class="st">"CHIMP_1"</span>)</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>quartets <span class="ot">&lt;-</span> <span class="fu">permn</span>(quartet)</span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>quartets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "AFR_1"   "EUR_1"   "NEA_1"   "CHIMP_1"

[[2]]
[1] "AFR_1"   "EUR_1"   "CHIMP_1" "NEA_1"  

[[3]]
[1] "AFR_1"   "CHIMP_1" "EUR_1"   "NEA_1"  

[[4]]
[1] "CHIMP_1" "AFR_1"   "EUR_1"   "NEA_1"  

[[5]]
[1] "CHIMP_1" "AFR_1"   "NEA_1"   "EUR_1"  

[[6]]
[1] "AFR_1"   "CHIMP_1" "NEA_1"   "EUR_1"  

[[7]]
[1] "AFR_1"   "NEA_1"   "CHIMP_1" "EUR_1"  

[[8]]
[1] "AFR_1"   "NEA_1"   "EUR_1"   "CHIMP_1"

[[9]]
[1] "NEA_1"   "AFR_1"   "EUR_1"   "CHIMP_1"

[[10]]
[1] "NEA_1"   "AFR_1"   "CHIMP_1" "EUR_1"  

[[11]]
[1] "NEA_1"   "CHIMP_1" "AFR_1"   "EUR_1"  

[[12]]
[1] "CHIMP_1" "NEA_1"   "AFR_1"   "EUR_1"  

[[13]]
[1] "CHIMP_1" "NEA_1"   "EUR_1"   "AFR_1"  

[[14]]
[1] "NEA_1"   "CHIMP_1" "EUR_1"   "AFR_1"  

[[15]]
[1] "NEA_1"   "EUR_1"   "CHIMP_1" "AFR_1"  

[[16]]
[1] "NEA_1"   "EUR_1"   "AFR_1"   "CHIMP_1"

[[17]]
[1] "EUR_1"   "NEA_1"   "AFR_1"   "CHIMP_1"

[[18]]
[1] "EUR_1"   "NEA_1"   "CHIMP_1" "AFR_1"  

[[19]]
[1] "EUR_1"   "CHIMP_1" "NEA_1"   "AFR_1"  

[[20]]
[1] "CHIMP_1" "EUR_1"   "NEA_1"   "AFR_1"  

[[21]]
[1] "CHIMP_1" "EUR_1"   "AFR_1"   "NEA_1"  

[[22]]
[1] "EUR_1"   "CHIMP_1" "AFR_1"   "NEA_1"  

[[23]]
[1] "EUR_1"   "AFR_1"   "CHIMP_1" "NEA_1"  

[[24]]
[1] "EUR_1"   "AFR_1"   "NEA_1"   "CHIMP_1"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many permutations there are in total?</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   4! = 4 * 3 * 2 * 1 = 24</span></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a><span class="fu">factorial</span>(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24</code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We should therefore have 24 different quartet combinations of samples</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(quartets)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24</code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop across all quartets, computing the corresponding f4 statistic (we want</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to do this using branch lengths, not mutations, as the mutation-based computation</span></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="co"># would involve statistical noise)</span></span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>all_f4s <span class="ot">&lt;-</span> <span class="fu">lapply</span>(quartets, <span class="cf">function</span>(q) <span class="fu">ts_f4</span>(ts, q[<span class="dv">1</span>], q[<span class="dv">2</span>], q[<span class="dv">3</span>], q[<span class="dv">4</span>], <span class="at">mode =</span> <span class="st">"branch"</span>))</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind the list of f4 results into a single data frame and inspect the results</span></span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a>all_f4s <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_f4s) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">abs</span>(f4))</span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(all_f4s, <span class="at">n =</span> <span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24 × 5
   W       X       Y       Z            f4
   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
 1 AFR_1   EUR_1   NEA_1   CHIMP_1   -853.
 2 AFR_1   EUR_1   CHIMP_1 NEA_1      853.
 3 NEA_1   CHIMP_1 AFR_1   EUR_1     -853.
 4 CHIMP_1 NEA_1   AFR_1   EUR_1      853.
 5 CHIMP_1 NEA_1   EUR_1   AFR_1     -853.
 6 NEA_1   CHIMP_1 EUR_1   AFR_1      853.
 7 EUR_1   AFR_1   CHIMP_1 NEA_1     -853.
 8 EUR_1   AFR_1   NEA_1   CHIMP_1    853.
 9 AFR_1   NEA_1   CHIMP_1 EUR_1   -16428.
10 AFR_1   NEA_1   EUR_1   CHIMP_1  16428.
11 NEA_1   AFR_1   EUR_1   CHIMP_1 -16428.
12 NEA_1   AFR_1   CHIMP_1 EUR_1    16428.
13 EUR_1   CHIMP_1 NEA_1   AFR_1   -16428.
14 CHIMP_1 EUR_1   NEA_1   AFR_1    16428.
15 CHIMP_1 EUR_1   AFR_1   NEA_1   -16428.
16 EUR_1   CHIMP_1 AFR_1   NEA_1    16428.
17 AFR_1   CHIMP_1 EUR_1   NEA_1    17281.
18 CHIMP_1 AFR_1   EUR_1   NEA_1   -17281.
19 CHIMP_1 AFR_1   NEA_1   EUR_1    17281.
20 AFR_1   CHIMP_1 NEA_1   EUR_1   -17281.
21 NEA_1   EUR_1   CHIMP_1 AFR_1    17281.
22 NEA_1   EUR_1   AFR_1   CHIMP_1 -17281.
23 EUR_1   NEA_1   AFR_1   CHIMP_1  17281.
24 EUR_1   NEA_1   CHIMP_1 AFR_1   -17281.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Narrow down the results to only unique f4 values</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="fu">distinct</span>(all_f4s, f4, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  W       X       Y       Z            f4
  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
1 AFR_1   EUR_1   NEA_1   CHIMP_1   -853.
2 AFR_1   EUR_1   CHIMP_1 NEA_1      853.
3 AFR_1   NEA_1   CHIMP_1 EUR_1   -16428.
4 AFR_1   NEA_1   EUR_1   CHIMP_1  16428.
5 AFR_1   CHIMP_1 EUR_1   NEA_1    17281.
6 CHIMP_1 AFR_1   EUR_1   NEA_1   -17281.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">distinct</span>(all_f4s, <span class="fu">abs</span>(f4), <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  W     X       Y       Z            f4 `abs(f4)`
  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 AFR_1 EUR_1   NEA_1   CHIMP_1   -853.      853.
2 AFR_1 NEA_1   CHIMP_1 EUR_1   -16428.    16428.
3 AFR_1 CHIMP_1 EUR_1   NEA_1    17281.    17281.</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</div>
</div>
</div>
<!-- End of Bonus exercises -->
</section>
</section>
<section id="simulation-based-inference-of-n_e" class="level1 page-columns page-full">
<h1>Simulation-based inference of <span class="math inline">\(N_e\)</span></h1>
<p>So far we’ve learned how <em>slendr</em> provides an easy way to define demographic models in R and simulate (even very large!) tree sequences from them. This allows us to quickly verify our intuition about some popgen problem (things like <em>“Hmmm, I wonder what would an <span class="math inline">\(f_4\)</span> statistic look like if my model includes this particular gene-flow event?</em>), in just a few lines of R. There have been instances in which we’ve been able to even answer questions like this directly in a meeting, pretty much on the spot! This makes <em>slendr</em> a very powerful “popgen calculator”.</p>
<p>Now let’s take things one step further. Imagine you gathered some empirical data, like an allele frequency spectrum (AFS) from a population that you study. That data was, in the real world, produced by some (hidden) biological process (demographic history) that we want to learn about. For instance, the population we study had some <span class="math inline">\(N_e\)</span>, which we don’t know the value of (the only thing we have is the observed AFS) but we want to infer that value.</p>
<p>Simulations can be a great tool to estimate the most likely value of such an unknown parameter. Briefly speaking, in this particular toy example, we can simulate a large number of AFS vectors (each resulting from a different assumed <span class="math inline">\(N_e\)</span> value) and then pick just those <span class="math inline">\(N_e\)</span> values (or just one <span class="math inline">\(N_e\)</span> value) which produced a simulated AFS closest to the observed AFS.</p>
<p>This is exactly what you’ll be doing just now in Exercise 3.</p>
<section id="part-1-a-self-contained-slendr-function-of-n_e-rightarrow-textrmafs" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-1-a-self-contained-slendr-function-of-n_e-rightarrow-textrmafs">Part 1: A self-contained <em>slendr</em> function of <span class="math inline">\(N_e \rightarrow \textrm{AFS}\)</span></h2>
<p><strong>In a new script <code>exercise3.R</code> write a custom R function called <code>simulate_afs()</code>, which will take <code>Ne</code> as its only parameter. Use this function to compute (and return) <a href="https://en.wikipedia.org/wiki/Allele_frequency_spectrum">AFS vectors</a> for a couple of <code>Ne</code> values of your choosing, but staying between <code>Ne = 1000</code> and <code>Ne = 30000</code> Plot those AFS vectors and observe how (and why?) do they differ based on <code>Ne</code> parameter you used in each respective simulation.</strong></p>
<p><strong>Hint:</strong> The function should create a one-population <em>forward-time</em> model (our population starting at <code>time = 1</code>, with the model <code>simulation_length = 100000</code> and <code>generation_time = 1</code> in <code>compile_model()</code>), simulate 10Mb tree sequence using <code>msprime()</code> (recombination rate 1e-8) and then overlay neutral mutations on it at <code>mutation_rate = 1e-8</code>), compute AFS for 10 samples and return the AFS vector as result of this custom function.</p>
<p><strong>Hint:</strong> If you’ve never programmed before, the concept of a “custom function” might be very alien to you. Again, if you need help, feel free to start building your <code>exercise3.R</code> solution based on this “template” (just fill in missing relevant bits of <em>slendr</em> code that you should be already familiar with):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>simulate_afs <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne) {</span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># In here you should write code which will:</span></span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   1. create one population with a given Ne (provided as a function argument)</span></span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   2. compile a model using `simulation_length =` and `generation_time =`</span></span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   3. simulate a tree sequence</span></span>
<span id="cb137-9"><a href="#cb137-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   4. select names of 10 samples (doesn't matter which, "pop_1", "po2_", ...)</span></span>
<span id="cb137-10"><a href="#cb137-10" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   5. compute AFS vector from those 10 individuals using `ts_afs()`</span></span>
<span id="cb137-11"><a href="#cb137-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb137-12"><a href="#cb137-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># `result` is a variable with your 10-sample AFS vector (we remove the</span></span>
<span id="cb137-13"><a href="#cb137-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first element because it's not meaningful for our example)</span></span>
<span id="cb137-14"><a href="#cb137-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result[<span class="sc">-</span><span class="dv">1</span>]) </span>
<span id="cb137-15"><a href="#cb137-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb137-16"><a href="#cb137-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-17"><a href="#cb137-17" aria-hidden="true" tabindex="-1"></a>afs_1 <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>) <span class="co"># simulate AFS from a Ne = 1000 model...</span></span>
<span id="cb137-18"><a href="#cb137-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_1, <span class="at">type =</span><span class="st">"o"</span>)           <span class="co"># ... and plot it</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Remember that you should drop the first element of the AFS vector produced by <code>ts_afs()</code> (for instance with something like <code>result[-1]</code> if <code>result</code> contains the output of <code>ts_afs()</code>) technical reasons related to <em>tskit</em>. You don’t have to worry about that here, but you can read <a href="https://tskit.dev/tutorials/analysing_tree_sequences.html#sec-tutorial-afs-zeroth-entry">this</a> for more detail.</p>
</div></div><p><strong>Hint:</strong> <strong>If the above still doesn’t make any sense to you, feel free to copy-paste the function from the solution below into your script and work with that function instead!</strong></p>
<p>When used in R, your custom function should work like this (the simulation is stochastic, so your numbers will be different, of course):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This gives us a vector of singletons, doubletons, etc., etc., all the way</span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to the number of fixed mutations in our sample of 10 individuals</span></span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a><span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 299 191  99 104  79  50  77  49  35  48  29  30  40  23  28  33  15  24  15
[20]  14</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># An R function can be understood as a block of a computer program which executes</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a block of code inside the {...} brackets given a certain value of a parameter</span></span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="co"># (here 'Ne' just after the word 'function')</span></span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a>simulate_afs <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne) {</span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create a slendr model with a single population of size Ne = N</span></span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>  pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">N =</span> Ne, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a>  model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">100000</span>)</span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-9"><a href="#cb140-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># simulate a tree sequence</span></span>
<span id="cb140-10"><a href="#cb140-10" aria-hidden="true" tabindex="-1"></a>  ts <span class="ot">&lt;-</span></span>
<span id="cb140-11"><a href="#cb140-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb140-12"><a href="#cb140-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb140-13"><a href="#cb140-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-14"><a href="#cb140-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get a random sample of names of 10 individuals</span></span>
<span id="cb140-15"><a href="#cb140-15" aria-hidden="true" tabindex="-1"></a>  samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">sample</span>(<span class="dv">10</span>)</span>
<span id="cb140-16"><a href="#cb140-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-17"><a href="#cb140-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># compute the AFS vector (dropping the 0-th element added by tskit)</span></span>
<span id="cb140-18"><a href="#cb140-18" aria-hidden="true" tabindex="-1"></a>  afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(ts, <span class="at">sample_sets =</span> <span class="fu">list</span>(samples))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb140-19"><a href="#cb140-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-20"><a href="#cb140-20" aria-hidden="true" tabindex="-1"></a>  afs</span>
<span id="cb140-21"><a href="#cb140-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb140-22"><a href="#cb140-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-23"><a href="#cb140-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's use our custom function to simulate AFS vector for Ne = 1k, 10k, and 30k</span></span>
<span id="cb140-24"><a href="#cb140-24" aria-hidden="true" tabindex="-1"></a>afs_1k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="dv">1000</span>)</span>
<span id="cb140-25"><a href="#cb140-25" aria-hidden="true" tabindex="-1"></a>afs_10k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="dv">10000</span>)</span>
<span id="cb140-26"><a href="#cb140-26" aria-hidden="true" tabindex="-1"></a>afs_30k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="dv">30000</span>)</span>
<span id="cb140-27"><a href="#cb140-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-28"><a href="#cb140-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the three simulated AFS using base R plotting functionality</span></span>
<span id="cb140-29"><a href="#cb140-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_30k, <span class="at">type =</span> <span class="st">"o"</span>, <span class="at">main =</span> <span class="st">"AFS, Ne = 30000"</span>, <span class="at">col =</span> <span class="st">"cyan"</span>,)</span>
<span id="cb140-30"><a href="#cb140-30" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_10k, <span class="at">type =</span> <span class="st">"o"</span>, <span class="at">main =</span> <span class="st">"AFS, Ne = 10000"</span>, <span class="at">col =</span> <span class="st">"purple"</span>)</span>
<span id="cb140-31"><a href="#cb140-31" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_1k, <span class="at">type =</span> <span class="st">"o"</span>, <span class="at">main =</span> <span class="st">"AFS, Ne = 1000"</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span>
<span id="cb140-32"><a href="#cb140-32" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"Ne = 1k"</span>, <span class="st">"Ne = 10k"</span>, <span class="st">"Ne = 30k"</span>),</span>
<span id="cb140-33"><a href="#cb140-33" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"purple"</span>, <span class="st">"cyan"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-2-estimating-unknown-n_e-from-empirical-afs" class="level2">
<h2 class="anchored" data-anchor-id="part-2-estimating-unknown-n_e-from-empirical-afs">Part 2: Estimating unknown <span class="math inline">\(N_e\)</span> from empirical AFS</h2>
<p>Imagine you sequenced 10 samples from a population and computed the following AFS vector (which contains, sequentially, the number of singletons, doubletons, etc., in your sample from a population):</p>
<!-- dput(as.vector(observed_afs)) -->
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You know (maybe from some fossil evidence) that the population probably had a constant <span class="math inline">\(N_e\)</span> somewhere between 1000 and 30000 for the past 100,000 generations, and had mutation and recombination rates of 1e-8 (i.e., parameters already implemented by your <code>simulate_afs()</code> function – how convenient!).</p>
<p><strong>Use <em>slendr</em> simulations to guess the true (and hidden!) <span class="math inline">\(N_e\)</span> given the observed AFS by running simulations for a range of <span class="math inline">\(N_e\)</span> values and finding out which <span class="math inline">\(N_e\)</span> produces the closest AFS vector to the <code>afs_observed</code> vector above using one of the following two approaches.</strong></p>
<ul>
<li><p><strong>Option 1</strong> [easy]: Plot AFS vectors for various <span class="math inline">\(N_e\)</span> values (i.e.&nbsp;simulate several of them using your function <code>simulate_afs()</code>), then eyeball which looks closest to the observed AFS based on the figures alone. (This is, of course, not how proper statistical inference is done, but it will be good enough for this exercie!)</p></li>
<li><p><strong>Option 2</strong> [hard]: Simulate AFS vectors in steps of possible <code>Ne</code> (maybe <code>lapply()</code>?), and find the <span class="math inline">\(N_e\)</span> which gives the closest AFS to the observed AFS based on <a href="https://en.wikipedia.org/wiki/Mean_squared_error">Mean squared error</a>.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution to “Option 1”
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is our starting observed AFS which we want to compare simulated AFS vectors to</span></span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We know that the Ne is between 1000 and 30000, so let's simulate</span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a><span class="co"># a bunch of AFS vectors for different Ne values using our custom</span></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a><span class="co"># AFS simulation function</span></span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a>afs_Ne1k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">1000</span>)</span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a>afs_Ne5k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">5000</span>)</span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>afs_Ne6k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">6000</span>)</span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>afs_Ne10k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">10000</span>)</span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>afs_Ne20k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">20000</span>)</span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a>afs_Ne30k <span class="ot">&lt;-</span> <span class="fu">simulate_afs</span>(<span class="at">Ne =</span> <span class="dv">30000</span>)</span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot all simulated AFS vectors, highlighting the observed AFS in black</span></span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_observed, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">lwd =</span> <span class="dv">3</span>,</span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>, <span class="at">ylab =</span> <span class="st">"count"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">13000</span>))</span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne1k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne5k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"green"</span>)</span>
<span id="cb142-20"><a href="#cb142-20" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne6k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"pink"</span>)</span>
<span id="cb142-21"><a href="#cb142-21" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne10k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"purple"</span>)</span>
<span id="cb142-22"><a href="#cb142-22" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne20k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"orange"</span>)</span>
<span id="cb142-23"><a href="#cb142-23" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs_Ne30k, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"cyan"</span>)</span>
<span id="cb142-24"><a href="#cb142-24" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>,</span>
<span id="cb142-25"><a href="#cb142-25" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"observed AFS"</span>, <span class="st">"Ne = 1000"</span>, <span class="st">"Ne = 5000"</span>,</span>
<span id="cb142-26"><a href="#cb142-26" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"Ne = 6000"</span>, <span class="st">"Ne = 10000"</span>, <span class="st">"Ne = 20000"</span>, <span class="st">"Ne = 30000"</span>),</span>
<span id="cb142-27"><a href="#cb142-27" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"pink"</span>, <span class="st">"purple"</span>, <span class="st">"orange"</span>, <span class="st">"cyan"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!!! SPOILER </span><span class="al">ALERT</span><span class="co"> BEFORE REVEALING THE CORRECT ANSWER !!!!!</span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-6"><a href="#cb143-6" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-7"><a href="#cb143-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-8"><a href="#cb143-8" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-9"><a href="#cb143-9" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-10"><a href="#cb143-10" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-11"><a href="#cb143-11" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-12"><a href="#cb143-12" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-13"><a href="#cb143-13" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-14"><a href="#cb143-14" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-15"><a href="#cb143-15" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-16"><a href="#cb143-16" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-17"><a href="#cb143-17" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-18"><a href="#cb143-18" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-19"><a href="#cb143-19" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-20"><a href="#cb143-20" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-21"><a href="#cb143-21" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-22"><a href="#cb143-22" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-23"><a href="#cb143-23" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-24"><a href="#cb143-24" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-25"><a href="#cb143-25" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-26"><a href="#cb143-26" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-27"><a href="#cb143-27" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-28"><a href="#cb143-28" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-29"><a href="#cb143-29" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-30"><a href="#cb143-30" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-31"><a href="#cb143-31" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-32"><a href="#cb143-32" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-33"><a href="#cb143-33" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-34"><a href="#cb143-34" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-35"><a href="#cb143-35" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-36"><a href="#cb143-36" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-37"><a href="#cb143-37" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-38"><a href="#cb143-38" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-39"><a href="#cb143-39" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-40"><a href="#cb143-40" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-41"><a href="#cb143-41" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-42"><a href="#cb143-42" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-43"><a href="#cb143-43" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-44"><a href="#cb143-44" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-45"><a href="#cb143-45" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-46"><a href="#cb143-46" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-47"><a href="#cb143-47" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-48"><a href="#cb143-48" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb143-49"><a href="#cb143-49" aria-hidden="true" tabindex="-1"></a><span class="co"># true Ne was 6543!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution to “Option 2”
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is our starting observed AFS which we want to compare simulated AFS vectors to</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>,</span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>                  <span class="dv">264</span>, <span class="dv">218</span>,  <span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate regularly spaced values of potential Ne values (our parameter grid)</span></span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a>Ne_grid <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">1000</span>, <span class="at">to =</span> <span class="dv">30000</span>, <span class="at">by =</span> <span class="dv">500</span>)</span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a>Ne_grid</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1000  1500  2000  2500  3000  3500  4000  4500  5000  5500  6000  6500
[13]  7000  7500  8000  8500  9000  9500 10000 10500 11000 11500 12000 12500
[25] 13000 13500 14000 14500 15000 15500 16000 16500 17000 17500 18000 18500
[37] 19000 19500 20000 20500 21000 21500 22000 22500 23000 23500 24000 24500
[49] 25000 25500 26000 26500 27000 27500 28000 28500 29000 29500 30000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># I'm not entirely sure if your workshop cloud instances support big</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="co"># parallelization runs -- if not, you can modify the `mc.cores =` argument</span></span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="co"># a couple of lines below to a smaller number (`mc.cores = 1` would make</span></span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a><span class="co"># the simulation run on a single processor, i.e. no parallelization).</span></span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute AFS (in parallel, to make things faster) across the entire grid of possible Ne values</span></span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a>afs_grid <span class="ot">&lt;-</span> <span class="fu">mclapply</span>(Ne_grid, simulate_afs, <span class="at">mc.cores =</span> <span class="fu">detectCores</span>())</span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(afs_grid) <span class="ot">&lt;-</span> Ne_grid</span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the first five simulated AFS vectors, for brevity</span></span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>afs_grid[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`1000`
 [1] 398 216 147  68  72  72  78  43  30  58  35  26  30  23  27  18  21  28  13
[20]   3

$`1500`
 [1] 590 326 219 127 124  80  90  82  76  51  50  59  48  35  43  38  46  29  15
[20]  22

$`2000`
 [1] 726 426 272 185 147 100 102  74  88  87  64  37  34  47  53  46  37  49  42
[20]  33

$`2500`
 [1] 940 561 341 260 212 159 145 131 117  85  78  85  78  66  72  67  49  63  52
[20]  26

$`3000`
 [1] 1208  528  430  279  194  207  186  154  113  129  113   87   94  103   77
[16]   77   83   55   82   88</code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the observed AFS...</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_observed, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">xlab =</span> <span class="st">"allele count bin"</span>, <span class="at">ylab =</span> <span class="st">"count"</span>)</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ... and overlay the simulated AFS vectors on top of it</span></span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(Ne_grid)) {</span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(afs_grid[[i]], <span class="at">lwd =</span> <span class="fl">0.5</span>)</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"observed AFS"</span>, <span class="st">"simulated AFS"</span>), <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"gray"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute mean-squared error of the AFS produced by each Ne value across the grid</span></span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>errors <span class="ot">&lt;-</span> <span class="fu">sapply</span>(afs_grid, <span class="cf">function</span>(sim_afs) {</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sum</span>((sim_afs <span class="sc">-</span> afs_observed)<span class="sc">^</span><span class="dv">2</span>) <span class="sc">/</span> <span class="fu">length</span>(sim_afs)</span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Ne_grid, errors, <span class="at">ylab =</span> <span class="st">"error"</span>)</span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> Ne_grid[<span class="fu">which.min</span>(errors)], <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">paste</span>(<span class="st">"minimum error Ne ="</span>, Ne_grid[<span class="fu">which.min</span>(errors)]), <span class="at">fill =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-43-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the AFS again, but this time highlight the most likely spectrum</span></span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (i.e. the one which gave the lowest RMSE value)</span></span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs_observed, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">xlab =</span> <span class="st">"allele count bin"</span>, <span class="at">ylab =</span> <span class="st">"count"</span>)</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(Ne_grid)) {</span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>  color <span class="ot">&lt;-</span> <span class="cf">if</span> (i <span class="sc">==</span> <span class="fu">which.min</span>(errors)) <span class="st">"red"</span> <span class="cf">else</span> <span class="st">"gray"</span></span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>  width <span class="ot">&lt;-</span> <span class="cf">if</span> (i <span class="sc">==</span> <span class="fu">which.min</span>(errors)) <span class="dv">2</span> <span class="cf">else</span> <span class="fl">0.75</span></span>
<span id="cb150-7"><a href="#cb150-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(afs_grid[[i]], <span class="at">lwd =</span> width, <span class="at">col =</span> color)</span>
<span id="cb150-8"><a href="#cb150-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb150-9"><a href="#cb150-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"observed AFS"</span>, <span class="fu">paste</span>(<span class="st">"best fitting Ne ="</span>, Ne_grid[<span class="fu">which.min</span>(errors)])),</span>
<span id="cb150-10"><a href="#cb150-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-43-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!!! SPOILER </span><span class="al">ALERT</span><span class="co"> BEFORE REVEALING THE CORRECT ANSWER !!!!!</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-18"><a href="#cb151-18" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-19"><a href="#cb151-19" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-20"><a href="#cb151-20" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-21"><a href="#cb151-21" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-22"><a href="#cb151-22" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-23"><a href="#cb151-23" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-24"><a href="#cb151-24" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-25"><a href="#cb151-25" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-26"><a href="#cb151-26" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-27"><a href="#cb151-27" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-28"><a href="#cb151-28" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-29"><a href="#cb151-29" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-30"><a href="#cb151-30" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-31"><a href="#cb151-31" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-32"><a href="#cb151-32" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-33"><a href="#cb151-33" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-34"><a href="#cb151-34" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-35"><a href="#cb151-35" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-36"><a href="#cb151-36" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-37"><a href="#cb151-37" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-38"><a href="#cb151-38" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-39"><a href="#cb151-39" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-40"><a href="#cb151-40" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-41"><a href="#cb151-41" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-42"><a href="#cb151-42" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-43"><a href="#cb151-43" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-44"><a href="#cb151-44" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-45"><a href="#cb151-45" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-46"><a href="#cb151-46" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-47"><a href="#cb151-47" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-48"><a href="#cb151-48" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb151-49"><a href="#cb151-49" aria-hidden="true" tabindex="-1"></a><span class="co"># true Ne was 6543!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>Congratulations, you now know how to infer parameters of evolutionary models using simulations! What you just did is really very similar to how simulation-based inference is done in practice (even with methods such as ABC). Hopefully you can also see how easy does <em>slendr</em> make it.</p>
<p>This kind of approach can be used to infer all sorts of demographic parameters, even using other summary statistics that you’ve also learned to compute… including selection parameters, which we delve into in the next exercise.</p>
</section>
</section>
<section id="simulating-and-detecting-natural-selection" class="level1 page-columns page-full">
<h1>Simulating (and detecting) natural selection</h1>
<p>The primary motivation for designing <em>slendr</em> was to make demographic modelling in R as trivially easy and fast as possible, focusing exclusively on neutral models. However, as <em>slendr</em> became popular, people have been asking for the possibility of simulating natural selection. After all, a large part of <em>slendr</em>’s functionality deals with population genetic <a href="https://www.slendr.net/articles/vignette-06-locations.html">models across geographical landscapes</a>, which requires SLiM. So why not support selection simulations using <em>slendr</em> as well?</p>
<p>In December 2024 I caved in and added support for modifying <em>slendr</em> demographic models with bits of SLiM code, which allows simulating pretty much any arbitrary selection scenario you might be interested in.</p>
<p>This exercise is a quick demonstration of how this works and how you might simulate selection using <em>slendr</em>. We will do this using another toy model of ancient human history, which we will first use as a basis for simulating the frequency trajectory of an allele under positive selection, and then implementing a toy selection scan using Tajima’s D.</p>
<p>To speed things up, <strong>create a new <code>exercise4.R</code> script and copy the following code as a starting point for this exercise</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>(<span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This line sources a script in which I provide a few useful helper functions</span></span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a><span class="co"># which you can use in this exercise</span></span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"utils.R"</span>))</span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a><span class="co"># African ancestral population</span></span>
<span id="cb152-9"><a href="#cb152-9" aria-hidden="true" tabindex="-1"></a>afr <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"AFR"</span>, <span class="at">time =</span> <span class="dv">65000</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb152-10"><a href="#cb152-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-11"><a href="#cb152-11" aria-hidden="true" tabindex="-1"></a><span class="co"># First migrants out of Africa</span></span>
<span id="cb152-12"><a href="#cb152-12" aria-hidden="true" tabindex="-1"></a>ooa <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"OOA"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="dv">60000</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">remove =</span> <span class="dv">27000</span>)</span>
<span id="cb152-13"><a href="#cb152-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-14"><a href="#cb152-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Eastern hunter-gatherers</span></span>
<span id="cb152-15"><a href="#cb152-15" aria-hidden="true" tabindex="-1"></a>ehg <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"EHG"</span>, <span class="at">parent =</span> ooa, <span class="at">time =</span> <span class="dv">28000</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">remove =</span> <span class="dv">6000</span>)</span>
<span id="cb152-16"><a href="#cb152-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-17"><a href="#cb152-17" aria-hidden="true" tabindex="-1"></a><span class="co"># European population</span></span>
<span id="cb152-18"><a href="#cb152-18" aria-hidden="true" tabindex="-1"></a>eur <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"EUR"</span>, <span class="at">parent =</span> ehg, <span class="at">time =</span> <span class="dv">25000</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb152-19"><a href="#cb152-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-20"><a href="#cb152-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Anatolian farmers</span></span>
<span id="cb152-21"><a href="#cb152-21" aria-hidden="true" tabindex="-1"></a>ana <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"ANA"</span>, <span class="at">time =</span> <span class="dv">28000</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> ooa, <span class="at">remove =</span> <span class="dv">4000</span>)</span>
<span id="cb152-22"><a href="#cb152-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-23"><a href="#cb152-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Yamnaya steppe population</span></span>
<span id="cb152-24"><a href="#cb152-24" aria-hidden="true" tabindex="-1"></a>yam <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"YAM"</span>, <span class="at">time =</span> <span class="dv">8000</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> ehg, <span class="at">remove =</span> <span class="dv">2500</span>)</span>
<span id="cb152-25"><a href="#cb152-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-26"><a href="#cb152-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Define gene-flow events</span></span>
<span id="cb152-27"><a href="#cb152-27" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb152-28"><a href="#cb152-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ana, <span class="at">to =</span> yam, <span class="at">rate =</span> <span class="fl">0.75</span>, <span class="at">start =</span> <span class="dv">7500</span>, <span class="at">end =</span> <span class="dv">6000</span>),</span>
<span id="cb152-29"><a href="#cb152-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> ana, <span class="at">to =</span> eur, <span class="at">rate =</span> <span class="fl">0.5</span>, <span class="at">start =</span> <span class="dv">6000</span>, <span class="at">end =</span> <span class="dv">5000</span>),</span>
<span id="cb152-30"><a href="#cb152-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> yam, <span class="at">to =</span> eur, <span class="at">rate =</span> <span class="fl">0.6</span>, <span class="at">start =</span> <span class="dv">4000</span>, <span class="at">end =</span> <span class="dv">3500</span>)</span>
<span id="cb152-31"><a href="#cb152-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb152-32"><a href="#cb152-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-33"><a href="#cb152-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Compile all populations into a single slendr model object</span></span>
<span id="cb152-34"><a href="#cb152-34" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb152-35"><a href="#cb152-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(afr, ooa, ehg, eur, ana, yam),</span>
<span id="cb152-36"><a href="#cb152-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf, <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb152-37"><a href="#cb152-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb152-38"><a href="#cb152-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-39"><a href="#cb152-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Schedule the sampling from four European populations roughly before their</span></span>
<span id="cb152-40"><a href="#cb152-40" aria-hidden="true" tabindex="-1"></a><span class="co"># disappearance (or before the end of the simulation)</span></span>
<span id="cb152-41"><a href="#cb152-41" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb152-42"><a href="#cb152-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(eur, <span class="dv">50</span>)),</span>
<span id="cb152-43"><a href="#cb152-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">6000</span>, <span class="fu">list</span>(ehg, <span class="dv">50</span>)),</span>
<span id="cb152-44"><a href="#cb152-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">4000</span>, <span class="fu">list</span>(ana, <span class="dv">50</span>)),</span>
<span id="cb152-45"><a href="#cb152-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">2500</span>, <span class="fu">list</span>(yam, <span class="dv">50</span>))</span>
<span id="cb152-46"><a href="#cb152-46" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Next, visualize the demographic model.</strong> If you did a bit of work in human population genetics, you might recognize it as a very simplified model of demographic history of Europe over the past 50 thousand years or so. As you can see, we are recording 50 individuals from four populations – for Europeans we sample 50 individuals at “present-day”, for the remaining populations we’re recording 50 individuals just before their disappearance. Also note that there’s quite a bit of gene-flow! This was an important thing we’ve learned about human history in the past 10 years or so – everyone is mixed with pretty much everyone, there isn’t (and never was) anything as a “pure population”.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> We didn’t discuss it earlier, but <em>slendr</em> also provides the option to specify a <code>remove =</code> argument in a <code>population()</code> call which instructs the simulation engine to delete a population from a simulation at a given point. For our <code>msprime()</code> simulations in earlier examples it wasn’t really important, but for the <code>slim()</code> simulation we will be running below, we want to make a population extinct at a certain timepoint. Which is why our ancient populations in the starting script model have the <code>remove =</code> parameter specified.</p>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<section id="part-1-simulating-a-tree-sequence-and-computing-tajimas-d" class="level3">
<h3 class="anchored" data-anchor-id="part-1-simulating-a-tree-sequence-and-computing-tajimas-d">Part 1: Simulating a tree sequence and computing Tajima’s D</h3>
<p>Although the point of this exercise is to simulate selection, let’s first simulate a normal neutral model using slendr’s <code>msprime()</code> engine as a sanity check. <strong>Simulate 10 Mb of sequence with a recombination rate <code>1e-8</code> and a sampling <code>schedule</code> defined above.</strong> Let’s not worry about adding any mutations, just to change things up a little bit. We’ll be working with branch-based statistics here (which means adding <code>mode = "branch"</code> whenever we will be computing a statistic, such as Tajima’s D).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-24-contents" aria-controls="callout-24" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-24" class="callout-24-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">samples =</span> schedule)</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a>ts <span class="co"># no mutations!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │      19063║
╟───────────────┼───────────╢
║Sequence Length│   10000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │        400║
╟───────────────┼───────────╢
║Total Size     │    3.8 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═════╤═════════╤════════════╗
║Table      │Rows │Size     │Has Metadata║
╠═══════════╪═════╪═════════╪════════════╣
║Edges      │82833│  2.5 MiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Individuals│  200│  5.5 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Migrations │    0│  8 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Mutations  │    0│ 16 Bytes│          No║
╟───────────┼─────┼─────────┼────────────╢
║Nodes      │22626│618.7 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Populations│    6│417 Bytes│         Yes║
╟───────────┼─────┼─────────┼────────────╢
║Provenances│    1│  3.6 KiB│          No║
╟───────────┼─────┼─────────┼────────────╢
║Sites      │    0│ 16 Bytes│          No║
╚═══════════╧═════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><strong>Inspect the table of all individuals recorded in our tree sequence using the function <code>ts_samples()</code>, making sure we have all the individuals scheduled for tree-sequence recording.</strong> (Again, there’s no such a thing as too many sanity checks when doing research!)</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 × 3
   name    time pop  
   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 EHG_1   6000 EHG  
 2 EHG_2   6000 EHG  
 3 EHG_3   6000 EHG  
 4 EHG_4   6000 EHG  
 5 EHG_5   6000 EHG  
 6 EHG_6   6000 EHG  
 7 EHG_7   6000 EHG  
 8 EHG_8   6000 EHG  
 9 EHG_9   6000 EHG  
10 EHG_10  6000 EHG  
# ℹ 190 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop, time) <span class="sc">%&gt;%</span> tally</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
# Groups:   pop [4]
  pop    time     n
  &lt;chr&gt; &lt;dbl&gt; &lt;int&gt;
1 ANA    4000    50
2 EHG    6000    50
3 EUR       0    50
4 YAM    2500    50</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>As you’ve already learned in an earlier exercise, <em>tskit</em> functions in <em>slendr</em> generally operate on vectors (or lists) of individual names, like those produced by <code>ts_names()</code> above. <strong>Get a vector of names of individuals in every population recorded in the tree sequence, then use this to compute Tajima’s D using the <em>slendr</em> function <code>ts_tajima()</code>.</strong> (Use the same approach as you have with <code>ts_diversity()</code> or <code>ts_divergence()</code> above, using the list of names of individuals as the <code>sample_sets =</code> argument for <code>ts_tajima()</code>). <strong>Do you see any striking differences in the Tajima’s D values across populations? Check <a href="https://en.wikipedia.org/wiki/Tajima%27s_D#Interpreting_Tajima's_D">this</a> for some general guidance.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ANA
 [1] "ANA_1"  "ANA_2"  "ANA_3"  "ANA_4"  "ANA_5"  "ANA_6"  "ANA_7"  "ANA_8" 
 [9] "ANA_9"  "ANA_10" "ANA_11" "ANA_12" "ANA_13" "ANA_14" "ANA_15" "ANA_16"
[17] "ANA_17" "ANA_18" "ANA_19" "ANA_20" "ANA_21" "ANA_22" "ANA_23" "ANA_24"
[25] "ANA_25" "ANA_26" "ANA_27" "ANA_28" "ANA_29" "ANA_30" "ANA_31" "ANA_32"
[33] "ANA_33" "ANA_34" "ANA_35" "ANA_36" "ANA_37" "ANA_38" "ANA_39" "ANA_40"
[41] "ANA_41" "ANA_42" "ANA_43" "ANA_44" "ANA_45" "ANA_46" "ANA_47" "ANA_48"
[49] "ANA_49" "ANA_50"

$EHG
 [1] "EHG_1"  "EHG_2"  "EHG_3"  "EHG_4"  "EHG_5"  "EHG_6"  "EHG_7"  "EHG_8" 
 [9] "EHG_9"  "EHG_10" "EHG_11" "EHG_12" "EHG_13" "EHG_14" "EHG_15" "EHG_16"
[17] "EHG_17" "EHG_18" "EHG_19" "EHG_20" "EHG_21" "EHG_22" "EHG_23" "EHG_24"
[25] "EHG_25" "EHG_26" "EHG_27" "EHG_28" "EHG_29" "EHG_30" "EHG_31" "EHG_32"
[33] "EHG_33" "EHG_34" "EHG_35" "EHG_36" "EHG_37" "EHG_38" "EHG_39" "EHG_40"
[41] "EHG_41" "EHG_42" "EHG_43" "EHG_44" "EHG_45" "EHG_46" "EHG_47" "EHG_48"
[49] "EHG_49" "EHG_50"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30" "EUR_31" "EUR_32"
[33] "EUR_33" "EUR_34" "EUR_35" "EUR_36" "EUR_37" "EUR_38" "EUR_39" "EUR_40"
[41] "EUR_41" "EUR_42" "EUR_43" "EUR_44" "EUR_45" "EUR_46" "EUR_47" "EUR_48"
[49] "EUR_49" "EUR_50"

$YAM
 [1] "YAM_1"  "YAM_2"  "YAM_3"  "YAM_4"  "YAM_5"  "YAM_6"  "YAM_7"  "YAM_8" 
 [9] "YAM_9"  "YAM_10" "YAM_11" "YAM_12" "YAM_13" "YAM_14" "YAM_15" "YAM_16"
[17] "YAM_17" "YAM_18" "YAM_19" "YAM_20" "YAM_21" "YAM_22" "YAM_23" "YAM_24"
[25] "YAM_25" "YAM_26" "YAM_27" "YAM_28" "YAM_29" "YAM_30" "YAM_31" "YAM_32"
[33] "YAM_33" "YAM_34" "YAM_35" "YAM_36" "YAM_37" "YAM_38" "YAM_39" "YAM_40"
[41] "YAM_41" "YAM_42" "YAM_43" "YAM_44" "YAM_45" "YAM_46" "YAM_47" "YAM_48"
[49] "YAM_49" "YAM_50"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute genome-wide Tajima's D for each population -- note that we don't</span></span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="co"># expect to see any significant differences because no population experienced</span></span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a><span class="co"># natural selection (yet)</span></span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_tajima</span>(ts, <span class="at">sample_sets =</span> samples, <span class="at">mode =</span> <span class="st">"branch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  set        D
  &lt;chr&gt;  &lt;dbl&gt;
1 ANA   -0.110
2 EHG   -0.113
3 EUR   -0.521
4 YAM   -0.365</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-2-computing-tajimas-d-in-windows" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-2-computing-tajimas-d-in-windows">Part 2: Computing Tajima’s D in windows</h2>
<p>Let’s take this one step forward. Even if there is a locus under positive selection somewhere along our chromosome, it might be quite unlikely that we would find a Tajima’s D value significant enough for the entire chromosome (which is basically what we did in Part 1 now). Fortunately, thanks to the flexibility of the <em>tskit</em> module, the <em>slendr</em> function <code>ts_tajima()</code> has an argument <code>windows =</code>, which allows us to specify the coordinates of windows into which a sequence should be broken into, with Tajima’s D computed separately for each window. Perhaps this will allow us to see the impact of positive selection after we get to adding selection to our model. So let’s first built some code towards that.</p>
<p><strong>Define a variable <code>windows</code> which will contain a vector of coordinates of 100 windows, starting at position <code>0</code>, and ending at position <code>10e6</code> (i.e., the end of our chromosome). Then provide this variable as the <code>windows =</code> argument of <code>ts_tajima()</code> on a new, separate line of your script. Save the result of <code>ts_tajima()</code> into the variable <code>tajima_wins</code>, and inspect its contents in the R console.</strong></p>
<p><strong>Hint:</strong> You can use the R function <code>seq()</code> and its argument <code>length.out = 100</code>, to create the coordinates of window boundaries very easily.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pre-compute genomic windows for window-based computation of Tajima's D</span></span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>windows <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">seq</span>(<span class="dv">0</span>, ts<span class="sc">$</span>sequence_length, <span class="at">length.out =</span> <span class="dv">100</span>))</span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>windows</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1]        0   101010   202020   303030   404040   505051   606061   707071
  [9]   808081   909091  1010101  1111111  1212121  1313131  1414141  1515152
 [17]  1616162  1717172  1818182  1919192  2020202  2121212  2222222  2323232
 [25]  2424242  2525253  2626263  2727273  2828283  2929293  3030303  3131313
 [33]  3232323  3333333  3434343  3535354  3636364  3737374  3838384  3939394
 [41]  4040404  4141414  4242424  4343434  4444444  4545455  4646465  4747475
 [49]  4848485  4949495  5050505  5151515  5252525  5353535  5454545  5555556
 [57]  5656566  5757576  5858586  5959596  6060606  6161616  6262626  6363636
 [65]  6464646  6565657  6666667  6767677  6868687  6969697  7070707  7171717
 [73]  7272727  7373737  7474747  7575758  7676768  7777778  7878788  7979798
 [81]  8080808  8181818  8282828  8383838  8484848  8585859  8686869  8787879
 [89]  8888889  8989899  9090909  9191919  9292929  9393939  9494949  9595960
 [97]  9696970  9797980  9898990 10000000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute genome-wide Tajima's D for each population in individual windows</span></span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>tajima_wins <span class="ot">&lt;-</span> <span class="fu">ts_tajima</span>(ts, <span class="at">sample_sets =</span> samples, <span class="at">windows =</span> windows, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>tajima_wins</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  set   D           
  &lt;chr&gt; &lt;named list&gt;
1 ANA   &lt;dbl [99]&gt;  
2 EHG   &lt;dbl [99]&gt;  
3 EUR   &lt;dbl [99]&gt;  
4 YAM   &lt;dbl [99]&gt;  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You can see that the format of the result is slightly strange, with the</span></span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="co"># `D` column containing a vector of numbers (this is done for conciseness)</span></span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>tajima_wins[<span class="dv">1</span>, ]<span class="sc">$</span>D</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`1`
 [1]  0.330262761  0.384711676 -0.940977464  0.255359573  0.217527555
 [6] -0.624906565  0.154466323 -0.214262438  0.685501564 -0.094378783
[11]  0.508321766 -0.193201779 -0.317806367  0.031241946 -0.063548911
[16] -0.459291796  0.567431551  0.577834535 -0.510844513 -0.393633394
[21]  0.214767386  0.007402898 -0.492649357 -0.252155503 -0.853599267
[26] -0.523573570  0.118517676 -0.407553808 -0.706102858 -0.344431711
[31]  0.564918806 -0.217403008  0.890654646 -0.063069872  0.342618880
[36]  0.854535779  0.319980920 -0.175834010 -1.065479958 -0.591120480
[41] -0.216640481 -0.303778399  0.044259456  0.519252576 -0.529938471
[46]  0.617801836  0.256848792  0.133338111 -0.179171504 -0.711084369
[51]  0.462546091 -0.401406261 -0.722834537  0.113856215  0.147721980
[56]  0.268035749 -0.477384717 -0.293618126 -0.363408198 -0.267762455
[61]  0.666017494  0.172528058 -0.368281829 -1.132534150 -0.429854111
[66] -0.766990892 -0.625317319 -1.184192484 -0.864078268 -1.081634911
[71] -0.580380016 -0.478994250 -0.114837544  0.144771015  0.041724919
[76] -0.155742550 -0.597064881 -0.566913636  0.036274766 -0.327430851
[81] -0.500871820 -0.006603553 -0.102350497 -0.080518768  0.414648380
[86] -0.681210957  0.117383785  0.346241061  0.648263886  0.552595076
[91]  0.218529845 -0.263889481 -1.297795275 -0.378681896  0.084100860
[96] -0.161245492  0.115172661 -0.149005677 -0.321882137</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The default output format of <code>ts_tajima()</code> is not super user-friendly. <strong>Process the result using a helper function <code>process_tajima(tajima_wins)</code> that I provided for you (perhaps save it as <code>tajima_df</code>), and visualize it using another of my helper functions <code>plot_tajima(tajima_df)</code>.</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> Making the <code>process_tajima()</code> and <code>plot_tajima()</code> function available in your R code is the purpose of the <code>source(here::here("utils.R"))</code> command at the beginning of your script for this exercise.</p>
</div></div><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The helper function `process_tajima()` reformats the results into a normal</span></span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="co"># data frame, this time with a new column `window` which indicates the index</span></span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="co"># of the window that each `D` value was computed in</span></span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a>tajima_df <span class="ot">&lt;-</span> <span class="fu">process_tajima</span>(tajima_wins)</span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a>tajima_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 396 × 3
   set         D window
   &lt;chr&gt;   &lt;dbl&gt;  &lt;int&gt;
 1 ANA    0.330       1
 2 ANA    0.385       2
 3 ANA   -0.941       3
 4 ANA    0.255       4
 5 ANA    0.218       5
 6 ANA   -0.625       6
 7 ANA    0.154       7
 8 ANA   -0.214       8
 9 ANA    0.686       9
10 ANA   -0.0944     10
# ℹ 386 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now let's visualize the window-based Tajima's D along the simulated genome</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="co"># using another helper function `plot_tajima()`</span></span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_tajima</span>(tajima_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It’s no surprise that we don’t see any Tajima’s D outliers in any of our windows, because we’re still working with a tree sequence produced by our a purely neutral simulation. But we have everything set up for the next part, in which we will add selection acting on a beneficial allele.</p>
</div>
</div>
</div>
</section>
<section id="part-3-adding-positive-selection-to-the-base-demographic-model" class="level2">
<h2 class="anchored" data-anchor-id="part-3-adding-positive-selection-to-the-base-demographic-model">Part 3: Adding positive selection to the base demographic model</h2>
<p>Although primarily designed for neutral demographic models, <em>slendr</em> allows optional simulation of natural selection by providing a “SLiM extension code snippet” with customization SLiM code as an optional argument <code>extension =</code> of <code>compile_model()</code> (a function you’re closely familiar with at this point).</p>
<p>Unfortunately we don’t have any space to explain SLiM here (and I have no idea, at the time of writing, whether or not you will have worked with SLiM earlier in this workshop). Suffice to say that SLiM is another very popular population genetic simulator software which allows simulation of selection, and which requires you to write custom code in a different programming language called Eidos.</p>
<p><strong>Take a look at the file <code>slim_extension.txt</code> provided in your working directory (it’s also part of the GitHub repository <a href="https://github.com/bodkan/simgen/blob/main/slim_extension.txt">here</a>). If you worked with SLiM before, glance through the script casually and see if it makes any sense to you. If you have not worked with SLiM before, look for the strange <code>{elements}</code> in curly brackets in the first ten lines of the script.</strong> Those are the parameters of the selection model we will be customizing the standard neutral demographic model we started with in the next step.</p>
<p>Specifically, when you inspect the <code>slim_extension.txt</code> file, you can see that this “SLiM extension script” I provided for you has three parameters:</p>
<ul>
<li><code>origin_pop</code> – in which population should a beneficial allele appear,</li>
<li><code>s</code> – what should be the selection coefficient of the beneficial allele, and</li>
<li><code>onset_time</code> – at which time should the allele appear in the <code>origin_pop</code>.</li>
</ul>
<p>However, at the start, the SLiM extension snippet doesn’t contain any concrete values of those parameters, but only their <code>{origin_pop}</code>, <code>{s}</code>, and <code>{onset_time}</code> placeholders.</p>
<p><strong>Use the <em>slendr</em> function <code>substitute_values()</code> to substitute concrete values for those parameters like this:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>extension <span class="ot">&lt;-</span> <span class="fu">substitute_values</span>(</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">template =</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"slim_extension.txt"</span>),</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">origin_pop =</span> <span class="st">"EUR"</span>,</span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">s =</span> <span class="fl">0.15</span>,</span>
<span id="cb173-5"><a href="#cb173-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">onset_time =</span> <span class="dv">12000</span></span>
<span id="cb173-6"><a href="#cb173-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb173-7"><a href="#cb173-7" aria-hidden="true" tabindex="-1"></a>extension</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T//RtmplT1OZW/filee0822753264e"</code></pre>
</div>
</div>
<p>You can see that <code>substitute_values()</code> returned a path to a file. <strong>Take a look at that file in your terminal – you should see each of the three <code>{placeholder}</code> parameters replaced with a concrete given value.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-29-contents" aria-controls="callout-29" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-29" class="callout-29-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Let’s take a look at the first 15 lines of the extension file before and after calling <code>substitute_values()</code>. We’ll do this in R for simplicity, but you can use <code>less</code> in plain unix terminal.</p>
<p><strong>Before – see the {{placeholder}} parameters in their original form:</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>// Define model constants (to be substituted) all in one place
// (each {{placeholder}} will be replaced by a value passed from R).
// Note that string constant template patterns are surrounded by "quotes"!
initialize() {
    defineConstant("s", {{s}});
    defineConstant("onset_time", {{onset_time}});
    defineConstant("origin_pop", "{{origin_pop}}");

    // compose a trajectory file based on given parameters
    defineConstant("traj_file", PATH + "/" + "trajectory.tsv");
}</code></pre>
</div>
</div>
<p><strong>After – see the {{placeholder}} parameters with concrete values!</strong></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>// Define model constants (to be substituted) all in one place
// (each {{placeholder}} will be replaced by a value passed from R).
// Note that string constant template patterns are surrounded by "quotes"!
initialize() {
    defineConstant("s", 0.15);
    defineConstant("onset_time", 12000);
    defineConstant("origin_pop", "EUR");

    // compose a trajectory file based on given parameters
    defineConstant("traj_file", PATH + "/" + "trajectory.tsv");
}</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>And that’s all the extra work we need to turn our purely neutral demographic <em>slendr</em> model into a model which includes natural selection! (In this case, only a simple selection acting on a single locus, as you’ll see later, but this can be generalized to any imaginable selection scenario.)</p>
<p>How do we use the SLiM extension for our simulation? It’s very simple – we just have to provide the <code>extension</code> variable as an additional argument of good old <code>compile_model()</code>. This will compile a new <em>slendr</em> model which will now include the new functionality for simulating natural selection:</p>
<p><strong>Compile a new <code>model</code> of the history of populations <code>afr</code>, <code>ooa</code>, <code>ehg</code>, etc., by following the instructions above, providing a new <code>extension =</code> argument to the <code>compile_model()</code> function.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-30-contents" aria-controls="callout-30" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-30" class="callout-30-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(afr, ooa, ehg, eur, ana, yam),</span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf, <span class="at">generation_time =</span> <span class="dv">30</span>,</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">extension =</span> extension   <span class="co"># &lt;======== this is missing in the neutral example!</span></span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-4-running-a-selection-simulation-using-slim" class="level2">
<h2 class="anchored" data-anchor-id="part-4-running-a-selection-simulation-using-slim">Part 4: Running a selection simulation using <code>slim()</code></h2>
<p>Now we can finally run our selection simulation!</p>
<p>There are two modifications to our previous simulation workflows:</p>
<ol type="1">
<li>Because we need to run a non-neutral simulation, we have to switch from using the <code>msprime()</code> <em>slendr</em> engine to <code>slim()</code>. The latter can still interpret the same demographic model we programmed in R, just like the <code>msprime()</code> engine can, but will run the model using SLiM (and thus leveraging the new SLiM extension code that we have customized using <code>substitute_values()</code> above). We simply do this by switching from this:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>to this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">slim</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you can see, you don’t have to modify anything in your model code, just switching from <code>msprime</code> to <code>slim</code> in the line of code which produces the simulation result.</p>
<ol start="2" type="1">
<li>The customized model will not only produce a tree sequence, but will also generate a table of allele frequencies in each population (SLiM experts might have noticed the revelant SLiM code when they were inspecting <a href="https://github.com/bodkan/simgen/blob/main/slim_extension.txt"><code>slim_extension.txt</code></a>). We need to be able to load both of these files after the simulation and thus need a path to a location we can find those files. We can do this by calling the <code>slim()</code> function as <code>path &lt;- slim(..., path = TRUE)</code> (so with the extra <code>path =</code> argument). This will return a path to where the <code>slim()</code> engine saved all files with our desired results.</li>
</ol>
<p><strong>Run a simulation from the modified model of selection with the <code>slim()</code> engine as instructed in points number 1. and 2. above, then use the <code>list.files(path)</code> function in R to take a look in the directory. Which files were produced by the simulation?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-31-contents" aria-controls="callout-31" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution (you have a working SLiM installation)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-31" class="callout-31-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tstart &lt;- Sys.time()</span></span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">slim</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">samples =</span> schedule, <span class="at">path =</span> <span class="cn">TRUE</span>, <span class="at">random_seed =</span> <span class="dv">59879916</span>)</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a><span class="co"># tend &lt;- Sys.time()</span></span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a><span class="co"># tend - tstart # Time difference of 38.82014 secs</span></span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a><span class="co"># We can verify that the path not only contains a tree-sequence file but also</span></span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the table of allele frequencies.</span></span>
<span id="cb180-8"><a href="#cb180-8" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "slim.trees"     "trajectory.tsv"</code></pre>
</div>
</div>
<p>We can see that the <code>slim()</code> simulation generated a tree-sequence file (just like in previous exercises focused on <code>msprime()</code>) but it also created a new file – this was done by the SLiM customization snippet we provided to <code>compile_model()</code>.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-32-contents" aria-controls="callout-32" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution (you don’t have a working SLiM installation <em>or</em> the simulation takes too long)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-32" class="callout-32-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you don't have SLiM set up, just use the simulated results from my own</span></span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a><span class="co"># run of the same simulation</span></span>
<span id="cb182-3"><a href="#cb182-3" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/selection"</span>)</span>
<span id="cb182-4"><a href="#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="#cb182-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We can verify that the path not only contains a tree-sequence file but also</span></span>
<span id="cb182-6"><a href="#cb182-6" aria-hidden="true" tabindex="-1"></a><span class="co"># the table of allele frequencies.</span></span>
<span id="cb182-7"><a href="#cb182-7" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "slim.trees"     "trajectory.tsv"</code></pre>
</div>
</div>
<p>We can see that the <code>slim()</code> simulation generated a tree-sequence file (just like in previous exercises focused on <code>msprime()</code>) but it also created a new file – this was done by the SLiM customization snippet we provided to <code>compile_model()</code>.</p>
</div>
</div>
</div>
</section>
<section id="part-5-investigating-allele-frequency-trajectories" class="level2">
<h2 class="anchored" data-anchor-id="part-5-investigating-allele-frequency-trajectories">Part 5: Investigating allele frequency trajectories</h2>
<p><strong>Use another helper function <code>read_trajectory(path)</code> which I provided for this exercise to read the simulated frequency trajectories of the positively selected mutation in all of our populations into a variable <code>traj_df</code>. Then run a second helper function <code>plot_trajectory(traj_df)</code> to inspect the trajectories visually.</strong></p>
<p><strong>Recall that you used the function <code>substitute_values()</code> to parametrize your selection model so that the allele under selection occurs in Europeans 15 thousand years ago, and is programmed to be under very strong selection of <span class="math inline">\(s = 0.15\)</span>. Do the trajectories visualized by <code>plot_trajectory()</code> make sense given the demographic model of European prehistory plotted above?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-33-contents" aria-controls="callout-33" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-33" class="callout-33-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a>traj_df <span class="ot">&lt;-</span> <span class="fu">read_trajectory</span>(path)</span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>traj_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,604 × 4
    time pop      freq onset
   &lt;dbl&gt; &lt;fct&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 11990 EHG    0      12000
 2 11990 ANA    0      12000
 3 11990 EUR    0.0001 12000
 4 11990 YAM   NA      12000
 5 11960 EHG    0      12000
 6 11960 ANA    0      12000
 7 11960 EUR    0.0001 12000
 8 11960 YAM   NA      12000
 9 11930 EHG    0      12000
10 11930 ANA    0      12000
# ℹ 1,594 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_trajectory</span>(traj_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 554 rows containing missing values or values outside the scale range
(`geom_line()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Comparing the trajectories side-by-side with the demographic model reveals</span></span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a><span class="co"># some obvious patterns of both selection and demographic history.</span></span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(</span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_model</span>(model),</span>
<span id="cb188-5"><a href="#cb188-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_trajectory</span>(traj_df),</span>
<span id="cb188-6"><a href="#cb188-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">rel_widths =</span> <span class="fu">c</span>(<span class="fl">0.7</span>, <span class="dv">1</span>)</span>
<span id="cb188-7"><a href="#cb188-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 554 rows containing missing values or values outside the scale range
(`geom_line()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-59-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can see that the beneficial allele which appeared in the European population was under <em>extremely strong selection</em> (look how its allele frequency shoots up immediately after its first appearance!). However, we can also se how the following demographic history with multiple admixture events kept “diluting” the allele frequency (indicated by the dips in the trajectory).</p>
<p>This is the kind of <em>slendr</em> simulation which could be also very useful for simulation-based inference, like we did in the previous exercise. Just imagine having a comparable aDNA time series data with empirical allele frequency trajectory over time and using it in an ABC setting!</p>
</div>
</div>
</div>
</section>
<section id="part-6-tajimas-d-genome-wide-and-window-based-from-the-selection-model" class="level2">
<h2 class="anchored" data-anchor-id="part-6-tajimas-d-genome-wide-and-window-based-from-the-selection-model">Part 6: Tajima’s D (genome-wide and window-based) from the selection model</h2>
<p>Recall that your simulation run saved results in the location stored in the <code>path</code> variable:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "slim.trees"     "trajectory.tsv"</code></pre>
</div>
</div>
<p>From this <code>path</code>, we’ve already successfuly investigated the frequency trajectories.</p>
<p>Now let’s compute Tajima’s D on the tree sequence simulated from our selection model. Hopefully we should see an interesting pattern in our selection scan? For instance, we don’t know yet <em>where</em> in the genome is the putative locus under selection!</p>
<p>To read a tree sequence simulated with <code>slim()</code> by our customized selection setup, we need to do a bit of work. To simplify things a bit, here’s the R code which makes it possible to do. Just copy it in your <code>exercise4.R</code> script as it is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's use my own saved simulation results, so that we're all on the</span></span>
<span id="cb192-2"><a href="#cb192-2" aria-hidden="true" tabindex="-1"></a><span class="co"># same page going forward</span></span>
<span id="cb192-3"><a href="#cb192-3" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/selection"</span>)</span>
<span id="cb192-4"><a href="#cb192-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-5"><a href="#cb192-5" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb192-6"><a href="#cb192-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.path</span>(path, <span class="st">"slim.trees"</span>) <span class="sc">%&gt;%</span>  <span class="co"># 1. compose full path to the slim.trees file</span></span>
<span id="cb192-7"><a href="#cb192-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_read</span>(model) <span class="sc">%&gt;%</span>                 <span class="co"># 2. read the tree sequence file into R</span></span>
<span id="cb192-8"><a href="#cb192-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_recapitate</span>(<span class="at">Ne =</span> <span class="dv">5000</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="co"># 3. perform recapitation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Very briefly, because our tree sequence was generated by SLiM, it’s very likely that not all genealogies along the simulated genome will be fully coalesced (i.e., not all tree will have a single root). To explain why this is the case is out of the scope of this session, but read <a href="https://tskit.dev/pyslim/docs/latest/tutorial.html">here</a> if you’re interested in learning more. For the time being, it suffices to say that we can pass the (uncoalesced) tree sequence into the <code>ts_recapitate()</code> function, which then takes a SLiM tree sequence and simulates all necessary “ancestral history” that was missing on the uncoalesced trees, thus ensuring that the entire tree sequence is fully coalesced and can be correctly computed on.</p>
<p><strong>Now that you have a <code>ts</code> tree sequence object resulting from a new selection simulation run, repeat the analyses of genome-wide and window-based Tajima’s D from <em>Part 1</em> and <em>Part 2</em> of this exercise, again using the provided helper functions <code>process_tajima()</code> and <code>plot_tajima()</code>. Can you identify which locus has been the likely focal point of the positive selection? Which population shows evidence of selection? Which doesn’t and why (look again at the visualization of the demographic model above)?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-34-contents" aria-controls="callout-34" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-34" class="callout-34-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb193"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb193-2"><a href="#cb193-2" aria-hidden="true" tabindex="-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ANA
 [1] "ANA_1"  "ANA_2"  "ANA_3"  "ANA_4"  "ANA_5"  "ANA_6"  "ANA_7"  "ANA_8" 
 [9] "ANA_9"  "ANA_10" "ANA_11" "ANA_12" "ANA_13" "ANA_14" "ANA_15" "ANA_16"
[17] "ANA_17" "ANA_18" "ANA_19" "ANA_20" "ANA_21" "ANA_22" "ANA_23" "ANA_24"
[25] "ANA_25" "ANA_26" "ANA_27" "ANA_28" "ANA_29" "ANA_30" "ANA_31" "ANA_32"
[33] "ANA_33" "ANA_34" "ANA_35" "ANA_36" "ANA_37" "ANA_38" "ANA_39" "ANA_40"
[41] "ANA_41" "ANA_42" "ANA_43" "ANA_44" "ANA_45" "ANA_46" "ANA_47" "ANA_48"
[49] "ANA_49" "ANA_50"

$EHG
 [1] "EHG_1"  "EHG_2"  "EHG_3"  "EHG_4"  "EHG_5"  "EHG_6"  "EHG_7"  "EHG_8" 
 [9] "EHG_9"  "EHG_10" "EHG_11" "EHG_12" "EHG_13" "EHG_14" "EHG_15" "EHG_16"
[17] "EHG_17" "EHG_18" "EHG_19" "EHG_20" "EHG_21" "EHG_22" "EHG_23" "EHG_24"
[25] "EHG_25" "EHG_26" "EHG_27" "EHG_28" "EHG_29" "EHG_30" "EHG_31" "EHG_32"
[33] "EHG_33" "EHG_34" "EHG_35" "EHG_36" "EHG_37" "EHG_38" "EHG_39" "EHG_40"
[41] "EHG_41" "EHG_42" "EHG_43" "EHG_44" "EHG_45" "EHG_46" "EHG_47" "EHG_48"
[49] "EHG_49" "EHG_50"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30" "EUR_31" "EUR_32"
[33] "EUR_33" "EUR_34" "EUR_35" "EUR_36" "EUR_37" "EUR_38" "EUR_39" "EUR_40"
[41] "EUR_41" "EUR_42" "EUR_43" "EUR_44" "EUR_45" "EUR_46" "EUR_47" "EUR_48"
[49] "EUR_49" "EUR_50"

$YAM
 [1] "YAM_1"  "YAM_2"  "YAM_3"  "YAM_4"  "YAM_5"  "YAM_6"  "YAM_7"  "YAM_8" 
 [9] "YAM_9"  "YAM_10" "YAM_11" "YAM_12" "YAM_13" "YAM_14" "YAM_15" "YAM_16"
[17] "YAM_17" "YAM_18" "YAM_19" "YAM_20" "YAM_21" "YAM_22" "YAM_23" "YAM_24"
[25] "YAM_25" "YAM_26" "YAM_27" "YAM_28" "YAM_29" "YAM_30" "YAM_31" "YAM_32"
[33] "YAM_33" "YAM_34" "YAM_35" "YAM_36" "YAM_37" "YAM_38" "YAM_39" "YAM_40"
[41] "YAM_41" "YAM_42" "YAM_43" "YAM_44" "YAM_45" "YAM_46" "YAM_47" "YAM_48"
[49] "YAM_49" "YAM_50"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Overall Tajima's D across the 10Mb sequence still doesn't reveal any significant</span></span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a><span class="co"># deviations even in case of selection (again, not entirely unsurprising)</span></span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_tajima</span>(ts, <span class="at">sample_sets =</span> samples, <span class="at">mode =</span> <span class="st">"branch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  set         D
  &lt;chr&gt;   &lt;dbl&gt;
1 ANA    0.0882
2 EHG    0.0526
3 EUR   -0.341 
4 YAM   -0.286 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb197"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1" aria-hidden="true" tabindex="-1"></a><span class="co"># So let's look at the window-based computation again...</span></span>
<span id="cb197-2"><a href="#cb197-2" aria-hidden="true" tabindex="-1"></a>windows <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">seq</span>(<span class="dv">0</span>, ts<span class="sc">$</span>sequence_length, <span class="at">length.out =</span> <span class="dv">100</span>))</span>
<span id="cb197-3"><a href="#cb197-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb197-4"><a href="#cb197-4" aria-hidden="true" tabindex="-1"></a><span class="co"># compute genome-wide Tajima's D for each population in individual windows</span></span>
<span id="cb197-5"><a href="#cb197-5" aria-hidden="true" tabindex="-1"></a>tajima_wins <span class="ot">&lt;-</span> <span class="fu">ts_tajima</span>(ts, <span class="at">sample_sets =</span> samples, <span class="at">windows =</span> windows, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb197-6"><a href="#cb197-6" aria-hidden="true" tabindex="-1"></a>tajima_df <span class="ot">&lt;-</span> <span class="fu">process_tajima</span>(tajima_wins)</span>
<span id="cb197-7"><a href="#cb197-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb197-8"><a href="#cb197-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_tajima</span>(tajima_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You should see a clear dip in Tajima’s D around the midpoint of the DNA sequence, but only in Europeans. The beneficial allele appeared in the European population, and although the plot of the allele frequency trajectories shows that the selection dynamics has been <em>dramatically</em> affected by gene-flow events (generally causing a repeated “dilution” of the selection signal in Europeans), there has never been gene-flow (at least in our model) <em>from</em> Europeans to other populations, so the beneficial allele never had a chance to “make it” into those populations.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-38-contents" aria-controls="callout-38" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus exercises
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-38" class="callout-38-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="bonus-1-investigate-the-impact-of-recombination-around-the-selected-locus" class="level4">
<h4 class="anchored" data-anchor-id="bonus-1-investigate-the-impact-of-recombination-around-the-selected-locus">Bonus 1: Investigate the impact of recombination around the selected locus</h4>
<p>Vary the uniform recombination rate and observe what happens with Tajima’s D in windows along the genome.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-35-contents" aria-controls="callout-35" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-35" class="callout-35-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Solution: just modify the value of the <code>recombination_rate =</code> argument provided to the <code>slim()</code> function above.</p>
</div>
</div>
</div>
</section>
<section id="bonus-2-simulate-origin-of-the-allele-in-ehg" class="level4">
<h4 class="anchored" data-anchor-id="bonus-2-simulate-origin-of-the-allele-in-ehg">Bonus 2: Simulate origin of the allele in EHG</h4>
<p>Simulate the origin of the beneficial allele in the EHG population – what do the trajectories look like now? How does that change the Tajima’s D distribution along the genome in our European populations?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-36-contents" aria-controls="callout-36" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-36" class="callout-36-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use this extension in the <code>slim()</code> call, and repeat the rest of the selection-based workflow in this exercise.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>extension <span class="ot">&lt;-</span> <span class="fu">substitute_values</span>(</span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">template =</span> <span class="st">"slim_extension.txt"</span>,</span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">origin_pop =</span> <span class="st">"EHG"</span>,</span>
<span id="cb198-4"><a href="#cb198-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">s =</span> <span class="fl">0.1</span>,</span>
<span id="cb198-5"><a href="#cb198-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">onset_time =</span> <span class="dv">12000</span></span>
<span id="cb198-6"><a href="#cb198-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb198-7"><a href="#cb198-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb198-8"><a href="#cb198-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(afr, ooa, ehg, eur, ana, yam),</span>
<span id="cb198-9"><a href="#cb198-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf, <span class="at">generation_time =</span> <span class="dv">30</span>,</span>
<span id="cb198-10"><a href="#cb198-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">extension =</span> extension</span>
<span id="cb198-11"><a href="#cb198-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-3-other-statistics-in-windows" class="level4">
<h4 class="anchored" data-anchor-id="bonus-3-other-statistics-in-windows">Bonus 3: Other statistics in windows</h4>
<p>As a practice of your newly acquired tree-sequence computation skills with <em>slendr</em>, calculate some other statistics in the same windows along the simulated genome, visualize them yourself, and compare the results to the window-based Tajima’s D pattern. For instance, <code>ts_diversity()</code>, <code>ts_divergence()</code>, or <code>ts_segregating()</code> might be quite interesting to look at.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-37-contents" aria-controls="callout-37" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-37" class="callout-37-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use the same tree sequence file you’ve computed Tajima’s D on, and then apply the functions <code>ts_diversity()</code>, <code>ts_divergence()</code>, and <code>ts_segregating()</code> on that tree sequence.</p>
</div>
</div>
</div>
</section>
</div>
</div>
</div>
<!-- End of Bonus exercises -->
<hr>
</section>
</section>
<section id="playing-around-with-pca-patterns" class="level1 page-columns page-full">
<h1>Playing around with PCA patterns</h1>
<p><strong>Note:</strong> We’re unlikely to make it to this part, so this is extra bonus which can be ignored!</p>
<p>In earlier lectures you’ve learned about the <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">Principal Component Analysis</a>, a dimensional-reduction technique that’s a central piece of many (if not most) papers studying population history of many (hundreds) of individuals genotyped across many (millions) SNPs.</p>
<p>Although incredibly popular, PCA has recently been a topic of controversy, with a <a href="https://www.nature.com/articles/s41598-022-14395-4">highly combative</a> paper criticizing the application of PCA in all of population genetics, almost to a point of claiming that none of the results can be published. As an example, take a look at <a href="https://www.nature.com/articles/s41598-022-14395-4#Fig5">Figure 5</a> in the paper, demonstrating that an Indian population can be arbitrarily placed in proximity to Europeans, East Asians, and Africans, simply by wiggling the sample sizes of all groups before they applying PCA on their genotype data.</p>
<p>This activity session is not a place to debate the merit of PCA (or any other popgen method for that matter). What we can do (as we’ve done in previous exercises) is to show how <em>slendr</em> can be used in a very easy way to evaluate patterns, expected behavior, and overall dynamics of many popgen metrics in a controlled settings – and in this way, verify our intuition and check the robustness of a method in question – using PCA as another example. We’ll be specifically looking at patterns which arise in PCA depending on the exact spatio-temporal sampling from a demographic model.</p>
<section id="part-1-create-a-model" class="level2">
<h2 class="anchored" data-anchor-id="part-1-create-a-model">Part 1: Create a model</h2>
<p><strong>Start a new script named <code>exercise5.R</code> with the following setup of another toy demographic model:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb199-2"><a href="#cb199-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>(<span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb199-3"><a href="#cb199-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-4"><a href="#cb199-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"utils.R"</span>))</span>
<span id="cb199-5"><a href="#cb199-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-6"><a href="#cb199-6" aria-hidden="true" tabindex="-1"></a>popA <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popA"</span>, <span class="at">time =</span> <span class="dv">3000</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb199-7"><a href="#cb199-7" aria-hidden="true" tabindex="-1"></a>popB <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popB"</span>, <span class="at">time =</span> <span class="dv">1500</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> popA)</span>
<span id="cb199-8"><a href="#cb199-8" aria-hidden="true" tabindex="-1"></a>popC <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popC"</span>, <span class="at">time =</span> <span class="dv">1500</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> popA)</span>
<span id="cb199-9"><a href="#cb199-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-10"><a href="#cb199-10" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(<span class="fu">list</span>(popA, popB, popC), <span class="at">generation_time =</span> <span class="dv">1</span>)</span>
<span id="cb199-11"><a href="#cb199-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-12"><a href="#cb199-12" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">3000</span>, <span class="dv">0</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">200</span>),</span>
<span id="cb199-13"><a href="#cb199-13" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">list</span>(popA, <span class="dv">10</span>), <span class="fu">list</span>(popB, <span class="dv">10</span>), <span class="fu">list</span>(popC, <span class="dv">10</span>))</span>
<span id="cb199-14"><a href="#cb199-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-15"><a href="#cb199-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-65-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As you can see, this model describes the demographic history of three populations: one ancestral population “popA” starting at 3000 generations ago, which splits into two populations “popB” and “popC” the same time at 1500 generations ago. We den instruct <em>slendr</em> to record 10 individuals from each of the three populations starting from 3000 generations ago all the way to 0 generations ago (i.e. the “present”), every 200 generations (remeber the <code>seq()</code> R function!).</p>
</section>
<section id="part-2-simulate-a-mutated-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="part-2-simulate-a-mutated-tree-sequence">Part 2: Simulate a (mutated!) tree sequence</h2>
<p>To be able to run PCA using the <a href="https://christianhuber.github.io/smartsnp/"><em>smartsnp</em> R package</a> (below), we will need to simulate data in the <a href="https://bodkan.net/admixr/articles/01-tutorial.html#a-note-about-eigenstrat-format">EIGENSTRAT</a> file format. And to do <em>that</em>, we need our tree sequence with mutations.</p>
<p>Recall that all of our previous exercises managed to do away with mutations completely, owing to the amazing nature of the succint tree sequence data structure invented by the people behind the <em>tskit</em> project. However, all traditional popgen software and tools still rely on genotype data, which is why we now have to simulate mutations as well. Luckily, this is very easy – instead of the traditional</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> ..., <span class="at">recombination_rate =</span> ..., <span class="at">samples =</span> ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>we will run this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb201"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First run a normal msprime simulation creating a tree-sequence object, then</span></span>
<span id="cb201-2"><a href="#cb201-2" aria-hidden="true" tabindex="-1"></a><span class="co"># directly pipe it into a function which adds (neutral!) mutations to it</span></span>
<span id="cb201-3"><a href="#cb201-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> ..., <span class="at">recombination_rate =</span> ..., <span class="at">samples =</span> ...) <span class="sc">%&gt;%</span></span>
<span id="cb201-4"><a href="#cb201-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>which is equivalent to running this without the <code>%&gt;%</code> “pipe operator”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb202"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First run a normal msprime simulation creating a tree-sequence object...</span></span>
<span id="cb202-2"><a href="#cb202-2" aria-hidden="true" tabindex="-1"></a>ts_nomuts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> ..., <span class="at">recombination_rate =</span> ..., <span class="at">samples =</span> ...)</span>
<span id="cb202-3"><a href="#cb202-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ... then add (neutral!) mutations to it</span></span>
<span id="cb202-4"><a href="#cb202-4" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(ts_nomuts, <span class="at">mutation_rate =</span> ...)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With that out of the way, <strong>simulate a tree sequence from the popA/popB/popC <code>model</code> above, which will be 50 Mb (50e6) long, with a recombination rate 1e-8, and overlay mutations on it at a rate 1e-8. Check that it has mutations either by typing out <code>ts</code> in the console and looking for a “Mutations” section of the summary, or by using the function <code>ts_table(ts, "mutations")</code>. Then count how many individuals you have recorded for each population using the table produced by <code>ts_samples()</code>.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-39-contents" aria-controls="callout-39" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-39" class="callout-39-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First run a normal msprime simulation creating a tree-sequence object, then</span></span>
<span id="cb203-2"><a href="#cb203-2" aria-hidden="true" tabindex="-1"></a><span class="co"># directly pipe it into a function which adds (neutral!) mutations to it</span></span>
<span id="cb203-3"><a href="#cb203-3" aria-hidden="true" tabindex="-1"></a>ts_nomuts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>, <span class="at">random_seed =</span> <span class="dv">1702182272</span>)</span>
<span id="cb203-4"><a href="#cb203-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-5"><a href="#cb203-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Notice we have no mutations on the tree sequence, just as before...</span></span>
<span id="cb203-6"><a href="#cb203-6" aria-hidden="true" tabindex="-1"></a>ts_nomuts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     160114║
╟───────────────┼───────────╢
║Sequence Length│   50000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │        620║
╟───────────────┼───────────╢
║Total Size     │   26.1 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤══════╤═════════╤════════════╗
║Table      │Rows  │Size     │Has Metadata║
╠═══════════╪══════╪═════════╪════════════╣
║Edges      │619886│ 18.9 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Individuals│   310│  8.5 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Migrations │     0│  8 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Mutations  │     0│ 16 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Nodes      │ 92209│  2.5 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Populations│     3│303 Bytes│         Yes║
╟───────────┼──────┼─────────┼────────────╢
║Provenances│     1│  5.2 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Sites      │     0│ 16 Bytes│          No║
╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
<div class="sourceCode cell-code" id="cb205"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(ts_nomuts, <span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb205-2"><a href="#cb205-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb205-3"><a href="#cb205-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ... but we have them now!</span></span>
<span id="cb205-4"><a href="#cb205-4" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     160114║
╟───────────────┼───────────╢
║Sequence Length│   50000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │        620║
╟───────────────┼───────────╢
║Total Size     │   36.1 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤══════╤═════════╤════════════╗
║Table      │Rows  │Size     │Has Metadata║
╠═══════════╪══════╪═════════╪════════════╣
║Edges      │619886│ 18.9 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Individuals│   310│  8.5 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Migrations │     0│  8 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Mutations  │168839│  6.0 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Nodes      │ 92209│  2.5 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Populations│     3│303 Bytes│         Yes║
╟───────────┼──────┼─────────┼────────────╢
║Provenances│     2│  6.0 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Sites      │168537│  4.0 MiB│          No║
╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
<div class="sourceCode cell-code" id="cb207"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the table of individuals (and process it a bit for tidier plotting later)</span></span>
<span id="cb207-2"><a href="#cb207-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">pop =</span> <span class="fu">factor</span>(pop, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"popA"</span>, <span class="st">"popB"</span>, <span class="st">"popC"</span>)))</span>
<span id="cb207-3"><a href="#cb207-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb207-4"><a href="#cb207-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Count how many individuals do we have for each population</span></span>
<span id="cb207-5"><a href="#cb207-5" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop) <span class="sc">%&gt;%</span> <span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
# Groups:   pop [3]
  pop       n
  &lt;fct&gt; &lt;int&gt;
1 popA    150
2 popB     80
3 popC     80</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-3-converting-a-tree-sequence-into-eigenstrat" class="level2">
<h2 class="anchored" data-anchor-id="part-3-converting-a-tree-sequence-into-eigenstrat">Part 3: Converting a tree sequence into EIGENSTRAT</h2>
<p>The function to use for converting a tree-sequence object we have in R (in our exercises the thing we usually had in the <code>ts</code> variable) to disk in form of genotypes in the EIGENSTRAT format is called <code>ts_eigenstrat()</code>. The standard way to call it (but see <code>?ts_eigenstrat</code> for more options) is like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/a/desired/EIGENSTRAT/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Which creates three files <code>.ind</code>, <code>.snp</code>, and <code>.geno</code> as: - <code>path/to/a/desired/EIGENSTRAT/prefix.ind</code>, - <code>path/to/a/desired/EIGENSTRAT/prefix.snp</code>, and - <code>path/to/a/desired/EIGENSTRAT/prefix.geno</code>,</p>
<p>just as you would expect for any EIGENSTRAT file.</p>
<p><strong>Take your tree sequence <code>ts</code> just just simulated, and convert it to EIGENSTRAT format under the prefix <code>data/ABC_all</code>.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-40-contents" aria-controls="callout-40" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-40" class="callout-40-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb210"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="st">"data/ABC_all"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>194 multiallelic sites (0.115% out of 168537 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>EIGENSTRAT object
=================
components:
  ind file: data/ABC_all.ind
  snp file: data/ABC_all.snp
  geno file: data/ABC_all.geno</code></pre>
</div>
</div>
</div>
</div>
</div>
<p><strong>Check that the EIGENSTRAT files really appeared at the path that you specified (in the terminal).</strong></p>
</section>
<section id="part-4-inspect-the-eigenstrat-data-produced-by-slendr" class="level2">
<h2 class="anchored" data-anchor-id="part-4-inspect-the-eigenstrat-data-produced-by-slendr">Part 4: Inspect the EIGENSTRAT data produced by <em>slendr</em></h2>
<p>Years ago I developed a small R package to help me with <span class="math inline">\(f\)</span>-statistics based projects using the ADMIXTOOLS software (which operates on data in the EIGENSTRAT file format), called <a href="https://bodkan.net/admixr/">admixr</a></p>
<p><strong>Use the following code to examine one of the EIGENSTRAT data sets you’ve just created. Just look at the results and see if they make sense in terms of what you’ve learned about this in earlier lectures.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(admixr)</span>
<span id="cb213-2"><a href="#cb213-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb213-3"><a href="#cb213-3" aria-hidden="true" tabindex="-1"></a>eigen <span class="ot">&lt;-</span> <span class="fu">eigenstrat</span>(<span class="st">"&lt;prefix of a trio of EIGENSTRAT .ind/.snp/.geno files"</span>)</span>
<span id="cb213-4"><a href="#cb213-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb213-5"><a href="#cb213-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out a summary of the EIGENSTRAT data</span></span>
<span id="cb213-6"><a href="#cb213-6" aria-hidden="true" tabindex="-1"></a>eigen</span>
<span id="cb213-7"><a href="#cb213-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb213-8"><a href="#cb213-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .ind file as a table into R</span></span>
<span id="cb213-9"><a href="#cb213-9" aria-hidden="true" tabindex="-1"></a><span class="fu">read_ind</span>(eigen)</span>
<span id="cb213-10"><a href="#cb213-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .snp file as a table into R</span></span>
<span id="cb213-11"><a href="#cb213-11" aria-hidden="true" tabindex="-1"></a><span class="fu">read_snp</span>(eigen)</span>
<span id="cb213-12"><a href="#cb213-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .geno file as a table into R</span></span>
<span id="cb213-13"><a href="#cb213-13" aria-hidden="true" tabindex="-1"></a><span class="fu">read_geno</span>(eigen)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-41-contents" aria-controls="callout-41" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-41" class="callout-41-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb214"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(admixr)</span>
<span id="cb214-2"><a href="#cb214-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-3"><a href="#cb214-3" aria-hidden="true" tabindex="-1"></a>eigen <span class="ot">&lt;-</span> <span class="fu">eigenstrat</span>(<span class="st">"data/ABC_all"</span>)</span>
<span id="cb214-4"><a href="#cb214-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-5"><a href="#cb214-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out a summary of the EIGENSTRAT data</span></span>
<span id="cb214-6"><a href="#cb214-6" aria-hidden="true" tabindex="-1"></a>eigen</span>
<span id="cb214-7"><a href="#cb214-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb214-8"><a href="#cb214-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .ind file as a table into R</span></span>
<span id="cb214-9"><a href="#cb214-9" aria-hidden="true" tabindex="-1"></a><span class="fu">read_ind</span>(eigen)</span>
<span id="cb214-10"><a href="#cb214-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .snp file as a table into R</span></span>
<span id="cb214-11"><a href="#cb214-11" aria-hidden="true" tabindex="-1"></a><span class="fu">read_snp</span>(eigen)</span>
<span id="cb214-12"><a href="#cb214-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the .geno file as a table into R</span></span>
<span id="cb214-13"><a href="#cb214-13" aria-hidden="true" tabindex="-1"></a><span class="fu">read_geno</span>(eigen)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-5-principal-component-analysis-on-the-entire-simulated-data-set" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="part-5-principal-component-analysis-on-the-entire-simulated-data-set">Part 5: Principal Component Analysis on the entire simulated data set</h2>
<p>Now, at long last, we have everything we need to be able to run ABC on the data generated by our <em>slendr</em> model. To avoid making this exercise even longer, I provided a helper function for you called <code>plot_pca()</code>. But this function isn’t doing anything magical – it uses the <a href="https://christianhuber.github.io/smartsnp/"><em>smartsnp</em> R package</a> to compute the principal components and visualize the results using <em>ggplot2</em>. This is something many of you could do given enough time but we want to focus on simulations and PCA, not pure R coding. If you’re interested, take a look at my implementation of <code>plot_pca()</code> <a href="https://github.com/bodkan/simgen/blob/main/utils.R#L15">here</a>.</p>
<p>Here’s how you can use this function (remeber that you need to put <code>source(here::here("utils.R"))</code> into your script!):</p>
<ol type="1">
<li>Plot PCA while coloring each individual by their population assignment:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"path/to/prefix"</span>, <span class="sc">&lt;</span>tree sequence used to create EIGENSTRAT<span class="sc">&gt;</span>, <span class="at">color_by =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Plot PCA while coloring each individual by their time of sampling:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb216"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"path/to/prefix"</span>, <span class="sc">&lt;</span>tree sequence used to create EIGENSTRAT<span class="sc">&gt;</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>By default, the function plots PC 1 vs PC 2, but you can customize things by providing an optional argument <code>pc =</code> like this:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"path/to/prefix"</span>, <span class="sc">&lt;</span>tree sequence used to create EIGENSTRAT<span class="sc">&gt;</span>, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Use the provided <code>plot_pca()</code> function to run PCA based on genotypes for all recorded individuals that you just converted as EIGENSTRAT <code>"data/ABC_all"</code> from the <code>ts</code> tree sequence. Visualize PC 1 vs PC 2 by first ccolor each individual by their population label (<code>color_by = "pop"</code>) then by the time of their sampling (<code>color_by = "time"</code>).</strong></p>
<p><strong>Does the PCA of PC 1 vs PC 2 capture the relationship between all individuals across the populations <em>and across time</em>?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-42-contents" aria-controls="callout-42" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-42" class="callout-42-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb218"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was not found. Generating it now (this might take a moment)...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-77-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb220"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"time"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-77-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It looks like the PCA from PC 1 vs 2 cannot “partition out” the drift along the ancestral “popA” lineage prior to the population splits!</p>
</div>
</div>
</div>
<p><strong>Use <code>plot_pca()</code> to compute the PCAon this exact same data, but examine how does the shape of the PCA scatterplot change when you switch the pairs of PCs plotted (i.e., PC 2 vs PC 3, PC 3 vs PC 4, PC 4 vs PC 6, etc.). Which pair of PCs does the best job at recapitulating the demographic model?</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> We’re doing this purely for educational purposes and for fun, using an extremely idealistic demographic model which is perfectly known (by definition, because we simulated it) and perfect sampling scheme. The point is to explore what does doing a PCA mean in practice, visually, and to built intuition into it.</p>
</div></div><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-43-contents" aria-controls="callout-43" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-43" class="callout-43-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb222"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb222-1"><a href="#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  We can see that the overall shape of the demographic model tree is now nicely</span></span>
<span id="cb222-2"><a href="#cb222-2" aria-hidden="true" tabindex="-1"></a><span class="co"># reflected in the PCA shape</span></span>
<span id="cb222-3"><a href="#cb222-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb224"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"time"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb226"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb226-1"><a href="#cb226-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb228"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb228-1"><a href="#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"time"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb230"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb230-1"><a href="#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Things are getting progressively wilder! </span></span>
<span id="cb230-2"><a href="#cb230-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb232"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb232-1"><a href="#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"time"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb234"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb234-1"><a href="#cb234-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ...</span></span>
<span id="cb234-2"><a href="#cb234-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"pop"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">6</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb236"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb236-1"><a href="#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/ABC_all"</span>, ts, <span class="at">color_by =</span> <span class="st">"time"</span>, <span class="at">pc =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">6</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was found. Loading it to save computational time...</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-78-8.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-45-contents" aria-controls="callout-45" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus exercises
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-45" class="callout-45-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="bonus-1-tree-sequence-simplification-and-eigenstrat-conversion" class="level3">
<h3 class="anchored" data-anchor-id="bonus-1-tree-sequence-simplification-and-eigenstrat-conversion">Bonus 1: Tree-sequence simplification and EIGENSTRAT conversion</h3>
<p>One of our goals in this exercise was to investigate how does the shape of a PCA look like based on the sampling of individuals across populations and also across time – all of that from the <em>same</em> demographic history. In order to do that, we need to be able to select only a defined subset of individuals from a given tree sequence. Which brings us to the last tree-sequence processing function in <em>slendr</em> callsed <code>ts_simplify()</code>. Implemented on top of the <code>simplify()</code> method in <em>tskit</em>, it has a very simple interface:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb238"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb238-1"><a href="#cb238-1" aria-hidden="true" tabindex="-1"></a>ts_small <span class="ot">&lt;-</span> <span class="fu">ts_simplify</span>(ts_big, <span class="at">simplify_to =</span> <span class="fu">c</span>(<span class="sc">&lt;</span>subset of individuals as a vector<span class="sc">&gt;</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function call creates a <em>new tree sequence</em>, which is smaller and only contains a those individuals whose names were specified in <code>simplify_to =</code> (again, we’re talking about the “symbolic names” of individuals, such as “NEA_1”, “AFR_42”, etc., not integer numbers of <em>tskit</em> nodes).</p>
<p>Whenever you want to create smaller subsets of a large tree sequence, it is often helpful to work with the table of all individuals in the original tree sequence, because it contains every individual’s <code>name</code>, <code>pop</code> assignment and the <code>time</code> in which it lived, so let’s save it for further use now:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb239"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb239-1"><a href="#cb239-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts)</span>
<span id="cb239-2"><a href="#cb239-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb239-3"><a href="#cb239-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 310</code></pre>
</div>
</div>
<p>For instance, we can get only individuals from “popB” and “popC” sampled at the present using this code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb241"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb241-1"><a href="#cb241-1" aria-hidden="true" tabindex="-1"></a>subset <span class="ot">&lt;-</span> <span class="fu">filter</span>(samples, pop <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"popB"</span>, <span class="st">"popC"</span>), time <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb241-2"><a href="#cb241-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb241-3"><a href="#cb241-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(subset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20</code></pre>
</div>
</div>
<p>You know that the table of samples contains the <code>name</code> of each individual, which you can access as <code>subset$name</code>. <strong>Use the <code>ts_simplify()</code> function to create a new tree sequence called <code>ts_BC0</code> which contains only this subset of individuals. Check that it really does contain only the defined subset of individuals using <code>ts_samples(ts_BC0)</code>.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-44-contents" aria-controls="callout-44" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-44" class="callout-44-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb243"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb243-1"><a href="#cb243-1" aria-hidden="true" tabindex="-1"></a>ts_BC0 <span class="ot">&lt;-</span> <span class="fu">ts_simplify</span>(ts, <span class="at">simplify_to =</span> subset<span class="sc">$</span>name)</span>
<span id="cb243-2"><a href="#cb243-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb243-3"><a href="#cb243-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts_BC0)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 popB_71     0 popB 
 2 popB_72     0 popB 
 3 popB_73     0 popB 
 4 popB_74     0 popB 
 5 popB_75     0 popB 
 6 popB_76     0 popB 
 7 popB_77     0 popB 
 8 popB_78     0 popB 
 9 popB_79     0 popB 
10 popB_80     0 popB 
11 popC_71     0 popC 
12 popC_72     0 popC 
13 popC_73     0 popC 
14 popC_74     0 popC 
15 popC_75     0 popC 
16 popC_76     0 popC 
17 popC_77     0 popC 
18 popC_78     0 popC 
19 popC_79     0 popC 
20 popC_80     0 popC </code></pre>
</div>
</div>
</div>
</div>
</div>
<p>When you have a smaller tree sequence like this, you can convert it to an EIGENSTRAT file format using <code>ts_eigenstrat()</code> just like we did above.</p>
</section>
<section id="part-2-pca-visualization-on-subsets-of-individuals" class="level3">
<h3 class="anchored" data-anchor-id="part-2-pca-visualization-on-subsets-of-individuals">Part 2: PCA visualization on subsets of individuals</h3>
<p><strong>TODO</strong>: Try to replicate some features of <a href="https://www.nature.com/articles/s41598-022-14395-4#Fig5">the results</a> discovered for EUR/ASIAN/AFR/Indian populations.</p>
</section>
</div>
</div>
</div>
<!-- End of Bonus exercises -->
</section>
</section>
<section id="spatial-pca" class="level1">
<h1>Spatial PCA</h1>
<p><strong>WIP:</strong> Exploring how (and when) PCA patterns recapitulate geographical history of populations, inspired by this figure by Novembre <em>et al.</em> (2008).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/novembre2008a.jpg" class="img-fluid figure-img"></p>
<figcaption>Novembre et al., 2008</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb245"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb245-1"><a href="#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb245-2"><a href="#cb245-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb247"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb247-1"><a href="#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(admixr)</span>
<span id="cb247-2"><a href="#cb247-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb247-3"><a href="#cb247-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb247-4"><a href="#cb247-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"utils.R"</span>)</span>
<span id="cb247-5"><a href="#cb247-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb247-6"><a href="#cb247-6" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">landscape_model</span>(<span class="at">rate =</span> <span class="fl">0.3</span>, <span class="at">Ne =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `ne_110m_land' from data source 
  `/private/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T/RtmplT1OZW/naturalearth/ne_110m_land.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 127 features and 3 fields
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: -180 ymin: -90 xmax: 180 ymax: 83.64513
Geodetic CRS:  WGS 84</code></pre>
</div>
<div class="sourceCode cell-code" id="cb249"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb249-1"><a href="#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pdf("exercise5.pdf", width = 8, height = 5)</span></span>
<span id="cb249-2"><a href="#cb249-2" aria-hidden="true" tabindex="-1"></a><span class="co"># for (n in c(50, 25, 10, 5, 2, 1)) {</span></span>
<span id="cb249-3"><a href="#cb249-3" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb249-4"><a href="#cb249-4" aria-hidden="true" tabindex="-1"></a>  schedule <span class="ot">&lt;-</span> <span class="fu">landscape_sampling</span>(model, n)</span>
<span id="cb249-5"><a href="#cb249-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb249-6"><a href="#cb249-6" aria-hidden="true" tabindex="-1"></a>  ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule, <span class="at">sequence_length =</span> <span class="fl">20e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> <span class="fu">ts_mutate</span>(<span class="fl">1e-8</span>)</span>
<span id="cb249-7"><a href="#cb249-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb249-8"><a href="#cb249-8" aria-hidden="true" tabindex="-1"></a>  samples <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts)</span>
<span id="cb249-9"><a href="#cb249-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb249-10"><a href="#cb249-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_eigenstrat</span>(ts, <span class="st">"data/spatial_pca1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>446 multiallelic sites (0.276% out of 161460 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>EIGENSTRAT object
=================
components:
  ind file: data/spatial_pca1.ind
  snp file: data/spatial_pca1.snp
  geno file: data/spatial_pca1.geno</code></pre>
</div>
<div class="sourceCode cell-code" id="cb252"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb252-1"><a href="#cb252-1" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_pca</span>(<span class="st">"data/spatial_pca1"</span>, ts, <span class="at">model =</span> <span class="st">"map"</span>, <span class="at">color_by =</span> <span class="st">"pop"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was not found. Generating it now (this might take a moment)...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: All gene-flow event will be visualized at once. If you wish to visualize
gene flows at a particular point in time, use the `time` argument.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Non-spatial populations in your model won't be visualized</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-83-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb256"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb256-1"><a href="#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb256-2"><a href="#cb256-2" aria-hidden="true" tabindex="-1"></a><span class="co"># dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="per-population-ne-values" class="level3">
<h3 class="anchored" data-anchor-id="per-population-ne-values">Per-population Ne values</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb257"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb257-1"><a href="#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb257-2"><a href="#cb257-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb259"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb259-1"><a href="#cb259-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(admixr)</span>
<span id="cb259-2"><a href="#cb259-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb259-3"><a href="#cb259-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb259-4"><a href="#cb259-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"utils.R"</span>)</span>
<span id="cb259-5"><a href="#cb259-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb259-6"><a href="#cb259-6" aria-hidden="true" tabindex="-1"></a>Ne <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb259-7"><a href="#cb259-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">p1 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-8"><a href="#cb259-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">p2 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-9"><a href="#cb259-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">p3 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-10"><a href="#cb259-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">p4 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-11"><a href="#cb259-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">p5 =</span> <span class="dv">100</span>,</span>
<span id="cb259-12"><a href="#cb259-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">p6 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-13"><a href="#cb259-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">p7 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-14"><a href="#cb259-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">p8 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-15"><a href="#cb259-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">p9 =</span> <span class="dv">10000</span>,</span>
<span id="cb259-16"><a href="#cb259-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">p10 =</span> <span class="dv">10000</span></span>
<span id="cb259-17"><a href="#cb259-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb259-18"><a href="#cb259-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb259-19"><a href="#cb259-19" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">landscape_model</span>(<span class="at">rate =</span> <span class="fl">0.3</span>, <span class="at">Ne =</span> Ne)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `ne_110m_land' from data source 
  `/private/var/folders/70/b_q2zdh116b9pfg29p03sx600000gn/T/RtmplT1OZW/naturalearth/ne_110m_land.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 127 features and 3 fields
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: -180 ymin: -90 xmax: 180 ymax: 83.64513
Geodetic CRS:  WGS 84</code></pre>
</div>
<div class="sourceCode cell-code" id="cb261"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb261-1"><a href="#cb261-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb261-2"><a href="#cb261-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">p1 =</span> <span class="dv">50</span>,</span>
<span id="cb261-3"><a href="#cb261-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">p2 =</span> <span class="dv">50</span>,</span>
<span id="cb261-4"><a href="#cb261-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">p3 =</span> <span class="dv">50</span>,</span>
<span id="cb261-5"><a href="#cb261-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">p4 =</span> <span class="dv">50</span>,</span>
<span id="cb261-6"><a href="#cb261-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">p5 =</span> <span class="dv">5</span>,</span>
<span id="cb261-7"><a href="#cb261-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">p6 =</span> <span class="dv">50</span>,</span>
<span id="cb261-8"><a href="#cb261-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">p7 =</span> <span class="dv">50</span>,</span>
<span id="cb261-9"><a href="#cb261-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">p8 =</span> <span class="dv">50</span>,</span>
<span id="cb261-10"><a href="#cb261-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">p9 =</span> <span class="dv">50</span>,</span>
<span id="cb261-11"><a href="#cb261-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">p10 =</span> <span class="dv">50</span></span>
<span id="cb261-12"><a href="#cb261-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb261-13"><a href="#cb261-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb261-14"><a href="#cb261-14" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">landscape_sampling</span>(model, n)</span>
<span id="cb261-15"><a href="#cb261-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb261-16"><a href="#cb261-16" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule, <span class="at">sequence_length =</span> <span class="fl">20e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span> <span class="fu">ts_mutate</span>(<span class="fl">1e-8</span>)</span>
<span id="cb261-17"><a href="#cb261-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb261-18"><a href="#cb261-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="st">"data/spatial_pca2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>1020 multiallelic sites (0.404% out of 252489 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>EIGENSTRAT object
=================
components:
  ind file: data/spatial_pca2.ind
  snp file: data/spatial_pca2.snp
  geno file: data/spatial_pca2.geno</code></pre>
</div>
<div class="sourceCode cell-code" id="cb264"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb264-1"><a href="#cb264-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_ind</span>(<span class="fu">eigenstrat</span>(<span class="st">"data/spatial_pca2"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 455 × 3
   id    sex   label
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
 1 p1_1  U     p1_1 
 2 p1_2  U     p1_2 
 3 p1_3  U     p1_3 
 4 p1_4  U     p1_4 
 5 p1_5  U     p1_5 
 6 p1_6  U     p1_6 
 7 p1_7  U     p1_7 
 8 p1_8  U     p1_8 
 9 p1_9  U     p1_9 
10 p1_10 U     p1_10
# ℹ 445 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb266"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb266-1"><a href="#cb266-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 455 × 3
   name   time pop  
   &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
 1 p1_1  10000 p1   
 2 p1_2  10000 p1   
 3 p1_3  10000 p1   
 4 p1_4  10000 p1   
 5 p1_5  10000 p1   
 6 p1_6  10000 p1   
 7 p1_7  10000 p1   
 8 p1_8  10000 p1   
 9 p1_9  10000 p1   
10 p1_10 10000 p1   
# ℹ 445 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb268"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb268-1"><a href="#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(<span class="st">"data/spatial_pca2"</span>, ts, <span class="at">model =</span> <span class="st">"map"</span>, <span class="at">color =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PCA cache for the given EIGENSTRAT data was not found. Generating it now (this might take a moment)...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: All gene-flow event will be visualized at once. If you wish to visualize
gene flows at a particular point in time, use the `time` argument.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Non-spatial populations in your model won't be visualized</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-84-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="ancestry-tracts-and-chromosome-painintg" class="level1">
<h1>Ancestry tracts and chromosome painintg</h1>
<p><strong>WIP:</strong> Extracting ancestry tracts, chromosome painting, and using admixture tracts and LD patterns for dating admixture events.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb272"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb272-1"><a href="#cb272-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb272-2"><a href="#cb272-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb274"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb274-1"><a href="#cb274-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb274-2"><a href="#cb274-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-3"><a href="#cb274-3" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"utils.R"</span>)</span>
<span id="cb274-4"><a href="#cb274-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-5"><a href="#cb274-5" aria-hidden="true" tabindex="-1"></a>popZ <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popZ"</span>, <span class="at">time =</span> <span class="dv">3000</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb274-6"><a href="#cb274-6" aria-hidden="true" tabindex="-1"></a>popX <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popX"</span>, <span class="at">time =</span> <span class="dv">1500</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> popZ)</span>
<span id="cb274-7"><a href="#cb274-7" aria-hidden="true" tabindex="-1"></a>popY <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"popY"</span>, <span class="at">time =</span> <span class="dv">1500</span>, <span class="at">N =</span> <span class="dv">5000</span>, <span class="at">parent =</span> popZ)</span>
<span id="cb274-8"><a href="#cb274-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-9"><a href="#cb274-9" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> popX, <span class="at">to =</span> popY, <span class="at">rate =</span> <span class="fl">0.2</span>, <span class="at">start =</span> <span class="dv">800</span>, <span class="at">end =</span> <span class="dv">799</span>)</span>
<span id="cb274-10"><a href="#cb274-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-11"><a href="#cb274-11" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(<span class="fu">list</span>(popZ, popX, popY), <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">gene_flow =</span> gf)</span>
<span id="cb274-12"><a href="#cb274-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-13"><a href="#cb274-13" aria-hidden="true" tabindex="-1"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb274-14"><a href="#cb274-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">seq</span>(<span class="dv">1500</span>, <span class="dv">0</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">100</span>), <span class="fu">list</span>(popY, <span class="dv">50</span>)),</span>
<span id="cb274-15"><a href="#cb274-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(popX, <span class="dv">10</span>), <span class="fu">list</span>(popZ, <span class="dv">10</span>))</span>
<span id="cb274-16"><a href="#cb274-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb274-17"><a href="#cb274-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb274-18"><a href="#cb274-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the function plot_model() to make sure that the model and the sampling schedule</span></span>
<span id="cb274-19"><a href="#cb274-19" aria-hidden="true" tabindex="-1"></a><span class="co"># are defined correctly (there's no such thing as too many sanity checks when doing research)</span></span>
<span id="cb274-20"><a href="#cb274-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>, <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-85-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb275"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb275-1"><a href="#cb275-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb275-2"><a href="#cb275-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb275-3"><a href="#cb275-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb275-4"><a href="#cb275-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb275-5"><a href="#cb275-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb275-6"><a href="#cb275-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb275-7"><a href="#cb275-7" aria-hidden="true" tabindex="-1"></a>tracts <span class="ot">&lt;-</span> <span class="fu">ts_tracts</span>(ts, <span class="at">census =</span> <span class="dv">800</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
PopAncestry summary
Number of ancestral populations:    3
Number of sample chromosomes:       1540
Number of ancestors:            55565
Total length of genomes:        154000000000.000000
Ancestral coverage:             84000000000.000000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb277"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb277-1"><a href="#cb277-1" aria-hidden="true" tabindex="-1"></a>tracts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 51,105 × 10
# Groups:   name [400]
   name     haplotype  time pop   source_pop    left  right length source_pop_id
   &lt;chr&gt;        &lt;int&gt; &lt;dbl&gt; &lt;chr&gt; &lt;fct&gt;        &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;         &lt;dbl&gt;
 1 popY_351         1   700 popY  popX        0      2.15e6 2.15e6             1
 2 popY_351         1   700 popY  popX        1.55e7 1.69e7 1.44e6             1
 3 popY_351         1   700 popY  popX        1.85e7 1.96e7 1.09e6             1
 4 popY_351         1   700 popY  popX        2.12e7 2.26e7 1.36e6             1
 5 popY_351         1   700 popY  popX        3.95e7 4.07e7 1.24e6             1
 6 popY_351         1   700 popY  popX        4.24e7 4.28e7 3.76e5             1
 7 popY_351         1   700 popY  popX        4.41e7 4.75e7 3.38e6             1
 8 popY_351         1   700 popY  popX        5.89e7 5.94e7 5.47e5             1
 9 popY_351         1   700 popY  popX        5.97e7 6.04e7 7.07e5             1
10 popY_351         1   700 popY  popX        6.46e7 6.51e7 4.84e5             1
# ℹ 51,095 more rows
# ℹ 1 more variable: node_id &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb279"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb279-1"><a href="#cb279-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Select sampled individuals from different ages (remember, we recorded oldest individuals</span></span>
<span id="cb279-2"><a href="#cb279-2" aria-hidden="true" tabindex="-1"></a><span class="co"># at 1000 generations ago, the youngest individuals at 0 generations ago). Use the function</span></span>
<span id="cb279-3"><a href="#cb279-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot_tracts to visualize their ancestry tracts. What do you see in terms of the number of</span></span>
<span id="cb279-4"><a href="#cb279-4" aria-hidden="true" tabindex="-1"></a><span class="co"># tracts and the length of tracts across these individuals? Can you eyeball some distinctive</span></span>
<span id="cb279-5"><a href="#cb279-5" aria-hidden="true" tabindex="-1"></a><span class="co"># pattern?</span></span>
<span id="cb279-6"><a href="#cb279-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb279-7"><a href="#cb279-7" aria-hidden="true" tabindex="-1"></a>subset_inds <span class="ot">&lt;-</span> tracts <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(time) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>(name) <span class="sc">%&gt;%</span> <span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(name)</span>
<span id="cb279-8"><a href="#cb279-8" aria-hidden="true" tabindex="-1"></a>subset_inds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "popY_728" "popY_664" "popY_604" "popY_578" "popY_539" "popY_478" "popY_415"
[8] "popY_353"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb281"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb281-1"><a href="#cb281-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_tracts</span>(tracts, subset_inds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-85-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb282"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb282-1"><a href="#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="co"># It looks like the older individual is, the closer they lived to the start of the admixture event,</span></span>
<span id="cb282-2"><a href="#cb282-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and the longer the tracts they carry will be. Compute the average length of a ancestry tracts in each</span></span>
<span id="cb282-3"><a href="#cb282-3" aria-hidden="true" tabindex="-1"></a><span class="co"># sample age group and visualize the length distribution of these tracts based on their age:</span></span>
<span id="cb282-4"><a href="#cb282-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb282-5"><a href="#cb282-5" aria-hidden="true" tabindex="-1"></a>tracts <span class="sc">%&gt;%</span></span>
<span id="cb282-6"><a href="#cb282-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(time) <span class="sc">%&gt;%</span></span>
<span id="cb282-7"><a href="#cb282-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">mean</span>(length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 2
   time `mean(length)`
  &lt;dbl&gt;          &lt;dbl&gt;
1     0        159816.
2   100        177905.
3   200        214291.
4   300        250490.
5   400        307885.
6   500        426322.
7   600        631109.
8   700       1221911.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb284"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb284-1"><a href="#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(tracts, <span class="fu">aes</span>(length, <span class="at">color =</span> <span class="fu">factor</span>(time))) <span class="sc">+</span></span>
<span id="cb284-2"><a href="#cb284-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb284-3"><a href="#cb284-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">3e6</span>)) <span class="sc">+</span></span>
<span id="cb284-4"><a href="#cb284-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-85-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb285"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb285-1"><a href="#cb285-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, try to work backwards. Assuming you have the following distribution of tract lengths...</span></span>
<span id="cb285-2"><a href="#cb285-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-3"><a href="#cb285-3" aria-hidden="true" tabindex="-1"></a>tracts <span class="ot">&lt;-</span> <span class="fu">filter</span>(tracts, time <span class="sc">==</span> <span class="dv">0</span>, length <span class="sc">&lt;=</span> <span class="fl">1e6</span>)</span>
<span id="cb285-4"><a href="#cb285-4" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">hist</span>(tracts<span class="sc">$</span>length, <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb285-5"><a href="#cb285-5" aria-hidden="true" tabindex="-1"></a>length <span class="ot">&lt;-</span> bins<span class="sc">$</span>mids</span>
<span id="cb285-6"><a href="#cb285-6" aria-hidden="true" tabindex="-1"></a>density <span class="ot">&lt;-</span> bins<span class="sc">$</span>density</span>
<span id="cb285-7"><a href="#cb285-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-8"><a href="#cb285-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(length, density)</span>
<span id="cb285-9"><a href="#cb285-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb285-10"><a href="#cb285-10" aria-hidden="true" tabindex="-1"></a>lambda_mle <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">mean</span>(tracts<span class="sc">$</span>length)</span>
<span id="cb285-11"><a href="#cb285-11" aria-hidden="true" tabindex="-1"></a>lambda_mle <span class="sc">/</span> <span class="fl">1e-8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 635.2365</code></pre>
</div>
<div class="sourceCode cell-code" id="cb287"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb287-1"><a href="#cb287-1" aria-hidden="true" tabindex="-1"></a>y_mle <span class="ot">&lt;-</span> <span class="fu">dexp</span>(length, <span class="at">rate =</span> lambda_mle)</span>
<span id="cb287-2"><a href="#cb287-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(length, y_mle, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"darkgreen"</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb287-3"><a href="#cb287-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb287-4"><a href="#cb287-4" aria-hidden="true" tabindex="-1"></a>nls_res <span class="ot">&lt;-</span> <span class="fu">nls</span>(density <span class="sc">~</span> <span class="fu">SSasymp</span>(length, Asym, R0, lrc))</span>
<span id="cb287-5"><a href="#cb287-5" aria-hidden="true" tabindex="-1"></a>nls_res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Nonlinear regression model
  model: density ~ SSasymp(length, Asym, R0, lrc)
   data: parent.frame()
      Asym         R0        lrc 
-5.220e-09  6.255e-06 -1.199e+01 
 residual sum-of-squares: 7.182e-13

Number of iterations to convergence: 0 
Achieved convergence tolerance: 1.978e-07</code></pre>
</div>
<div class="sourceCode cell-code" id="cb289"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb289-1"><a href="#cb289-1" aria-hidden="true" tabindex="-1"></a>lambda_nls <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">unname</span>(<span class="fu">coef</span>(nls_res)[<span class="st">"lrc"</span>]))</span>
<span id="cb289-2"><a href="#cb289-2" aria-hidden="true" tabindex="-1"></a>lambda_nls <span class="sc">/</span> <span class="fl">1e-8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 621.1402</code></pre>
</div>
<div class="sourceCode cell-code" id="cb291"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb291-1"><a href="#cb291-1" aria-hidden="true" tabindex="-1"></a>y_nls <span class="ot">&lt;-</span> <span class="fu">predict</span>(nls_res, <span class="at">newdata =</span> <span class="fu">data.frame</span>(<span class="at">length =</span> length))</span>
<span id="cb291-2"><a href="#cb291-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(length, y_nls, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"purple"</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb291-3"><a href="#cb291-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb291-4"><a href="#cb291-4" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"darkgreen"</span>, <span class="st">"purple"</span>),</span>
<span id="cb291-5"><a href="#cb291-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="fu">paste</span>(<span class="st">"MLE, t ="</span>, <span class="fu">round</span>(lambda_mle <span class="sc">/</span> <span class="fl">1e-8</span>, <span class="dv">1</span>), <span class="st">"generations ago"</span>),</span>
<span id="cb291-6"><a href="#cb291-6" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">paste</span>(<span class="st">"MLE, t ="</span>, <span class="fu">round</span>(lambda_nls <span class="sc">/</span> <span class="fl">1e-8</span>, <span class="dv">1</span>), <span class="st">"generations ago"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr_files/figure-html/unnamed-chunk-85-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./intro.html" class="pagination-link" aria-label="Introduction">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Introduction</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>