<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Computing statistics on tree sequences from slendr – Simulation-Based Population Genomics and Inference in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./slendr-04-inference-afs.html" rel="next">
<link href="./slendr-02-models.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-626149efe8f5d16e1d391ba177679bf0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./slendr-00-intro.html">Introduction to <em>slendr</em></a></li><li class="breadcrumb-item"><a href="./slendr-03-statistics.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Computing statistics on tree sequences from <em>slendr</em></span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Simulation-Based Population Genomics and Inference in R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./slendr-00-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to _slendr_</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-01-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installation and setup</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-02-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Programming demographic models with <em>slendr</em></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-03-statistics.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Computing statistics on tree sequences from <em>slendr</em></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-04-inference-afs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Simulation-based inference of <span class="math inline">\(N_e\)</span></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-05-selection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Simulating (and detecting) natural selection</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-06-nonspatial-pca.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Playing around with PCA patterns</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-07-spatial-pca.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Spatial PCA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slendr-08-tracts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Ancestry tracts and chromosome painintg</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#part-1-computing-nucleotide-diversity" id="toc-part-1-computing-nucleotide-diversity" class="nav-link active" data-scroll-target="#part-1-computing-nucleotide-diversity"><span class="header-section-number">2.1</span> Part 1: Computing nucleotide diversity</a></li>
  <li><a href="#part-2-computing-pairwise-divergence" id="toc-part-2-computing-pairwise-divergence" class="nav-link" data-scroll-target="#part-2-computing-pairwise-divergence"><span class="header-section-number">2.2</span> Part 2: Computing pairwise divergence</a></li>
  <li><a href="#part-3-detecting-neanderthal-admixture-in-europeans" id="toc-part-3-detecting-neanderthal-admixture-in-europeans" class="nav-link" data-scroll-target="#part-3-detecting-neanderthal-admixture-in-europeans"><span class="header-section-number">2.3</span> Part 3: Detecting Neanderthal admixture in Europeans</a></li>
  <li><a href="#part-4-detecting-neanderthal-admixture-in-europeans-v2.0" id="toc-part-4-detecting-neanderthal-admixture-in-europeans-v2.0" class="nav-link" data-scroll-target="#part-4-detecting-neanderthal-admixture-in-europeans-v2.0"><span class="header-section-number">2.4</span> Part 4: Detecting Neanderthal admixture in Europeans v2.0</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./slendr-00-intro.html">Introduction to <em>slendr</em></a></li><li class="breadcrumb-item"><a href="./slendr-03-statistics.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Computing statistics on tree sequences from <em>slendr</em></span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Computing statistics on tree sequences from <em>slendr</em></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this exercise, you will build on top of the results from Exercise 1. Specifically, we will learn how to compute popgen statistics on <em>slendr</em>-simulated tree sequences using <a href="https://www.slendr.net/reference/index.html#tree-sequence-statistics"><em>slendr</em>’s interface</a> to the <em>tskit</em> Python module.</p>
<p><strong>First, create a new R script <code>exercise2.R</code> and paste in the following code.</strong> This is one of the possible solutions to the Exercise 1, and it’s easier if we all use it to be on the same page from now on, starting from the same model and the same simulated tree sequence:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## The interface to all required Python modules has been activated.</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"CHIMP"</span>, <span class="at">time =</span> <span class="fl">7e6</span>, <span class="at">N =</span> <span class="dv">5000</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>afr <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"AFR"</span>, <span class="at">parent =</span> chimp, <span class="at">time =</span> <span class="fl">6e6</span>, <span class="at">N =</span> <span class="dv">15000</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>eur <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"EUR"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">70e3</span>, <span class="at">N =</span> <span class="dv">3000</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>nea <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"NEA"</span>, <span class="at">parent =</span> afr, <span class="at">time =</span> <span class="fl">600e3</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">remove =</span> <span class="fl">40e3</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> nea, <span class="at">to =</span> eur, <span class="at">rate =</span> <span class="fl">0.03</span>, <span class="at">start =</span> <span class="dv">55000</span>, <span class="at">end =</span> <span class="dv">50000</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, nea, afr, eur),</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_flow =</span> gf,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co"># We will read a cached version of a tree sequence I simulated myself</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># to make sure we're all on the same page. That said, if you managed to</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="co"># do Exercise 1 on your own, feel free to stick with your own tree sequence!</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">ts_read</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data/introgression.trees"</span>), <span class="at">model =</span> model)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>),</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>, <span class="at">log =</span> <span class="cn">TRUE</span>),</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">1</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr-03-statistics_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As a sanity check, let’s use a couple of tidyverse table-munging tricks to make sure the tree sequence does contain a set of sample which matches our intended sampling schedule (particularly the time series of European individuals and the two Neanderthals):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># total number of recorded individuals in the tree sequence</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> nrow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 38</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># times of sampling of each recorded European individual</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"EUR"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 40000 38000 36000 34000 32000 30000 28000 26000 24000 22000 20000 18000
[13] 16000 14000 12000 10000  8000  6000  4000  2000     0     0     0     0
[25]     0     0     0     0     0     0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># times of sampling of each recorded Neanderthal</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"NEA"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 70000 40000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># count of individuals in each population</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pop, <span class="at">present_day =</span> time <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  tally <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(pop, present_day, n) <span class="sc">%&gt;%</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(present_day)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
# Groups:   pop [4]
  pop   present_day     n
  &lt;chr&gt; &lt;lgl&gt;       &lt;int&gt;
1 EUR   FALSE          20
2 NEA   FALSE           2
3 AFR   TRUE            5
4 CHIMP TRUE            1
5 EUR   TRUE           10</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> These bits of tidyverse code are extremely helpful when you’re working with large tree sequences with many individuals as sanity checks that your sampling worked as intended. I’m listing them here in case you’ve never worked with the tidyverse family of R packages before (such as the <em>dplyr</em> package where <code>filter()</code>, <code>group_by()</code>, <code>tally()</code>, and <code>pull()</code> come from).</p>
</div></div><p>Everything looks good! Having made sure that the <code>ts</code> object contains the individuals we want, let’s move to the exercise.</p>
<section id="part-1-computing-nucleotide-diversity" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="part-1-computing-nucleotide-diversity"><span class="header-section-number">2.1</span> Part 1: Computing nucleotide diversity</h2>
<p>The toy model of ancient human history plotted above makes a fairly clear prediction of what would be the nucleotide diversity expected in the simulated populations. <strong>Compute the nucleotide diversity in all populations using the <em>slendr</em> function <a href="https://www.slendr.net/reference/ts_diversity.html#ref-examples"><code>ts_diversity()</code></a> in your tree sequence <code>ts</code>. Do you get numbers that (relatively between all populations) match what would expect from the model given the <span class="math inline">\(N_e\)</span> that you programmed for each?</strong></p>
<p><strong>Hint:</strong> Nearly every <em>slendr</em> statistic function interfacing with <em>tskit</em> accepts a <code>ts</code> tree-sequence object as its first argument, with further arguments being either a vector of individual names representing a group of samples to compute a statistic on, or a (named) list of such vectors (each element of that list for a group of samples) – these lists are intended to be equivalent to the <code>sample_sets =</code> argument of many <em>tskit</em> Python methods (which you’ve learned about in your activity on <em>tskit</em>), except that they allow symbolic names of individuals, rather then integer indices of nodes in a tree sequence.</p>
<p>Although you can get all the above information by processing the table produced by the <code>ts_samples()</code> function, <em>slendr</em> provides a useful helper function <code>ts_names()</code> which only returns the names of individuals as a vector (or a named list of such vectors, one vector per population as shown below).</p>
<p>When you call it directly, you get a plain vector of individual names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NEA_1"   "EUR_1"   "NEA_2"   "EUR_2"   "EUR_3"   "EUR_4"   "EUR_5"  
 [8] "EUR_6"   "EUR_7"   "EUR_8"   "EUR_9"   "EUR_10"  "EUR_11"  "EUR_12" 
[15] "EUR_13"  "EUR_14"  "EUR_15"  "EUR_16"  "EUR_17"  "EUR_18"  "EUR_19" 
[22] "EUR_20"  "AFR_1"   "AFR_2"   "AFR_3"   "AFR_4"   "AFR_5"   "CHIMP_1"
[29] "EUR_21"  "EUR_22"  "EUR_23"  "EUR_24"  "EUR_25"  "EUR_26"  "EUR_27" 
[36] "EUR_28"  "EUR_29"  "EUR_30" </code></pre>
</div>
</div>
<p>This is not super helpful, unless we want to compute some statistic for <em>everyone</em> in the tree sequence, regardless of their population assignment. Perhaps a bit more useful is to call the function like this, because it will produce a result which can be immediately used as the <code>sample_sets =</code> argument mentioned in the <strong>Hint</strong> above:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"

$CHIMP
[1] "CHIMP_1"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
</div>
<p>As you can see, this gave us a normal R list, with each element containing a vector of individual names in a population. Note that we can use standard R list indexing to get subsets of individuals:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>names[<span class="st">"NEA"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>names[<span class="fu">c</span>(<span class="st">"EUR"</span>, <span class="st">"NEA"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
</div>
<p>etc.</p>
<p>Many of the following exercises will use these kinds of tricks to instruct various <em>slendr</em> / <em>tskit</em> functions to compute statistics on subsets of all individuals sub-sampled in this way.</p>
<p><strong>After you computed nucleotide diversity per-population, compute it for each individual separately using the same function <code>ts_diversity()</code></strong> (which, in this setting, gives you effectively the heterozygosity for each individual). <strong>If you are familiar with plotting in R, visualize the individual-based heterozygosities across all populations.</strong></p>
<p><strong>Hint:</strong> You can do this by giving a vector of names as <code>sample_sets =</code> (so not an R list of vectors). You could also use the data frame produced by <code>ts_samples(ts)</code> to get the names, just adding the heterozygosities to that data frame as a new column.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Population-based nucleotide diversity:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's first get a named list of individuals in each group we want to be</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># working with (slendr tree-sequence statistic functions generally operate</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with this kind of structure)</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>sample_sets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$AFR
[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"

$CHIMP
[1] "CHIMP_1"

$EUR
 [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"

$NEA
[1] "NEA_1" "NEA_2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use such `sample_sets` object to compute nucleotide diversity (pi)\</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># in each population, in a bit of a similar manner to how we would do it</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with the standard tskit in Python</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>pi_pop <span class="ot">&lt;-</span> <span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> sample_sets)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(pi_pop, diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 NEA   0.0000483
2 CHIMP 0.000196 
3 EUR   0.000510 
4 AFR   0.000600 </code></pre>
</div>
</div>
<p>You can see that this simple computation fits the extreme differences in long-term <span class="math inline">\(N_e\)</span> encoded by your <em>slendr</em> demografr model.</p>
<p><strong>Per-individual heterozygosity:</strong></p>
<p>We can do this by passing the vector of individual names directory as the <code>sample_sets =</code> argument, rather than in a list of groups as we did above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For convenience, we first get a table of all individuals (which of course</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># contains also their names) and in the next step, we'll just add their</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># heterozygosities as a new column.</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>pi_df <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>pi_df<span class="sc">$</span>name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "NEA_1"   "EUR_1"   "NEA_2"   "EUR_2"   "EUR_3"   "EUR_4"   "EUR_5"  
 [8] "EUR_6"   "EUR_7"   "EUR_8"   "EUR_9"   "EUR_10"  "EUR_11"  "EUR_12" 
[15] "EUR_13"  "EUR_14"  "EUR_15"  "EUR_16"  "EUR_17"  "EUR_18"  "EUR_19" 
[22] "EUR_20"  "AFR_1"   "AFR_2"   "AFR_3"   "AFR_4"   "AFR_5"   "CHIMP_1"
[29] "EUR_21"  "EUR_22"  "EUR_23"  "EUR_24"  "EUR_25"  "EUR_26"  "EUR_27" 
[36] "EUR_28"  "EUR_29"  "EUR_30" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pi_df<span class="sc">$</span>diversity <span class="ot">&lt;-</span> <span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> pi_df<span class="sc">$</span>name)<span class="sc">$</span>diversity</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>pi_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 38 × 4
   name   time pop   diversity
   &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 NEA_1 70000 NEA   0.0000431
 2 EUR_1 40000 EUR   0.000516 
 3 NEA_2 40000 NEA   0.0000410
 4 EUR_2 38000 EUR   0.000520 
 5 EUR_3 36000 EUR   0.000539 
 6 EUR_4 34000 EUR   0.000534 
 7 EUR_5 32000 EUR   0.000533 
 8 EUR_6 30000 EUR   0.000553 
 9 EUR_7 28000 EUR   0.000499 
10 EUR_8 26000 EUR   0.000506 
# ℹ 28 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's plot the results using the ggplot2 package</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (because a picture is worth a thousand numbers!)</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pi_df, <span class="fu">aes</span>(pop, diversity, <span class="at">color =</span> pop, <span class="at">group =</span> pop)) <span class="sc">+</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr-03-statistics_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="part-2-computing-pairwise-divergence" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="part-2-computing-pairwise-divergence"><span class="header-section-number">2.2</span> Part 2: Computing pairwise divergence</h2>
<p><strong>Use the function <a href="https://www.slendr.net/reference/ts_divergence.html#ref-examples"><code>ts_divergence()</code></a> to compute genetic divergence between all pairs of populations. Again, do you get results compatible with our demographic model in terms of expectation given the split times between populations as you programmed them for your model?</strong></p>
<p><strong>Hint:</strong> Again, you can use the same concept of <code>sample_sets =</code> we discussed in the previous part. In this case, the function computes <em>pairwise</em> divergence between each element of the list given as <code>sample_sets =</code> (i.e., for each vector of individual names).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>div_df <span class="ot">&lt;-</span> <span class="fu">ts_divergence</span>(ts, sample_sets)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(div_df, divergence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  x     y     divergence
  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
1 AFR   EUR     0.000649
2 EUR   NEA     0.000955
3 AFR   NEA     0.000983
4 CHIMP NEA     0.00416 
5 CHIMP EUR     0.00417 
6 AFR   CHIMP   0.00418 </code></pre>
</div>
</div>
<p>We can see that the pairwise nucleotide divergences between populations recapitulate the known population/species relationships we would expect from our model.</p>
</div>
</div>
</div>
</section>
<section id="part-3-detecting-neanderthal-admixture-in-europeans" class="level2 page-columns page-full" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="part-3-detecting-neanderthal-admixture-in-europeans"><span class="header-section-number">2.3</span> Part 3: Detecting Neanderthal admixture in Europeans</h2>
<p>Let’s now pretend its about 2008, we’ve sequenced the first Neanderthal genome, and we are working on a project that will <a href="https://www.science.org/doi/10.1126/science.1188021">change human evolution research forever</a>. We also have the genomes of a couple of people from Africa and Europe, which we want to use to answer the most burning question of all evolutionary anthropology: <em>“Do some people living today carry Neanderthal ancestry?”</em></p>
<p>Earlier you’ve learned about <span class="math inline">\(f\)</span>-statistics of various kinds. You have also heard that an <span class="math inline">\(f_4\)</span> statistic (or its equivalent <span class="math inline">\(D\)</span> statistic) can be used as a test of “treeness”. Simply speaking, for some “quartet” of individuals or population samples, they can be used as a hypothesis test of whether the history of those samples is compatible with there not having been an introgression.</p>
<p><strong>Compute the <span class="math inline">\(f_4\)</span> test of Neanderthal introgression in EUR individuals using the <em>slendr</em> function <code>ts_f4()</code>.</strong> When you’re running it, you will have to provide individuals to compute the statistic using a slightly different format. Take a look at the help page available as <code>?ts_f4</code> for more information. <strong>When you’re computing the <span class="math inline">\(f_4\)</span>, make sure to set <code>mode = "branch"</code> argument of the <code>ts_f4()</code> function (we will get to why a bit later).</strong></p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> By default, each <em>slendr</em> / <em>tskit</em> statistic function operates on mutations, and this will switch them to use branch length (as you might know, <span class="math inline">\(f\)</span>-statistics are mathematically defined using branch lengths in trees and <code>mode = "branch"</code> does exactly that).</p>
</div></div><p><strong>Hint:</strong> If you haven’t learned this in your <span class="math inline">\(f\)</span>-statistics lecture, you want to compute (and compare) the values of these two statistics using the <em>slendr</em> function <code>ts_f4()</code>:</p>
<ol type="1">
<li><p><span class="math inline">\(f_4\)</span>(&lt;some African&gt;, &lt;another African&gt;; &lt;Neanderthal&gt;, &lt;Chimp&gt;)</p></li>
<li><p><span class="math inline">\(f_4\)</span>(&lt;some African&gt;, &lt;a test European&gt;; &lt;Neanderthal&gt;, &lt;Chimp&gt;),</p></li>
</ol>
<p>here <code>&lt;individual&gt;</code> can be the name of any individual recorded in your tree sequence, such as names you saw as <code>name</code> column in the table returned by <code>ts_samples(ts)</code> (i.e.&nbsp;<code>"NEA_1"</code> could be used as a “representative” &lt;Neanderthal&gt; in those equations, similarly for <code>"CHIMP_1"</code> as the fourth sample in the <span class="math inline">\(f_4\)</span> quarted representing the outgroup).</p>
<p>To simplify things a lot, we can understand the above equations as comparing the counts of so-called BABA and ABBA allele patterns between the quarted of samples specified in the statistics:</p>
<p><span class="math display">\[
f_4(AFR, X; NEA, CHIMP) = \frac{\#BABA - \#ABBA}{\#SNPs}
\]</span></p>
<p>The first <span class="math inline">\(f_4\)</span> statistic above is not expected to give values “too different” from 0 (even in case of Neanderthal introgression into Europeans) because we don’t expect two African individuals to differ “significantly” in terms of how much alleles they share with a Neanderthal (because their ancestors never met Neanderthals!). The other should – if there was a Neanderthal introgression into Europeans some time in their history – be “significantly negative”.</p>
<p><strong>Is the second of those two statistics “much more negative” than the first, as expected assuming introgression from Neanderthals into Europeans?</strong></p>
<p><strong>Why am I putting “significantly” and “much more negative” in quotes in the previous sentence? What are we missing here for this being a true hypothesis test as you might be accustomed to from computing <span class="math inline">\(f\)</span>-statistics using a tool such as ADMIXTOOLS?</strong> (We will get to this again in the following part of this exercise.)</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the difference in the amount of allele sharing between two African</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># individuals and a Neanderthal</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>f4_null <span class="ot">&lt;-</span> <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> <span class="st">"AFR_2"</span>, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>f4_null</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z          f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 AFR_1 AFR_2 NEA_1 CHIMP_1  43.4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the difference in the amount of allele sharing between an African</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="co"># individual vs European individual and a Neanderthal</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>f4_alt <span class="ot">&lt;-</span> <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> <span class="st">"EUR_1"</span>, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>f4_alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z          f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 AFR_1 EUR_1 NEA_1 CHIMP_1 -853.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can see that the second test resulted in an f4 value about ~20 times more</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="co"># negative than the first test, indicating that a European in our test carries</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># "significantly more" Neanderthal alleles compared to the baseline expectation</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of no introgression established by the comparison to an African ...</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abs</span>(f4_alt<span class="sc">$</span>f4 <span class="sc">/</span> f4_null<span class="sc">$</span>f4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19.65719</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ... although this is not a real test of significance (we have no Z-score or</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co"># standard error which would give us something like a p-value for the hypothesis</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># test, as we get by jackknife procedure in ADMIXTOOLS)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="part-4-detecting-neanderthal-admixture-in-europeans-v2.0" class="level2 page-columns page-full" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="part-4-detecting-neanderthal-admixture-in-europeans-v2.0"><span class="header-section-number">2.4</span> Part 4: Detecting Neanderthal admixture in Europeans v2.0</h2>
<p>The fact that we don’t get something equivalent to a p-value in these kinds of simulations is generally not a problem, because we’re often interested in establishing a trend of a statistic under various conditions, and understanding when and how its <em>expected value</em> behaves in a certain way. If statistical noise is a problem, we work around this by computing a statistic on multiple simulation replicates or even increasing the sample sizes.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p><strong>Note:</strong> To see this in practice, you can check out a <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig02">paper</a> in which I used this approach quite successfully on a related problem.</p>
</div></div><p>On top of that, p-value of something like an <span class="math inline">\(f\)</span>-statistic (whether it’s significantly different from zero) is also strongly affected by quality of the data, sequencing errors, coverage, etc. (which can certainly be examined using simulations!). However, these are aspects of modeling which are quite orthogonal to the problem of investigating the expectations and trends of statistics given some underlying evolutionary model, which is what we’re after in these exercises.</p>
<p>That said, even in perfect simulated data, what exactly does “significantly different from zero compared to some baseline expectation” mean can be blurred by noise with simple single-individual comparisons that we did above. Let’s increase the sample size a bit to see if a statistical pattern expected in <span class="math inline">\(f_4\)</span> statistic from our Neanderthal introgression model becomes more apparent.</p>
<p><strong>Compute the first <span class="math inline">\(f_4\)</span> statistic (the baseline expectation between a pair of Africans) and the second <span class="math inline">\(f_4\)</span> statistic (comparing an African and a European), but this time on <em>all recorded Africans</em> and <em>all recorded Europeans</em>, respectively. Plot the <em>distributions</em> of those two sets of statistics. This should remove lots of the uncertainty and make a statistical trend stand out much more clearly.</strong></p>
<p><strong>Hint:</strong> Whenever you need to compute something for many things in sequence, looping is very useful. One way to do compute, say, an <span class="math inline">\(f_4\)</span> statistic over many individuals is by using this kind of pattern using R’s looping function <code>lapply()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over vector of individual names (variable x) and apply a given ts_f4()</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># expression on each individual (note the ts_f4(..., X = x, ...) in the code)</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>list_f4 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"ind_1"</span>, <span class="st">"ind_2"</span>, ...),</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="co"># The above gives us a list of data frames, so we need to bind them all into a</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># single table for easier interpretation and visualization</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>df_f4 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, list_f4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This gives us list of vectors of the names of all individuals in each</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># population...</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>sample_sets <span class="ot">&lt;-</span> <span class="fu">ts_names</span>(ts, <span class="at">split =</span> <span class="st">"pop"</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ... which we can then access like this</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>sample_sets<span class="sc">$</span>AFR <span class="co"># all Africans</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sample_sets<span class="sc">$</span>EUR <span class="co"># all Europeans</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" 
 [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"
[17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"
[25] "EUR_25" "EUR_26" "EUR_27" "EUR_28" "EUR_29" "EUR_30"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's compute the f4 statistic for all Africans... </span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>f4_afr_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  sample_sets<span class="sc">$</span>AFR,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ... and Europeans</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>f4_eur_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  sample_sets<span class="sc">$</span>EUR,</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> <span class="st">"AFR_1"</span>, <span class="at">X =</span> x, <span class="at">Y =</span> <span class="st">"NEA_1"</span>, <span class="at">Z =</span> <span class="st">"CHIMP_1"</span>, <span class="at">mode =</span> <span class="st">"branch"</span>)</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind each list of data frames into a single data frame</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>f4_afr <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, f4_afr_list)</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>f4_eur <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, f4_eur_list)</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's add population columns to each of the two results for easier plotting</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>f4_afr<span class="sc">$</span>pop <span class="ot">&lt;-</span> <span class="st">"AFR"</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>f4_eur<span class="sc">$</span>pop <span class="ot">&lt;-</span> <span class="st">"EUR"</span></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind both tables together</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>f4_results <span class="ot">&lt;-</span> <span class="fu">rbind</span>(f4_afr, f4_eur)</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the results</span></span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>f4_results <span class="sc">%&gt;%</span></span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(pop, f4, <span class="at">color =</span> pop)) <span class="sc">+</span></span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>() <span class="sc">+</span></span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"f4(AFR, EUR; NEA, CHIMP)"</span>) <span class="sc">+</span></span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr-03-statistics_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can see that the <span class="math inline">\(f_4\)</span> statistic test of Neanderthal introgression in Europeans indeed does give a much more negative distribution of values compared to the baseline expectation which compares two Africans to each other.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus exercises
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="bonus-1-mode-branch-vs-mode-site" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="bonus-1-mode-branch-vs-mode-site"><span class="header-section-number">2.4.1</span> Bonus 1: <code>mode = "branch"</code> vs <code>mode = "site"</code></h3>
<p><strong>Repeat the previous part of the exercise by setting <code>mode = "site"</code> in the <code>ts_f4()</code> function calls</strong> (this is actually the default behavior of all <em>slendr</em> tree-sequence based <em>tskit</em> functions). This will switch the <em>tskit</em> computation to using mutation counts along each branch of the tree sequence, rather than using branch length themselves. <strong>Why might the branch-based computation be a bit better if what we’re after is investigating the expected values of statistics under some model?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See <a href="https://tskit.dev/tutorials/no_mutations.html#genealogy-based-measures-are-less-noisy">this tutorial</a> (and particularly the directly linked section) for explanation.</p>
</div>
</div>
</div>
</section>
<section id="bonus-2-outgroup-f_3-statistic" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="bonus-2-outgroup-f_3-statistic"><span class="header-section-number">2.4.2</span> Bonus 2: Outgroup <span class="math inline">\(f_3\)</span> statistic</h3>
<p><strong>Use the function <code>ts_f3()</code> to compute the outgroup</strong> <span class="math inline">\(f_3\)</span> statistic between pairs of African-European, African-Neanderthal, and European-Neanderthal and a Chimpanzee outgroup.</p>
<p><strong>Hint:</strong> The <span class="math inline">\(f_3\)</span> statistic is traditionally expressed as <span class="math inline">\(f_3(A, B; C)\)</span>, where C represents the outgroup. Unfortunately, in <em>tskit</em> the outgroup is named A, with B and C being the pair of samples from which we trace the length of branches towards the outgroup, so the statistic is interpreted as <span class="math inline">\(f_3(B, C; A)\)</span>.</p>
<p><strong>How do the outgroup f3 results compare to your expectation based on simple population relationships (and to the divergence computation above)?</strong></p>
<p><strong>Do you see any impact of introgression on the <span class="math inline">\(f_3\)</span> value when a Neanderthal is included in the computation?</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># f3(A, B; C) = E[ (A - C) * (B - C) ]</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This means that in tskit, C is the outgroup (different from ADMIXTOOLS!)</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We can compute f3 for individuals...</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> <span class="st">"AFR_1"</span>, <span class="at">C =</span> <span class="st">"EUR_1"</span>, <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR_1 EUR_1 0.00375</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ... but also whole populations (or population samples)</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"AFR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"EUR"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR   EUR   0.00375</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"AFR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"NEA"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR   NEA   0.00358</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> sample_sets[<span class="st">"EUR"</span>], <span class="at">C =</span> sample_sets[<span class="st">"NEA"</span>], <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 EUR   NEA   0.00359</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-3-outgroup-f_3-statistic-as-a-linear-combination-of-f_2-statistics" class="level3" data-number="2.4.3">
<h3 data-number="2.4.3" class="anchored" data-anchor-id="bonus-3-outgroup-f_3-statistic-as-a-linear-combination-of-f_2-statistics"><span class="header-section-number">2.4.3</span> Bonus 3: Outgroup <span class="math inline">\(f_3\)</span> statistic as a linear combination of <span class="math inline">\(f_2\)</span> statistics</h3>
<p>You might have learned that any complex <span class="math inline">\(f\)</span>-statistic can be expressed as a linear combination of multiple <span class="math inline">\(f_2\)</span> statistics (which represent simple branch length separating two lineages). <strong>Verify that this is the case by looking up equation <em>(20b)</em> in <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">this amazing paper</a> and compute an</strong> <span class="math inline">\(f_3\)</span> statistic for any arbitrary trio of individuals of your choosing using this linear combination of <span class="math inline">\(f_2\)</span> statistics.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># standard f3</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">B =</span> <span class="st">"AFR_1"</span>, <span class="at">C =</span> <span class="st">"AFR_2"</span>, <span class="at">A =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  A       B     C          f3
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 CHIMP_1 AFR_1 AFR_2 0.00378</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a "homemade" f3 statistic as a linear combination of f2 statistics</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co"># f3(A, B; C) = f2(A, C) + f2(B, C) - f2(A, B) / 2</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>homemade_f3 <span class="ot">&lt;-</span> (</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_1"</span>, <span class="at">B =</span> <span class="st">"CHIMP_1"</span>)<span class="sc">$</span>f2 <span class="sc">+</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_2"</span>, <span class="at">B =</span> <span class="st">"CHIMP_1"</span>)<span class="sc">$</span>f2 <span class="sc">-</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_f2</span>(ts, <span class="at">A =</span> <span class="st">"AFR_1"</span>, <span class="at">B =</span> <span class="st">"AFR_2"</span>)<span class="sc">$</span>f2</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>homemade_f3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.003778796</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-4-trajectory-of-neanderthal-ancestry-in-europe-over-time" class="level3" data-number="2.4.4">
<h3 data-number="2.4.4" class="anchored" data-anchor-id="bonus-4-trajectory-of-neanderthal-ancestry-in-europe-over-time"><span class="header-section-number">2.4.4</span> Bonus 4: Trajectory of Neanderthal ancestry in Europe over time</h3>
<p>There used to be a lot of controversy about the question of whether or not did Neanderthal ancestry proportion in Europeans decline or not over the past 40 thousand years (see figure 1 in <a href="https://www.pnas.org/doi/full/10.1073/pnas.1814338116">this paper</a> figure 2 in <a href="https://www.nature.com/articles/nature17993">this paper</a>).</p>
<p>Your simulated tree sequence contains a time-series of European individuals over time. Use the <em>slendr</em> function <code>ts_f4ratio()</code> to compute (and then plot) the proportion (commonly designated as <code>alpha</code>) of Neanderthal ancestry in Europe over time. Use <span class="math inline">\(f_4\)</span>-ratio statistic of the following form:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> <span class="sc">&lt;</span>vector of ind. names<span class="sc">&gt;</span>, <span class="at">A =</span> <span class="st">"NEA_1"</span>, <span class="at">B =</span> <span class="st">"NEA_2"</span>, <span class="at">C =</span> <span class="st">"AFR_1"</span>, <span class="at">O =</span> <span class="st">"CHIMP_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract table with names and times of sampled Europeans (ancient and present day)</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>eur_inds <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">==</span> <span class="st">"EUR"</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>eur_inds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 3
   name    time pop  
   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 EUR_1  40000 EUR  
 2 EUR_2  38000 EUR  
 3 EUR_3  36000 EUR  
 4 EUR_4  34000 EUR  
 5 EUR_5  32000 EUR  
 6 EUR_6  30000 EUR  
 7 EUR_7  28000 EUR  
 8 EUR_8  26000 EUR  
 9 EUR_9  24000 EUR  
10 EUR_10 22000 EUR  
# ℹ 20 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute f4-ration statistic (this will take ~30s) -- note that we can provide</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a vector of names for the X sample set to the `ts_f4ratio()` function</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>nea_ancestry <span class="ot">&lt;-</span> <span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> eur_inds<span class="sc">$</span>name, <span class="at">A =</span> <span class="st">"NEA_1"</span>, <span class="at">B =</span> <span class="st">"NEA_2"</span>, <span class="at">C =</span> <span class="st">"AFR_1"</span>, <span class="at">O =</span> <span class="st">"CHIMP_1"</span>)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the age of each sample to the table of proportions</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>nea_ancestry<span class="sc">$</span>time <span class="ot">&lt;-</span> eur_inds<span class="sc">$</span>time</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>nea_ancestry</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 7
   X      A     B     C     O        alpha  time
   &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;
 1 EUR_1  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0153 40000
 2 EUR_2  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0170 38000
 3 EUR_3  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0117 36000
 4 EUR_4  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0201 34000
 5 EUR_5  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0261 32000
 6 EUR_6  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0132 30000
 7 EUR_7  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0151 28000
 8 EUR_8  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0243 26000
 9 EUR_9  NEA_1 NEA_2 AFR_1 CHIMP_1 0.0195 24000
10 EUR_10 NEA_1 NEA_2 AFR_1 CHIMP_1 0.0243 22000
# ℹ 20 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>nea_ancestry <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(time, alpha)) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">40000</span>, <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>)) <span class="sc">+</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"time [years ago]"</span>, <span class="at">y =</span> <span class="st">"Neanderthal ancestry proportion"</span>) <span class="sc">+</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Neanderthal ancestry proportion in Europeans over time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="slendr-03-statistics_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For good measure, let's test the significance of the decline using a linear model</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(alpha <span class="sc">~</span> time, <span class="at">data =</span> nea_ancestry))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = alpha ~ time, data = nea_ancestry)

Residuals:
       Min         1Q     Median         3Q        Max 
-0.0102725 -0.0029417  0.0001505  0.0040630  0.0073792 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  2.205e-02  1.214e-03  18.173   &lt;2e-16 ***
time        -1.031e-07  6.204e-08  -1.662    0.108    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.004642 on 28 degrees of freedom
Multiple R-squared:  0.08979,   Adjusted R-squared:  0.05728 
F-statistic: 2.762 on 1 and 28 DF,  p-value: 0.1077</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-5-how-many-unique-f4-quartets-are-there" class="level3" data-number="2.4.5">
<h3 data-number="2.4.5" class="anchored" data-anchor-id="bonus-5-how-many-unique-f4-quartets-are-there"><span class="header-section-number">2.4.5</span> Bonus 5: How many unique f4 quartets are there?</h3>
<p>In your lecture about <span class="math inline">\(f\)</span>-statistics, you’ve probably learned about various symmetries in <span class="math inline">\(f_4\)</span> (but also other <span class="math inline">\(f\)</span>-statistics) depending on the arrangement of the “quartet”. As a trivial example, an <span class="math inline">\(f_3(A; B, C)\)</span> and <span class="math inline">\(f_3(A; C, B)\)</span> will give you exactly the same value, and the same thing applies even to more complex <span class="math inline">\(f\)</span>-statistics like <span class="math inline">\(f_4\)</span>.</p>
<p><strong>Use simulations to compute how manu unique</strong> <span class="math inline">\(f_4\)</span> values involving a single quartet are there.</p>
<p><strong>Hint:</strong> Draw some trees to figure out why could that be true. Also, when computing <code>ts_f4()</code>, set <code>mode = "branch"</code> to avoid the effect of statistical noise due to mutations.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # install a combinatorics R package</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("combinat")</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(combinat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'combinat'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    combn</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># These are the four samples we can create quartet combinations from</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>quartet <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"AFR_1"</span>, <span class="st">"EUR_1"</span>, <span class="st">"NEA_1"</span>, <span class="st">"CHIMP_1"</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>quartets <span class="ot">&lt;-</span> <span class="fu">permn</span>(quartet)</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>quartets</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "AFR_1"   "EUR_1"   "NEA_1"   "CHIMP_1"

[[2]]
[1] "AFR_1"   "EUR_1"   "CHIMP_1" "NEA_1"  

[[3]]
[1] "AFR_1"   "CHIMP_1" "EUR_1"   "NEA_1"  

[[4]]
[1] "CHIMP_1" "AFR_1"   "EUR_1"   "NEA_1"  

[[5]]
[1] "CHIMP_1" "AFR_1"   "NEA_1"   "EUR_1"  

[[6]]
[1] "AFR_1"   "CHIMP_1" "NEA_1"   "EUR_1"  

[[7]]
[1] "AFR_1"   "NEA_1"   "CHIMP_1" "EUR_1"  

[[8]]
[1] "AFR_1"   "NEA_1"   "EUR_1"   "CHIMP_1"

[[9]]
[1] "NEA_1"   "AFR_1"   "EUR_1"   "CHIMP_1"

[[10]]
[1] "NEA_1"   "AFR_1"   "CHIMP_1" "EUR_1"  

[[11]]
[1] "NEA_1"   "CHIMP_1" "AFR_1"   "EUR_1"  

[[12]]
[1] "CHIMP_1" "NEA_1"   "AFR_1"   "EUR_1"  

[[13]]
[1] "CHIMP_1" "NEA_1"   "EUR_1"   "AFR_1"  

[[14]]
[1] "NEA_1"   "CHIMP_1" "EUR_1"   "AFR_1"  

[[15]]
[1] "NEA_1"   "EUR_1"   "CHIMP_1" "AFR_1"  

[[16]]
[1] "NEA_1"   "EUR_1"   "AFR_1"   "CHIMP_1"

[[17]]
[1] "EUR_1"   "NEA_1"   "AFR_1"   "CHIMP_1"

[[18]]
[1] "EUR_1"   "NEA_1"   "CHIMP_1" "AFR_1"  

[[19]]
[1] "EUR_1"   "CHIMP_1" "NEA_1"   "AFR_1"  

[[20]]
[1] "CHIMP_1" "EUR_1"   "NEA_1"   "AFR_1"  

[[21]]
[1] "CHIMP_1" "EUR_1"   "AFR_1"   "NEA_1"  

[[22]]
[1] "EUR_1"   "CHIMP_1" "AFR_1"   "NEA_1"  

[[23]]
[1] "EUR_1"   "AFR_1"   "CHIMP_1" "NEA_1"  

[[24]]
[1] "EUR_1"   "AFR_1"   "NEA_1"   "CHIMP_1"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many permutations there are in total?</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   4! = 4 * 3 * 2 * 1 = 24</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">factorial</span>(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We should therefore have 24 different quartet combinations of samples</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(quartets)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 24</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop across all quartets, computing the corresponding f4 statistic (we want</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to do this using branch lengths, not mutations, as the mutation-based computation</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="co"># would involve statistical noise)</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>all_f4s <span class="ot">&lt;-</span> <span class="fu">lapply</span>(quartets, <span class="cf">function</span>(q) <span class="fu">ts_f4</span>(ts, q[<span class="dv">1</span>], q[<span class="dv">2</span>], q[<span class="dv">3</span>], q[<span class="dv">4</span>], <span class="at">mode =</span> <span class="st">"branch"</span>))</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind the list of f4 results into a single data frame and inspect the results</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>all_f4s <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_f4s) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="fu">abs</span>(f4))</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(all_f4s, <span class="at">n =</span> <span class="cn">Inf</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24 × 5
   W       X       Y       Z            f4
   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
 1 AFR_1   EUR_1   NEA_1   CHIMP_1   -853.
 2 AFR_1   EUR_1   CHIMP_1 NEA_1      853.
 3 NEA_1   CHIMP_1 AFR_1   EUR_1     -853.
 4 CHIMP_1 NEA_1   AFR_1   EUR_1      853.
 5 CHIMP_1 NEA_1   EUR_1   AFR_1     -853.
 6 NEA_1   CHIMP_1 EUR_1   AFR_1      853.
 7 EUR_1   AFR_1   CHIMP_1 NEA_1     -853.
 8 EUR_1   AFR_1   NEA_1   CHIMP_1    853.
 9 AFR_1   NEA_1   CHIMP_1 EUR_1   -16428.
10 AFR_1   NEA_1   EUR_1   CHIMP_1  16428.
11 NEA_1   AFR_1   EUR_1   CHIMP_1 -16428.
12 NEA_1   AFR_1   CHIMP_1 EUR_1    16428.
13 EUR_1   CHIMP_1 NEA_1   AFR_1   -16428.
14 CHIMP_1 EUR_1   NEA_1   AFR_1    16428.
15 CHIMP_1 EUR_1   AFR_1   NEA_1   -16428.
16 EUR_1   CHIMP_1 AFR_1   NEA_1    16428.
17 AFR_1   CHIMP_1 EUR_1   NEA_1    17281.
18 CHIMP_1 AFR_1   EUR_1   NEA_1   -17281.
19 CHIMP_1 AFR_1   NEA_1   EUR_1    17281.
20 AFR_1   CHIMP_1 NEA_1   EUR_1   -17281.
21 NEA_1   EUR_1   CHIMP_1 AFR_1    17281.
22 NEA_1   EUR_1   AFR_1   CHIMP_1 -17281.
23 EUR_1   NEA_1   AFR_1   CHIMP_1  17281.
24 EUR_1   NEA_1   CHIMP_1 AFR_1   -17281.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Narrow down the results to only unique f4 values</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">distinct</span>(all_f4s, f4, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  W       X       Y       Z            f4
  &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
1 AFR_1   EUR_1   NEA_1   CHIMP_1   -853.
2 AFR_1   EUR_1   CHIMP_1 NEA_1      853.
3 AFR_1   NEA_1   CHIMP_1 EUR_1   -16428.
4 AFR_1   NEA_1   EUR_1   CHIMP_1  16428.
5 AFR_1   CHIMP_1 EUR_1   NEA_1    17281.
6 CHIMP_1 AFR_1   EUR_1   NEA_1   -17281.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">distinct</span>(all_f4s, <span class="fu">abs</span>(f4), <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 6
  W     X       Y       Z            f4 `abs(f4)`
  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 AFR_1 EUR_1   NEA_1   CHIMP_1   -853.      853.
2 AFR_1 NEA_1   CHIMP_1 EUR_1   -16428.    16428.
3 AFR_1 CHIMP_1 EUR_1   NEA_1    17281.    17281.</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</div>
</div>
</div>
<!-- End of Bonus exercises -->


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./slendr-02-models.html" class="pagination-link" aria-label="Programming demographic models with _slendr_">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Programming demographic models with <em>slendr</em></span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./slendr-04-inference-afs.html" class="pagination-link" aria-label="Simulation-based inference of $N_e$">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Simulation-based inference of <span class="math inline">\(N_e\)</span></span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>