<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Advanced data science – Simulation-Based Population Genomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./repro-proj.html" rel="next">
<link href="./tidy-tables.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-485d01fc63b59abcd3ee1bf1e8e2748d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./tidy-tables.html">Welcome to tidyverse</a></li><li class="breadcrumb-item"><a href="./tidy-viz.html"><span class="chapter-title">Advanced data science</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Simulation-Based Population Genomics</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./r-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R language</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-bootcamp.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">R bootcamp</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Welcome to tidyverse</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidy-tables.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Basics of data science</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidy-viz.html" class="sidebar-item-text sidebar-link active"><span class="chapter-title">Advanced data science</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Reproducible computing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./repro-proj.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Structuring projects</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./repro-quarto.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Quarto reports and slides</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./repro-cli.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Command-line scripts</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial data science</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial-sf.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Spatial data science</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Modeling fundamentals</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-models.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Demographic models</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-stats.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Tree-sequence statistics</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Simulation-based inference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inference-afs.html" class="sidebar-item-text sidebar-link"><span class="chapter-title"><span class="math inline">\(N_e\)</span> inference with AFS</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">WIP</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-thinking.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Algorithmic thinking</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-nonspatial-pca.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Exploring PCA patterns</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-spatial-pca.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Spatial PCA</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-tracts-dating.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Admixture tracts and dating</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./popgen-selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Natural selection</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./handouts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Handouts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./handouts_slendr.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Introduction to <em>slendr</em></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Installation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Slides</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#exercise-1-exploring-new-data" id="toc-exercise-1-exploring-new-data" class="nav-link active" data-scroll-target="#exercise-1-exploring-new-data">Exercise 1: Exploring new data</a></li>
  <li><a href="#exercise-2-ibd-processing" id="toc-exercise-2-ibd-processing" class="nav-link" data-scroll-target="#exercise-2-ibd-processing">Exercise 2: IBD processing</a></li>
  <li><a href="#exercise-4-metadata-processing" id="toc-exercise-4-metadata-processing" class="nav-link" data-scroll-target="#exercise-4-metadata-processing">Exercise 4: Metadata processing</a></li>
  <li><a href="#exercise-6-reproducibility-intermezzo" id="toc-exercise-6-reproducibility-intermezzo" class="nav-link" data-scroll-target="#exercise-6-reproducibility-intermezzo">Exercise 6: Reproducibility intermezzo</a></li>
  <li><a href="#exercise-6-data-visualization-with-ggplot2" id="toc-exercise-6-data-visualization-with-ggplot2" class="nav-link" data-scroll-target="#exercise-6-data-visualization-with-ggplot2">Exercise 6: Data visualization with <em>ggplot2</em></a>
  <ul class="collapse">
  <li><a href="#distribution-of-categorical-variable" id="toc-distribution-of-categorical-variable" class="nav-link" data-scroll-target="#distribution-of-categorical-variable">Distribution of categorical variable</a></li>
  <li><a href="#distribution-of-a-numerical-variable" id="toc-distribution-of-a-numerical-variable" class="nav-link" data-scroll-target="#distribution-of-a-numerical-variable">Distribution of a numerical variable</a></li>
  <li><a href="#faceting-intermezzo" id="toc-faceting-intermezzo" class="nav-link" data-scroll-target="#faceting-intermezzo">Faceting intermezzo</a></li>
  <li><a href="#combining-figures" id="toc-combining-figures" class="nav-link" data-scroll-target="#combining-figures">Combining figures</a></li>
  <li><a href="#two-numerical-variables" id="toc-two-numerical-variables" class="nav-link" data-scroll-target="#two-numerical-variables">Two numerical variables</a></li>
  </ul></li>
  <li><a href="#exercise-7-summarizing-ibd-data" id="toc-exercise-7-summarizing-ibd-data" class="nav-link" data-scroll-target="#exercise-7-summarizing-ibd-data">Exercise 7: Summarizing IBD data</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./tidy-tables.html">Welcome to tidyverse</a></li><li class="breadcrumb-item"><a href="./tidy-viz.html"><span class="chapter-title">Advanced data science</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-title">Advanced data science</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this chapter, we will be delving into the data visualization R package called <a href="https://ggplot2.tidyverse.org"><em>ggplot2</em></a>, which is possibly the most famous piece of the <em>tidyverse</em> ecosystem. So much so that people who otherwise don’t use any <em>tidyverse</em> functions (or even who don’t even use R for data analysis itself) still use <em>ggplot2</em> for making figures. It really is that good.</p>
<p><strong>Here is how the start of our solutions script for the day will look like.</strong> Note the addition of <code>library(ggplot2)</code> to the list of R packages used in the previous chapter. Although you could use <em>ggplot2</em> without the rest, it ties so neatly with every other part of the <em>tidyverse</em> ecosystem that you almost always end up using them together.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co"># the star of the day!</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Let us also read in data which we will be using in these exercises.</strong> These are coordinates of identity-by-descent (IBD) segments between pairs of individuals in a <a href="https://www.nature.com/articles/s41586-023-06865-0">huge aDNA study</a> we’ve already talked about. This data is huge and quite complex – I’ve prefiltered it to contain only IBD segments for chromosome 1.</p>

<div class="no-row-height column-margin column-container"><div class="margin-aside">
<p>Why use IBD for these exercises? First, IBD data are increasingly popular across population genomics and evolutionary biology, so it’s very likely you will encounter it in your own work. Second, it’s an excellent data se on which you can practice and develop your data science skills.</p>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>gz_file <span class="ot">&lt;-</span> <span class="fu">tempfile</span>()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://tinyurl.com/simgen-ibd"</span>, gz_file, <span class="at">mode =</span> <span class="st">"wb"</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>ibd_all <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(gz_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- load("~/Desktop/neo.impute.1000g.ibd.filter.strict.RData") -->
<!-- ibd_all <- dd -->
<!-- ibd_all <- dplyr::as_tibble(ibd_all) -->
<!-- ibd_all <- dplyr::select(ibd_all, sample1, sample2, chrom = chromosome, start = posCmStart, end = posCmEnd) -->
<!-- ibd_all <- dplyr::filter(ibd_all, chrom == 1) -->
<!-- ibd_all <- dplyr::filter( -->
<!--   ibd_all, -->
<!--   sample1 != "K1" & sample2 != "K1" &  # individual without age -->
<!--   sample1 != "Denisova" & sample2 != "Denisova" & -->
<!--   sample1 != "Vindija33.19" & sample2 != "Vindija33.19" & -->
<!--   sample1 != "AltaiNeandertal" & sample2 != "AltaiNeandertal" -->
<!-- ) -->
<!-- readr::write_tsv(ibd_all, "files/tidy/ibd.tsv.gz") -->
<p>And <strong>we also need to read the corresponding metadata, with which you are already very closely familiar with</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>metadata_all <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"https://tinyurl.com/simgen-metadata"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Create a new R script in RStudio, (<code>File</code> <code>-&gt;</code> <code>New file</code> <code>-&gt;</code> <code>R Script</code>) and save it somewhere on your computer as <code>tidy-viz.R</code> (<code>File</code> -&gt; <code>Save</code>). Copy the two chunks of code above into it and let’s get started!</strong></p>
<p>You just got a new data set from a bioinformatics software, in this case the IBDseq software for detecting IBD segments between pairs of individuals. Before you proceed with doing any kind of statistical analysis you need to do two things:</p>
<ol type="1">
<li>Exploring the data to see <em>what</em> is it that you just got.</li>
<li>Filtering and processing it in a way which will make your work easier.</li>
</ol>
<p>Let’s tackle step 1 first.</p>
<hr>
<section id="exercise-1-exploring-new-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-exploring-new-data">Exercise 1: Exploring new data</h2>
<p><strong>As a sanity check, do you have metadata information for every individual in the <code>sample1</code> and <code>sample2</code> columns of the IBD table? What about the other way around – do all individuals in the metadata have some IBD relationship to another individual? If not, find out which individuals are these.</strong></p>
<p>This is another sort of sanity checking you will be doing all the time. We can only analyze data for which we have metadata information (population assignment, geographical location, dating information), so let’s make sure we have what we need.</p>
<p><strong>Hint:</strong> Another way to phrase this question is this: does every name that appears in either <code>sample1</code> or <code>sample2</code> column of <code>ibd</code> have a record in the <code>sampleId</code> column of <code>metadata</code> (i.e., is information in one vector of names a perfect subset of the second vector of names)? Remember that you can use the function <code>unique()</code> to get all unique values in a given vector (as in, all unique values in vector <code>ibd$sample1</code> which has otherwise many duplicated entries). And remember the existence of <code>%in%</code> and <code>!</code> operators!</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We can get unique values across a combination of vectors by first binding everything together using <code>c(ibd$sample1, ibd$sample2)</code> which gets us a single vector, then calling <code>unique()</code> on that:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ibd_samples <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(ibd_all<span class="sc">$</span>sample1, ibd_all<span class="sc">$</span>sample2))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>metadata_samples <span class="ot">&lt;-</span> metadata_all<span class="sc">$</span>sampleId</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can use our well-known operator <code>%in%</code> to check that every sample in the IBD data has a representative in the metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(ibd_samples <span class="sc">%in%</span> metadata_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Let’s also look at the inverse test:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(metadata_samples <span class="sc">%in%</span> ibd_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<p>Interesting, some individuals don’t have any IBD with anybody else. Who are they?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>metadata_all[<span class="sc">!</span>metadata_samples <span class="sc">%in%</span> ibd_samples, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 23
  sampleId        popId  site  country region groupLabel groupAge flag  latitude
  &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;    &lt;chr&gt;    &lt;dbl&gt;
1 NEO962          NEO962 Drag… Denmark North… Denmark_M… Ancient  0         55.8
2 Vindija33.19    Vindi… Vind… Croatia Centr… Neanderta… Archaic  0         46.3
3 AltaiNeandertal Altai… Deni… Russia  North… Neanderta… Archaic  0         51.4
4 Denisova        Denis… Deni… Russia  North… Denisova_… Archaic  0         51.4
# ℹ 14 more variables: longitude &lt;dbl&gt;, dataSource &lt;chr&gt;, age14C &lt;dbl&gt;,
#   ageHigh &lt;dbl&gt;, ageLow &lt;dbl&gt;, ageAverage &lt;dbl&gt;, datingSource &lt;chr&gt;,
#   coverage &lt;dbl&gt;, sex &lt;chr&gt;, hgMT &lt;chr&gt;, gpAvg &lt;dbl&gt;, ageRaw &lt;chr&gt;,
#   hgYMajor &lt;chr&gt;, hgYMinor &lt;chr&gt;</code></pre>
</div>
</div>
<p>Why doesn’t that ancient Danish person have any IBD with anyone is interesting, but perhaps not a red flag we should be worrying about in our analyses. Same goes for the Archaic individuals – even if they had some IBD shared with modern humans (which we would expect in principle), perhaps the author of the IBD data set filtered those IBD segments out.</p>
</div>
</div>
</div>
<hr>
<p><strong>Which columns does the IBD data have? What’s the format of the data? What ranges or distributions of values do you have available, and with what data types? Do you have information for the entire genome?</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ibd_all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  sample1   sample2 chrom start   end
  &lt;chr&gt;     &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 Gjerrild5 kzb005      1  2.23  4.73
2 Gjerrild5 HG01438     1  2.23  4.73
3 Gjerrild5 NA19649     1  2.23  4.73
4 DA248     DA341       1  2.23  4.73
5 DA249     DA340       1  2.23  4.73
6 DA249     DA341       1  2.23  4.73</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(ibd_all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 2,291,452
Columns: 5
$ sample1 &lt;chr&gt; "Gjerrild5", "Gjerrild5", "Gjerrild5", "DA248", "DA249", "DA24…
$ sample2 &lt;chr&gt; "kzb005", "HG01438", "NA19649", "DA341", "DA340", "DA341", "NE…
$ chrom   &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ start   &lt;dbl&gt; 2.228832, 2.228832, 2.228832, 2.228832, 2.228832, 2.228832, 2.…
$ end     &lt;dbl&gt; 4.734423, 4.734423, 4.734423, 4.734423, 4.734423, 4.734423, 4.…</code></pre>
</div>
</div>
<p>We have information just for chromosome 1 to make the data set a little smaller and easier for your laptops to handle.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ibd_all<span class="sc">$</span>chrom)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
      1 
2291452 </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-2-ibd-processing" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-ibd-processing">Exercise 2: IBD processing</h2>
<p><strong>Add a new column using the <code>mutate()</code> function called <code>length</code>, which contains the length of each IBD segment in centimorgans (<code>end - start</code>). Save the result to a new variable <code>ibd</code>.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> ibd_all <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">length =</span> end <span class="sc">-</span> start)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,291,452 × 6
   sample1   sample2 chrom start   end length
   &lt;chr&gt;     &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 Gjerrild5 kzb005      1  2.23  4.73   2.51
 2 Gjerrild5 HG01438     1  2.23  4.73   2.51
 3 Gjerrild5 NA19649     1  2.23  4.73   2.51
 4 DA248     DA341       1  2.23  4.73   2.51
 5 DA249     DA340       1  2.23  4.73   2.51
 6 DA249     DA341       1  2.23  4.73   2.51
 7 DA335     NEO230      1  2.23  4.73   2.51
 8 DA336     DA338       1  2.23  4.73   2.51
 9 DA336     DA343       1  2.23  4.73   2.51
10 DA336     RISE662     1  2.23  4.73   2.51
# ℹ 2,291,442 more rows</code></pre>
</div>
</div>
<p>Note that you don’t always use a <em>tidyverse</em> approach. Sometimes doing a simple thing using a simple method is just… simpler:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we first make a copy of the original data</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> ibd_all</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and then add the new column</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>ibd<span class="sc">$</span>length <span class="ot">&lt;-</span> ibd_all<span class="sc">$</span>end <span class="sc">-</span> ibd_all<span class="sc">$</span>start</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,291,452 × 6
   sample1   sample2 chrom start   end length
   &lt;chr&gt;     &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 Gjerrild5 kzb005      1  2.23  4.73   2.51
 2 Gjerrild5 HG01438     1  2.23  4.73   2.51
 3 Gjerrild5 NA19649     1  2.23  4.73   2.51
 4 DA248     DA341       1  2.23  4.73   2.51
 5 DA249     DA340       1  2.23  4.73   2.51
 6 DA249     DA341       1  2.23  4.73   2.51
 7 DA335     NEO230      1  2.23  4.73   2.51
 8 DA336     DA338       1  2.23  4.73   2.51
 9 DA336     DA343       1  2.23  4.73   2.51
10 DA336     RISE662     1  2.23  4.73   2.51
# ℹ 2,291,442 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p>We have processed our IBD table a bit, so let’s proceed with the metadata. There’s much more information than we need for now, so let’s make the data a bit smaller and easier to look at at a glance.</p>
<p><strong><code>select()</code> a subset of the metadata with the following columns and store it in the same variable <code>metadata</code>: <code>sampleId</code>, <code>country</code>, <code>ageAverage</code>. Then <code>rename()</code> <code>sampleId</code> to <code>sample</code>, <code>popId</code> to <code>pop</code>, and <code>ageAverage</code> to <code>age</code> just to save ourselves some typing.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">select</span>(metadata_all, <span class="at">sample =</span> sampleId, country, <span class="at">age =</span> ageAverage)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>metadata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,171 × 3
   sample  country   age
   &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;
 1 NA18486 Nigeria    NA
 2 NA18488 Nigeria    NA
 3 NA18489 Nigeria    NA
 4 NA18498 Nigeria    NA
 5 NA18499 Nigeria    NA
 6 NA18501 Nigeria    NA
 7 NA18502 Nigeria    NA
 8 NA18504 Nigeria    NA
 9 NA18505 Nigeria    NA
10 NA18507 Nigeria    NA
# ℹ 4,161 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>Just as you did in the previous chapter, use <code>mutate()</code> and <code>if_else()</code> to make sure that the “Modern” individuals have the <code>age</code> set to 0, instead of <code>NA</code> (and everyone else’s age stays the same). In other words, replace the <code>NA</code> values of <code>age</code> with 0.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">mutate</span>(metadata, <span class="at">age =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(age), <span class="dv">0</span>, age))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s make sure we haven’t missed anything else:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(metadata<span class="sc">$</span>age))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>Our analyses will exclusively focus on modern humans. Filter out the three archaics in the <code>metadata</code>, saving the results into the same <code>metadata</code> variable again. As a reminder, these are individuals whose <code>sample</code> name is among <code>c("Vindija33.19", "AltaiNeandertal", "Denisova")</code> which you can test in a <code>filter()</code> command using the <code>%in%</code> operator.</strong></p>
<p><strong>Hint:</strong> Remember that you can do not only <code>column %in% c(... some values...)</code> but also do the opposite test as <code>!column %in% c(... some values...)</code> (notice the <code>!</code> operator in the second version).</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">filter</span>(metadata, <span class="sc">!</span>sample <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Vindija33.19"</span>, <span class="st">"AltaiNeandertal"</span>, <span class="st">"Denisova"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-4-metadata-processing" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-metadata-processing">Exercise 4: Metadata processing</h2>
<p>We now have a cleaner IBD table looking like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ibd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  sample1   sample2 chrom start   end length
  &lt;chr&gt;     &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 Gjerrild5 kzb005      1  2.23  4.73   2.51
2 Gjerrild5 HG01438     1  2.23  4.73   2.51
3 Gjerrild5 NA19649     1  2.23  4.73   2.51
4 DA248     DA341       1  2.23  4.73   2.51
5 DA249     DA340       1  2.23  4.73   2.51
6 DA249     DA341       1  2.23  4.73   2.51</code></pre>
</div>
</div>
<p>And here’s our metadata information:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  sample  country   age
  &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;
1 NA18486 Nigeria     0
2 NA18488 Nigeria     0
3 NA18489 Nigeria     0
4 NA18498 Nigeria     0
5 NA18499 Nigeria     0
6 NA18501 Nigeria     0</code></pre>
</div>
</div>
<p>Our <code>set</code> has only two values, “Modern” and “Ancient”. Although useful, we might want to do some fancier analyses later down the line, looking at IBD as a time series.</p>
<p><strong>Let’s introduce an incredibly useful function called <code>cut()</code>. Take a look at <code>?cut</code> help page and skim through it to figure out what it does. As a bit of a hint, we will want to add a new metadata column which will indicate in which age bin (maybe, split in steps of 5000 years) do our individuals belong to.</strong></p>
<p>Here’s a small example to help us get started. Think about what the <code>cut()</code> function does here based on the result it gives you. You can pretend for now that the <code>ages</code> variable corresponds to age of our samples in the huge <code>metadata</code> table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a toy example data frame mimicking the age column in our huge metadata table</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">age =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1000</span>, <span class="dv">0</span>, <span class="dv">5000</span>, <span class="dv">10000</span>, <span class="dv">7000</span>, <span class="dv">13000</span>, <span class="dv">18000</span>, <span class="dv">21000</span>, <span class="dv">27000</span>, <span class="dv">30000</span>))</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># let's first generate the breakpoints for our bins (check out `?seq` if</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># you're confused by this!)</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>breakpoints <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50000</span>, <span class="at">by =</span> <span class="dv">5000</span>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>breakpoints</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]     0  5000 10000 15000 20000 25000 30000 35000 40000 45000 50000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># binning the age into groups using our breakpoints</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>age_bin <span class="ot">&lt;-</span> <span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> breakpoints)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     age         age_bin
1      0            &lt;NA&gt;
2      0            &lt;NA&gt;
3   1000       (0,5e+03]
4      0            &lt;NA&gt;
5   5000       (0,5e+03]
6  10000   (5e+03,1e+04]
7   7000   (5e+03,1e+04]
8  13000 (1e+04,1.5e+04]
9  18000 (1.5e+04,2e+04]
10 21000 (2e+04,2.5e+04]
11 27000 (2.5e+04,3e+04]
12 30000 (2.5e+04,3e+04]</code></pre>
</div>
</div>
<p><strong>Note:</strong> The function <code>cut()</code> is extremely useful whenever you want to discretize some continuous variable in bins (basically, a little similar to what a histogram does in context of plotting). Doing statistics on this kind of binned data is something we do very often. So never forget that <code>cut()</code> is there to help you!</p>
<hr>
<p><strong>In the example of the <code>cut()</code> function right above, what is the data type of the column <code>age_bin</code> created by the <code>cut()</code> function? Use <code>glimpse(df)</code> to see this data type, then skim through the documentation of <code>?factor</code>. Additionally, why is the <code>age_bin</code> value equal to <code>NA</code> for some of the values? (We will fix this later, don’t worry.)</strong></p>
<p>As you’re thinking about these questions, consider the two following vectors. When you print them out in your R console (by typing <code>x1</code> and <code>x2</code>) you will get something that looks almost identical. **What happens when you apply the <code>typeof()</code> function on both of them? <code>x2</code> gives you a strange result – why? What do you get when you run the following command <code>levels(x2)</code>? What do you get when you run <code>as.character(x2)</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"hello"</span>, <span class="st">"hello"</span>, <span class="st">"these"</span>, <span class="st">"are"</span>, <span class="st">"characters/strings"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">factor</span>(x1)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>x1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "hello"              "hello"              "these"             
[4] "are"                "characters/strings"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>x2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] hello              hello              these              are               
[5] characters/strings
Levels: are characters/strings hello these</code></pre>
</div>
</div>
<p><strong>Note:</strong> The discussion of “factors” should’ve been technically part of our R bootcamp chapter, on the topic of “data types”. However, that section was already too technical, so I decided to move it here to the data analysis section, because I think it makes more sense to explain it in a wider context.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The reason you get the <code>typeof(x1)</code> as “character” but <code>typeof(x2)</code> as “integer” although both are “basically strings” is because factors are a special data type for encoding categorical variables which can only have a set of fixed possible values. Internally, for efficiency, levels of a factor variables are stored as integers, and their values or just “labels” for those integers.</p>
<p>Importantly, factor levels are actually, which is very useful for plotting, as we will see later.</p>
<p>For now, don’t worry about this too much. We needed to introduce the concept of factors because they are very frequently used whenever we need to bin continuous data, like we will now for assigning samples into bins based on their <code>age</code> value.</p>
<p>Oh, and the answer to why <code>cut()</code> gave us the <code>NA</code> for every 0 value is because of its <code>include.lowest</code> argument is set to <code>FALSE</code>. Compare these two:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50000</span>, <span class="at">by =</span> <span class="dv">5000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] &lt;NA&gt;            &lt;NA&gt;            (0,5e+03]       &lt;NA&gt;           
 [5] (0,5e+03]       (5e+03,1e+04]   (5e+03,1e+04]   (1e+04,1.5e+04]
 [9] (1.5e+04,2e+04] (2e+04,2.5e+04] (2.5e+04,3e+04] (2.5e+04,3e+04]
10 Levels: (0,5e+03] (5e+03,1e+04] (1e+04,1.5e+04] ... (4.5e+04,5e+04]</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50000</span>, <span class="at">by =</span> <span class="dv">5000</span>), <span class="at">include.lowest =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] [0,5e+03]       [0,5e+03]       [0,5e+03]       [0,5e+03]      
 [5] [0,5e+03]       (5e+03,1e+04]   (5e+03,1e+04]   (1e+04,1.5e+04]
 [9] (1.5e+04,2e+04] (2e+04,2.5e+04] (2.5e+04,3e+04] (2.5e+04,3e+04]
10 Levels: [0,5e+03] (5e+03,1e+04] (1e+04,1.5e+04] ... (4.5e+04,5e+04]</code></pre>
</div>
</div>
<p>However, this isn’t what we want, because we want to treat present-day individuals separately as their own category, not include them with other ancient people less than 5000 years old.</p>
<p>The <code>levels()</code> function is useful for getting the “labels” of the categories as they are presented to you in various outputs (rather than their internal numerical encoding).</p>
<p>Consider the original vector <code>x2</code> (note the duplicated “hello” value):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>x2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] hello              hello              these              are               
[5] characters/strings
Levels: are characters/strings hello these</code></pre>
</div>
</div>
<p>And the <code>levels()</code> output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(x2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "are"                "characters/strings" "hello"             
[4] "these"             </code></pre>
</div>
</div>
<p>When we run <code>as.character()</code>, we effectively convert the (categorical) factor values into normal strings (i.e., basically convert those values into their plain labels). This is very useful whenever we want to manipulate factors, as we’ll se below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.character</span>(x2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "hello"              "hello"              "these"             
[4] "are"                "characters/strings"</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>The scientific notation format of bin labels with all those <code>+</code> is very annoying to look at. How can you use the <code>dig.lab =</code> argument of the <code>cut()</code> functions to make this prettier? Experiment in the R console to figure this out, then modify the <code>df$age_bin &lt;- cut(df$age, breaks = breakpoints)</code> command in your script accordingly.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Let’s experiment first:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> breakpoints)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] &lt;NA&gt;            &lt;NA&gt;            (0,5e+03]       &lt;NA&gt;           
 [5] (0,5e+03]       (5e+03,1e+04]   (5e+03,1e+04]   (1e+04,1.5e+04]
 [9] (1.5e+04,2e+04] (2e+04,2.5e+04] (2.5e+04,3e+04] (2.5e+04,3e+04]
10 Levels: (0,5e+03] (5e+03,1e+04] (1e+04,1.5e+04] ... (4.5e+04,5e+04]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> breakpoints, <span class="at">dig.lab =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] &lt;NA&gt;            &lt;NA&gt;            (0,5e+03]       &lt;NA&gt;           
 [5] (0,5e+03]       (5e+03,1e+04]   (5e+03,1e+04]   (1e+04,1.5e+04]
 [9] (1.5e+04,2e+04] (2e+04,2.5e+04] (2.5e+04,3e+04] (2.5e+04,3e+04]
10 Levels: (0,5e+03] (5e+03,1e+04] (1e+04,1.5e+04] ... (4.5e+04,5e+04]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> breakpoints, <span class="at">dig.lab =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] &lt;NA&gt;          &lt;NA&gt;          (0,5000]      &lt;NA&gt;          (0,5000]     
 [6] (5000,10000]  (5000,10000]  (10000,15000] (15000,20000] (20000,25000]
[11] (25000,30000] (25000,30000]
10 Levels: (0,5000] (5000,10000] (10000,15000] (15000,20000] ... (45000,50000]</code></pre>
</div>
</div>
<p>Our solution to get rid of the ugly scientific notation in our labels:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>age_bin <span class="ot">&lt;-</span> <span class="fu">cut</span>(df<span class="sc">$</span>age, <span class="at">breaks =</span> breakpoints, <span class="at">dig.lab =</span> <span class="dv">10</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     age       age_bin
1      0          &lt;NA&gt;
2      0          &lt;NA&gt;
3   1000      (0,5000]
4      0          &lt;NA&gt;
5   5000      (0,5000]
6  10000  (5000,10000]
7   7000  (5000,10000]
8  13000 (10000,15000]
9  18000 (15000,20000]
10 21000 (20000,25000]
11 27000 (25000,30000]
12 30000 (25000,30000]</code></pre>
</div>
</div>
<p>Much nicer to look at!</p>
</div>
</div>
</div>
<hr>
<p>You have now learned that <code>cut()</code> has an optional argument called <code>include.lowest =</code>, which includes the lowest value of 0 (representing the “present-day” age of our samples) in the lowest bin [0, 5]. However, in the case of our assignment of samples from present-day, this is not what we want. We want present-day individuals to have their own category called “present-day”.</p>
<p>Here’s a useful bit of code I use often for this exact purpose. If we start from the original toy example data frame (with the <code>NA</code> values assigned to present-day ages of 0):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     age       age_bin
1      0          &lt;NA&gt;
2      0          &lt;NA&gt;
3   1000      (0,5000]
4      0          &lt;NA&gt;
5   5000      (0,5000]
6  10000  (5000,10000]
7   7000  (5000,10000]
8  13000 (10000,15000]
9  18000 (15000,20000]
10 21000 (20000,25000]
11 27000 (25000,30000]
12 30000 (25000,30000]</code></pre>
</div>
</div>
<p>We can fix this by the following three-step process, described in the comments:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>bin_levels <span class="ot">&lt;-</span> <span class="fu">levels</span>(df<span class="sc">$</span>age_bin) <span class="co"># 1. extract labels (no "present-day" category yet)</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  df <span class="sc">%&gt;%</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">age_bin =</span> <span class="fu">as.character</span>(age_bin),                      <span class="co"># 2. convert factors into strings</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">age_bin =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(age_bin), <span class="st">"present-day"</span>, age_bin) <span class="co"># 3. replace NA with "present-day"</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When we print the modified <code>df</code> table, we see all bins properly assigned now, including the present-day samples with ages at 0:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     age       age_bin
1      0   present-day
2      0   present-day
3   1000      (0,5000]
4      0   present-day
5   5000      (0,5000]
6  10000  (5000,10000]
7   7000  (5000,10000]
8  13000 (10000,15000]
9  18000 (15000,20000]
10 21000 (20000,25000]
11 27000 (25000,30000]
12 30000 (25000,30000]</code></pre>
</div>
</div>
<p><strong>This is a very useful pattern which you will get to practice now on the full <code>metadata</code> table.</strong></p>
<hr>
<p><strong>Now that you’re familiar with <code>cut()</code>, use the functions <code>mutate()</code> and <code>cut()</code> again to create a new column <code>age_bin</code> this time on the whole <code>metadata</code> table. The new column should carry a category age of each individual in steps of 2000 years again. Use the example from above to do this. Use `table(metadata$age_bin) to sanity check your final results.</strong></p>
<p><strong>Hint:</strong> As we did above, first assign the <code>age_bin</code> column using the <code>cut()</code> function, then modify the column accordingly with the <code>mutate()</code> snippet to set “present-day” in the <code>age_bin</code> for all present-day individuals.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we first bin samples according to their age:</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>age_bin <span class="ot">&lt;-</span> <span class="fu">cut</span>(metadata<span class="sc">$</span>age, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">50000</span>, <span class="at">by =</span> <span class="dv">5000</span>), <span class="at">dig.lab =</span> <span class="dv">10</span>)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(metadata<span class="sc">$</span>age_bin)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     (0,5000]  (5000,10000] (10000,15000] (15000,20000] (20000,25000] 
         1211           412            31             0             1 
(25000,30000] (30000,35000] (35000,40000] (40000,45000] (45000,50000] 
            1             6             1             0             1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then we modify the bins to include "present-day" labels:</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">age_bin =</span> <span class="fu">as.character</span>(age_bin),</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">age_bin =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(age_bin), <span class="st">"present-day"</span>, age_bin)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check the binning result to see that we’ve been successful!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(metadata<span class="sc">$</span>age_bin)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     (0,5000] (10000,15000] (20000,25000] (25000,30000] (30000,35000] 
         1211            31             1             1             6 
(35000,40000] (45000,50000]  (5000,10000]   present-day 
            1             1           412          2504 </code></pre>
</div>
</div>
<section id="exercise-5-merging-metadata" class="level2">
<h2 class="anchored" data-anchor-id="exercise-5-merging-metadata">Exercise 5: Merging metadata</h2>
<p>After filtering and clean up, we often need to annotate our popgen data with some metadata information. What does this mean? Consider the IBD data in its current state:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ibd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  sample1   sample2 chrom start   end length
  &lt;chr&gt;     &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 Gjerrild5 kzb005      1  2.23  4.73   2.51
2 Gjerrild5 HG01438     1  2.23  4.73   2.51
3 Gjerrild5 NA19649     1  2.23  4.73   2.51
4 DA248     DA341       1  2.23  4.73   2.51
5 DA249     DA340       1  2.23  4.73   2.51
6 DA249     DA341       1  2.23  4.73   2.51</code></pre>
</div>
</div>
<p>When we analyze things down the line, we might want to look at IBD patterns over space and time, look at some temporal changes, etc. However, our IBD data has none of the information in it, so even if we were to do run our friends <code>group_by()</code> and <code>summarize()</code> to get some summary statistics, we have no idea to correlate them to other variables. We only have names of individuals, and that’s not enough.</p>
<p>The annotation that we need is in the metadata table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  sample  country   age age_bin    
  &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;      
1 NA18486 Nigeria     0 present-day
2 NA18488 Nigeria     0 present-day
3 NA18489 Nigeria     0 present-day
4 NA18498 Nigeria     0 present-day
5 NA18499 Nigeria     0 present-day
6 NA18501 Nigeria     0 present-day</code></pre>
</div>
</div>
<p><strong>What we need to do is join our IBD data with a selection of the most important variables in our metadata, which is what we’re going to do now.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>metadata1 <span class="ot">&lt;-</span> metadata</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>metadata2 <span class="ot">&lt;-</span> metadata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(metadata1) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(metadata1), <span class="st">"1"</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(metadata2) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">colnames</span>(metadata2), <span class="st">"2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(ibd, metadata1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample1)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(ibd, metadata2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample2)`</code></pre>
</div>
</div>
<hr>
<p>The <code>paste()</code> function (and also <code>paste0()</code>) are very useful whenver we need to join the elements of two (or more) character vectors together. The difference between then is that <code>paste0()</code> doesn’t add a space between the individual values and <code>paste()</code> does (the latter also allows you to customize the string which should be used for joining instead of the space). For instance, consider these two vectors and the result of joining them together:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>v1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Denmark"</span>, <span class="st">"Czech Republic"</span>, <span class="st">"Estonia"</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>v2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Finland"</span>, <span class="st">"Italy"</span>, <span class="st">"Estonia"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(v1, v2, <span class="at">sep =</span> <span class="st">":"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Denmark:Finland"      "Czech Republic:Italy" "Estonia:Estonia"     </code></pre>
</div>
</div>
<p><strong>Use the same principle to add a new column to your <code>ibd</code> table named <code>geo_pair</code>, which will contain a vector of values created by joining of the columns <code>country1</code> and <code>country2</code>. Add this column <code>.after</code> the <code>sample1</code> and <code>sample2</code> and <code>.before</code> the <code>chrom</code> columns for clearer visibility. Save the result back to the <code>ibd</code> variable.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">mutate</span>(ibd, <span class="at">geo_pair =</span> <span class="fu">paste</span>(country1, country2, <span class="at">sep =</span> <span class="st">":"</span>), <span class="at">.before =</span> chrom)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,291,452 × 13
   sample1   sample2 geo_pair   chrom start   end length country1  age1 age_bin1
   &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;   
 1 Gjerrild5 kzb005  Denmark:R…     1  2.23  4.73   2.51 Denmark  4110  (0,5000]
 2 Gjerrild5 HG01438 Denmark:C…     1  2.23  4.73   2.51 Denmark  4110  (0,5000]
 3 Gjerrild5 NA19649 Denmark:M…     1  2.23  4.73   2.51 Denmark  4110  (0,5000]
 4 DA248     DA341   Russia:Ru…     1  2.23  4.73   2.51 Russia   7645  (5000,1…
 5 DA249     DA340   Russia:Ru…     1  2.23  4.73   2.51 Russia   7840. (5000,1…
 6 DA249     DA341   Russia:Ru…     1  2.23  4.73   2.51 Russia   7840. (5000,1…
 7 DA335     NEO230  Russia:Ru…     1  2.23  4.73   2.51 Russia   4200  (0,5000]
 8 DA336     DA338   Russia:Ru…     1  2.23  4.73   2.51 Russia   4200  (0,5000]
 9 DA336     DA343   Russia:Ru…     1  2.23  4.73   2.51 Russia   4200  (0,5000]
10 DA336     RISE662 Russia:Ru…     1  2.23  4.73   2.51 Russia   4200  (0,5000]
# ℹ 2,291,442 more rows
# ℹ 3 more variables: country2 &lt;chr&gt;, age2 &lt;dbl&gt;, age_bin2 &lt;chr&gt;</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>In addition to computing statistics across pairs of geographical regions, we will also probably want to look at temporal patterns. Create a new column <code>time_pair</code>, which will in this case contains a <code>paste()</code> combination of <code>age_bin1</code> and <code>age_bin2</code>, added <code>.after</code> the <code>geo_pair</code> column. Save the result back to the <code>ibd</code> variable.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">mutate</span>(ibd, <span class="at">time_pair =</span> <span class="fu">paste</span>(age_bin1, age_bin2, <span class="at">sep =</span> <span class="st">":"</span>), <span class="at">.after =</span> geo_pair)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,291,452 × 14
   sample1   sample2 geo_pair  time_pair chrom start   end length country1  age1
   &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt;
 1 Gjerrild5 kzb005  Denmark:… (0,5000]…     1  2.23  4.73   2.51 Denmark  4110 
 2 Gjerrild5 HG01438 Denmark:… (0,5000]…     1  2.23  4.73   2.51 Denmark  4110 
 3 Gjerrild5 NA19649 Denmark:… (0,5000]…     1  2.23  4.73   2.51 Denmark  4110 
 4 DA248     DA341   Russia:R… (5000,10…     1  2.23  4.73   2.51 Russia   7645 
 5 DA249     DA340   Russia:R… (5000,10…     1  2.23  4.73   2.51 Russia   7840.
 6 DA249     DA341   Russia:R… (5000,10…     1  2.23  4.73   2.51 Russia   7840.
 7 DA335     NEO230  Russia:R… (0,5000]…     1  2.23  4.73   2.51 Russia   4200 
 8 DA336     DA338   Russia:R… (0,5000]…     1  2.23  4.73   2.51 Russia   4200 
 9 DA336     DA343   Russia:R… (0,5000]…     1  2.23  4.73   2.51 Russia   4200 
10 DA336     RISE662 Russia:R… (0,5000]…     1  2.23  4.73   2.51 Russia   4200 
# ℹ 2,291,442 more rows
# ℹ 4 more variables: age_bin1 &lt;chr&gt;, country2 &lt;chr&gt;, age2 &lt;dbl&gt;,
#   age_bin2 &lt;chr&gt;</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p>As a final step, because we have all annotation available in our two <code>pair</code> columns, drop the individual <code>country</code>, <code>age</code>, and <code>age_bin</code> columns using the <code>select()</code> function (remember the <code>starts_with()</code> helper functions?), and save the result back into the <code>ibd</code> variable.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">select</span>(ibd, <span class="sc">-</span><span class="fu">starts_with</span>(<span class="st">"country"</span>), <span class="sc">-</span><span class="fu">starts_with</span>(<span class="st">"age"</span>))</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2,291,452 × 8
   sample1   sample2 geo_pair         time_pair         chrom start   end length
   &lt;chr&gt;     &lt;chr&gt;   &lt;chr&gt;            &lt;chr&gt;             &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 Gjerrild5 kzb005  Denmark:Russia   (0,5000]:(0,5000]     1  2.23  4.73   2.51
 2 Gjerrild5 HG01438 Denmark:Colombia (0,5000]:present…     1  2.23  4.73   2.51
 3 Gjerrild5 NA19649 Denmark:Mexico   (0,5000]:present…     1  2.23  4.73   2.51
 4 DA248     DA341   Russia:Russia    (5000,10000]:(50…     1  2.23  4.73   2.51
 5 DA249     DA340   Russia:Russia    (5000,10000]:(50…     1  2.23  4.73   2.51
 6 DA249     DA341   Russia:Russia    (5000,10000]:(50…     1  2.23  4.73   2.51
 7 DA335     NEO230  Russia:Russia    (0,5000]:(0,5000]     1  2.23  4.73   2.51
 8 DA336     DA338   Russia:Russia    (0,5000]:(0,5000]     1  2.23  4.73   2.51
 9 DA336     DA343   Russia:Russia    (0,5000]:(0,5000]     1  2.23  4.73   2.51
10 DA336     RISE662 Russia:Russia    (0,5000]:(0,5000]     1  2.23  4.73   2.51
# ℹ 2,291,442 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-6-reproducibility-intermezzo" class="level2">
<h2 class="anchored" data-anchor-id="exercise-6-reproducibility-intermezzo">Exercise 6: Reproducibility intermezzo</h2>
<p>You have written a lot of code in this chapter, both for filtering and processing of the IBD data and the metadata. Your script is already very long. If you need to ever update or change something (which you always will during a project, usually many times over) it can be hard to figure out which line of your script should be modified. If you forget to update something somewhere in a long script, you’re in big trouble (and sometimes you won’t even find out where’s the problem until its too late).</p>
<p>Our session on reproducibility is in the following chapter, but let’s work towards making your current workflow more reproducible even now.</p>
<hr>
<p><strong>Create a new script called <code>ibd_utils.R</code> in your RStudio and create the following functions. Then take bits and pieces from the exercises used above and add then to the bodies of these functions accordingly.</strong></p>
<ol type="1">
<li>function <code>process_ibd()</code> which will (without accepting any arguments):</li>
</ol>
<ul>
<li>download the IBD data from the internet like you did above,</li>
<li>process it in the same way you did above,</li>
<li>return the processed <code>tibble</code> object with IBD segments</li>
</ul>
<ol start="2" type="1">
<li>function <code>process_metadata()</code>, which will (without accepting any arguments):</li>
</ol>
<ul>
<li>download the metadata from the internet like you did above,</li>
<li>process and filter it in the same way like above (replacing <code>NA</code> in <code>age</code> column, filtering out unwanted individuals, adding the <code>age_bin</code> column, etc.)</li>
<li>return the processed <code>tibble</code> object with metadata</li>
</ul>
<ol start="3" type="1">
<li>function <code>join_metadata()</code>, which will accept arguments <code>ibd</code> and <code>metadata</code> (produced by the two functions above) and then:</li>
</ol>
<ul>
<li>join the <code>ibd</code> table with the <code>metadata</code> table</li>
<li>add the <code>geo_pair</code> column with <code>country1-country2</code> information</li>
<li>return the finalized <code>tibble</code> with all information</li>
</ul>
<p><strong>Do not write all the code at once! Start with the first function, test it by executing it in your R console, and check that your results make sense before you move on to the other function. Building up more complex pipelines from simple components is absolutely critical to minimize bugs!</strong></p>
<p><strong>Make sure to add comments to your code! Reproducibility is only half of the story – you (and other people) need to be able to understand your code a long time after you wrote it.</strong></p>
<p><strong>After you’re done with this exercise, you should create a new script (you can name it something like <code>ibd_analysis.R</code>) and be able to add the following code.</strong></p>
<p><strong>Note:</strong> The <code>source()</code> command executes all the code in your utility script and therefore makes your custom-built functions available in your R session. This is an incredibly useful trick which you should use all the time whenever you modularize your code into reproducible functions in their own scripts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"ibd_utils.R"</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">process_ibd</span>()</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">process_metadata</span>()</span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> <span class="fu">join_metadata</span>(ibd, metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Downloading and processing IBD data...</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloading and processing metadata...</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Joining IBD data and metadata...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample1)`
Joining with `by = join_by(sample2)`</code></pre>
</div>
</div>
<p><strong>Hint:</strong> If this looks overwhelming and scary, then remember that function is a block of code (optionally accepting arguments) which simply wraps in <code>{</code> and <code>}</code> the code already written by you above, and calls <code>return</code> on the result.</p>
<p><strong>Hint:</strong> An extremely useful shortcut when writing R functions is “Run function definition” under <code>Code</code> <code>-&gt;</code> <code>Run region</code>. Remember that keyboard shortcut! (On macOS it’s <code>Option+CMD+F</code>).</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You can find my solution <a href="https://github.com/bodkan/simgen/blob/main/files/repro/ibd_utils.R">here</a>.</p>
</div>
</div>
</div>
<p><strong>It’s worth pausing here and reflecting on what we’ve just done. Yes, we took (lots!) of code already written and basically “just moved it elsewhere”. But consider how clean your new script <code>ibd_analysis.R</code> is and how easily readable it is. What we’ve done is sometimes called “abstraction” in software engineering. We took components of our pipeline, and “hidden” them away so that we no longer have to think in terms of <code>select()</code>, <code>filter()</code>, <code>group_by()</code>, etc. We now can think about steps called <code>process_ibd()</code>, <code>process_metadata()</code>, and <code>join_metadata()</code>. Much less code to look at and think about! Plus, if we do need to take a deep dive into details, we just have to look at the code of our functions!</strong></p>
<hr>
<p>**Point your cursor somewhere inside the <code>process_ibd()</code> call in your new script <code>ibd_analysis.R</code> and press “CTRL+.”. This is how easy it is to navigate code, even when we modularized it!</p>
<hr>
<p>Having your processing and filtering code available in a self-contained modularized way will make a huge difference in later parts of our workshop.</p>
</section>
<section id="exercise-6-data-visualization-with-ggplot2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-6-data-visualization-with-ggplot2">Exercise 6: Data visualization with <em>ggplot2</em></h2>
<section id="distribution-of-categorical-variable" class="level3">
<h3 class="anchored" data-anchor-id="distribution-of-categorical-variable">Distribution of categorical variable</h3>
<section id="geom_bar" class="level4">
<h4 class="anchored" data-anchor-id="geom_bar"><code>geom_bar()</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span> <span class="fu">geom_bar</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Let’s add more layers! Use a <code>xlab()</code> function to add an x-axis label element using the <code>+</code> operator.</strong></p>
<p>Just as we did with <em>tidyverse</em> <code>%&gt;%</code> data transformation pipelines, as a <em>ggplot2</em> visualization pipeline gets more complex, it’s a good idea to introduce indentation so that each visualization steps is on its own line.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Time period [years before present]"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>Now continue adding y-axis label with the <code>ylab()</code> function.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Time period [years before present]"</span>) <span class="sc">+</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Number of individuals"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>Give your figure a proper main title using the function <code>ggtitle()</code>.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Time period [years before present]"</span>) <span class="sc">+</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Number of individuals"</span>) <span class="sc">+</span></span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Distribution of sample counts in each time period"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>I don’t know about you, but I don’t like how the x-axis labels are overlapping. A solution to this is to use custom theming of plots, but for now, a good workaround is a function <code>coord_flip()</code>. Add it to your figure again using the <code>+</code> operator and observe what happens. Notice what this does to your x- and y-axis labels, and fix the problem!</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Time period [years before present]"</span>) <span class="sc">+</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Number of individuals"</span>) <span class="sc">+</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Distribution of sample counts in each time period"</span>) <span class="sc">+</span></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>Although individual functions <code>xlab()</code>, <code>ylab()</code>, <code>ggtitle()</code> are useful, oftentimes it’s better to use a general function <code>labs()</code>. Look up its documentation under <code>?lab</code>, then rewrite the code for your figure to use only this function, replacing your uses of <code>xlab()</code>, <code>ylab()</code>, and <code>ggtitle()</code> just with <code>labs()</code>. Note that the function has other useful arguments – go ahead and use as many of them as you want.</strong></p>
<p><strong>Note:</strong> Don’t worry about making your figure cluttered, this is just for practice. In a real paper, you wouldn’t use title or caption directly in a <em>ggplot2</em> figure, but it’s definitely useful for work-in-progress reports at meetings, etc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Time period [years before present]"</span>,</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Distribution of sample counts in each time period"</span>,</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">"Counts for individuals in 1000 Genomes Project and MesoNeo data"</span>,</span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">"Here is an optional caption for the figure, similarly to what</span></span>
<span id="cb97-9"><a href="#cb97-9" aria-hidden="true" tabindex="-1"></a><span class="st">      you might do in a real scientific article as a more detailed description of</span></span>
<span id="cb97-10"><a href="#cb97-10" aria-hidden="true" tabindex="-1"></a><span class="st">      of what is going on in the plot."</span></span>
<span id="cb97-11"><a href="#cb97-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb97-12"><a href="#cb97-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>Does the order of adding layers with the <code>+</code> operator matter? Do a little experiment to figure it out!</strong></p>
<p>No it doesn’t. This is another very useful feature of <em>ggplot2</em>. Remember that a huge important part of an R data science workflow is interactive communication with an R console. You can start with a little bit of code which produces a basic outline of your figure and continue adding useful bits and pieces to it, modifying the outcome as you go. Almost like an artist!</p>
<hr>
<p>By the way, it’s useful to keep in mind that you can always pipe a data frame into a <code>ggplot()</code> function using the <code>%&gt;%</code> operator. I.e., instead of writing <code>ggplot(metadata, aes(...))</code>, you can also do <code>metadata %&gt;% ggplot(aes(...))</code>. This allows you to transform or summarize data before plotting, which makes the combination of <em>tidyverse</em> and <em>ggplot2</em> infinitely stronger.</p>
<p><strong>Just for practice, <code>filter()</code> the metadata first for individuals who are 10000 years or older, discarding the rest. Then pipe this <code>filter()</code> result into your <code>ggplot()</code> code, keeping the plotting part exactly the same.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">10000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Time period [years before present]"</span>,</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Distribution of sample counts in each time period"</span>,</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">"Counts for individuals in 1000 Genomes Project and MesoNeo data"</span>,</span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">"Here is an optional caption for the figure, similarly to what</span></span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a><span class="st">      you might do in a real scientific article as a more detailed description of</span></span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a><span class="st">      of what is going on in the plot."</span></span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Believe it or not, with this basic structure of <em>ggplot2</em> visualization and combining this with any other type of <em>tidyverse</em> manipulation, filtering, and (as we’ll later see), summarization, you’re ready to make any figure imaginable. Let’s take this one step further.</p>
<hr>
<p><strong>Let’s show how flexible the <em>ggplot2</em> “grammar of graphics” is. Modify the visuals of your figure by replacing <code>geom_bar()</code> with `geom_point() in the final version of your barplot figure. Why doesn’t the code work? Carefully read the message you get from R when you do this, and try to figure out what the problem is. The solution comes in the next part.</strong></p>
<p>The solution on its own would be this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Time period [years before present]"</span>,</span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Distribution of sample counts in each time period"</span>,</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">"Counts for individuals in 1000 Genomes Project and MesoNeo data"</span>,</span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">"Here is an optional caption for the figure, similarly to what</span></span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a><span class="st">      you might do in a real scientific article as a more detailed description of</span></span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a><span class="st">      of what is going on in the plot."</span></span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `geom_point()`:
! Problem while setting up geom.
ℹ Error occurred in the 1st layer.
Caused by error in `compute_geom_1()`:
! `geom_point()` requires the following missing aesthetics: y.</code></pre>
</div>
</div>
<p>But it gives you an error! Don’t panic!</p>
<hr>
<p>The previous answer gives you an error stating: **</p>
<blockquote class="blockquote">
<p><code>geom_point()</code> requires the following missing aesthetics: y.</p>
</blockquote>
<p>You’ve already heard about “mapping aesthetics”, and the error tells you that <code>geom_point()</code> need both <code>x</code> and <code>y</code> coordinates to plot the elements of a figure. <code>geom_bar()</code> counts the number of individuals in each age group automatically which is why <code>aes(x)</code> was enough. But <code>geom_point()</code> needs us to specify <code>aes(x, y)</code> explicitly.</p>
<p><strong>You already know how to do it! As you learned in the previous chapter, use the <code>metadata %&gt;% group_by(age_bin) %&gt;% summarize(count = n())</code> pipeline to compute the <code>count</code> in each age bin, and then use that <code>count</code> as the <code>y = count</code> mapping!</strong></p>
<p><strong>Hint:</strong> This is a very similar technique to the one you used to only plot individuals older than 10000 years using <code>filter()</code> above.</p>
<p>Note how our code is almost the same, we just needed to change a couple of things to get a different figure! This is why <em>tidyverse</em> is so elegant and powerful (and even easy, once you get familiar with it)! We can use the same patterns and actions over and over and over again, and get new figures in infinite number of variations!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span>                      <span class="co"># here is data processing / summarization part</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">group_by</span>(age_bin) <span class="sc">%&gt;%</span></span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span> </span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age_bin, count)) <span class="sc">+</span> <span class="co"># and this is the plotting part</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Number of individuals"</span>,</span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Time period [years before present]"</span>,</span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Distribution of sample counts in each time period"</span>,</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle =</span> <span class="st">"Counts for individuals in 1000 Genomes Project and MesoNeo data"</span>,</span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">caption =</span> <span class="st">"Here is an optional caption for the figure, similarly to what</span></span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a><span class="st">      you might do in a real scientific article as a more detailed description of</span></span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a><span class="st">      of what is going on in the plot."</span></span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb101-15"><a href="#cb101-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="distribution-of-a-numerical-variable" class="level3">
<h3 class="anchored" data-anchor-id="distribution-of-a-numerical-variable">Distribution of a numerical variable</h3>
<section id="geom_histogram" class="level4">
<h4 class="anchored" data-anchor-id="geom_histogram"><code>geom_histogram()</code></h4>
<p>In the previous exercise you visualized a distribution of a categorical variable, specifically the number of counts of samples in an age category. Now let’s look at continuous variables and consider your <code>metadata</code> table again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  sample  population country   age coverage longitude latitude age_bin    
  &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;   &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;      
1 NA18486 YRI        Nigeria     0       NA        NA       NA present-day
2 NA18488 YRI        Nigeria     0       NA        NA       NA present-day
3 NA18489 YRI        Nigeria     0       NA        NA       NA present-day
4 NA18498 YRI        Nigeria     0       NA        NA       NA present-day
5 NA18499 YRI        Nigeria     0       NA        NA       NA present-day
6 NA18501 YRI        Nigeria     0       NA        NA       NA present-day</code></pre>
</div>
</div>
<hr>
<p><strong>Let’s visualize the distribution of <code>coverage</code> values across our individuals!</strong></p>
<p><strong>You can visualize a density of values using the function <code>geom_histogram()</code>. For instance, your data has a column <code>coverage</code>. Visualize the density distribution of coverages across all individuals in your data set, starting from the (admittedly overly simplified – for now!) pattern of <code>ggplot() + geom_histogram()</code>. Of course, you have to fill in the <code>aes()</code> accordingly. Try to use the knowledge you obtained in the previous exercise!</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage)) <span class="sc">+</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 2504 rows containing non-finite outside the scale range
(`stat_bin()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>You got a warning that 2504 rows have non-finite values. This is because present-day individuals in our metadata lack a <code>coverage</code> value. In the spirit of building composite <em>tidyverse</em> pipelines, <code>filter()</code> out present-day individuals before you plot the <code>coverage</code> distribution.</strong></p>
<p><strong>Note:</strong> Never ever ignore warnings, unless you know exactly where they come from and can be sure they are harmless! Imagine dropping a warning like this in a situation in which an important category would be completely missing values. Your results would be very skewed, leading you to the wrong conclusion!</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage)) <span class="sc">+</span></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
<p>The distribution of coverages across all individuals is not that informative. Maybe we can take a look at those distributions across different categories?</p>
<p>We clearly have two sets of categories we might be interested in. One is perhaps the <code>country</code> of origin, the other the <code>age_bin</code>. <strong>Let’s introduce a new mapping of an aesthetic called <code>fill</code>. In your <code>aes()</code> call, in addition to <code>aes(x = age)</code>, write <code>aes(x = age, fill = age_bin)</code>. What happens when you do this?</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>Staying with this exercise, what’s the difference between setting the <code>fill</code> aesthetic with <code>aes(x = age, fill = age_bin)</code> and the <code>color</code> aesthetic with <code>aes(x = age, fill = age_bin)</code>? What if you do both?</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <code>fill</code> mapping is useful for filling out geometries which involve “areas” of some kind (like a histogram or a barplot). The <code>color</code> mapping is more useful for outlines alone, like we’ll see in a following exercise on plotting densities. <strong>For now just remember that <code>fill</code> and <code>color</code> are useful parameters that you can set for basically every plotting function in <em>ggplot2</em>!</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">color =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>There are two ways in which you can adjust the binning of a <em>ggplot2</em> histogram. Check out <code>?geom_histogram</code> and experiment a little bit.</strong></p>
<p><strong>Note:</strong> The arguments you find which modify the visualization of a histogram are not mapping, so they don’t belong in the <code>aes()</code> call, but either after it (in the <code>ggplot()</code> call) or you can put them directly inside the <code>geom_histogram(...)</code> call. <em>I always get this wrong, so whenever you have issues with this, take a look where you’re specifying these optional arguments!</em></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>We have the <code>bins</code> argument which changes the number of bins. This sets the number of histogram bars to a hardcoded value of 100:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-61-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ol start="2" type="1">
<li>And we have the <code>binwidth</code> argument which changes the width of each bin. This spaces each bar of the histogram at 1 point of coverage:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<hr>
<p>In addition to <code>bins</code> and <code>binwidth</code>, the function <code>geom_histogram()</code> has two additional arguments:</p>
<ul>
<li><code>position</code> has the default value <code>"stack"</code>, but can also be set to <code>"identity"</code> or <code>"dodge"</code>,</li>
<li><code>alpha</code>, which can have a value between 0 and 1 (1 being the default).</li>
</ul>
<p><strong>What happens when you set <code>position = "identity"</code> and <code>alpha = 0.5</code>? How is the result different from the default (which is either <code>geom_histogram()</code> alone or <code>geom_histogram(position = "stack")</code>)?</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>Here is the default:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="co"># same as geom_histogram(position = "stack")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ol start="2" type="1">
<li>Here is the <code>"dodge"</code> variant:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">"dodge"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-64-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ol start="3" type="1">
<li>And here is the <code>"identity"</code> variant. Note that we should probably specify the <code>alpha</code> transparency too!</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">"identity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-65-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="faceting-intermezzo" class="level3">
<h3 class="anchored" data-anchor-id="faceting-intermezzo">Faceting intermezzo</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">%&gt;%</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(coverage, <span class="at">fill =</span> age_bin)) <span class="sc">+</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> age_bin, <span class="at">scales =</span> <span class="st">"free_y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="combining-figures" class="level3">
<h3 class="anchored" data-anchor-id="combining-figures">Combining figures</h3>
</section>
<section id="two-numerical-variables" class="level3">
<h3 class="anchored" data-anchor-id="two-numerical-variables">Two numerical variables</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>geo_df <span class="ot">&lt;-</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>  metadata <span class="sc">%&gt;%</span></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(age <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(country) <span class="sc">%&gt;%</span></span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(</span>
<span id="cb127-6"><a href="#cb127-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">count =</span> <span class="fu">n</span>(),</span>
<span id="cb127-7"><a href="#cb127-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">avg_lat =</span> <span class="fu">mean</span>(latitude, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb127-8"><a href="#cb127-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">avg_lon =</span> <span class="fu">mean</span>(longitude, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb127-9"><a href="#cb127-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">avg_cov =</span> <span class="fu">mean</span>(coverage, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb127-10"><a href="#cb127-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(geo_df, <span class="fu">aes</span>(<span class="at">x =</span> avg_lat, <span class="at">y =</span> count)) <span class="sc">+</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">90</span>, <span class="dv">90</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(geo_df, <span class="fu">aes</span>(<span class="at">x =</span> avg_lon, <span class="at">y =</span> count)) <span class="sc">+</span></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">90</span>, <span class="dv">90</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-71-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> p1 <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p2 <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-74-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-7-summarizing-ibd-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-7-summarizing-ibd-data">Exercise 7: Summarizing IBD data</h2>
<p>OK, data munging (widely acknowledged to be the most annoying part of doing data science) is done. Time to do some data exploration and science! In this and later exercises you will get to practice your <em>tidyverse</em> skills from the previous chapter on your new exciting IBD data.</p>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>ibd_today <span class="ot">&lt;-</span> <span class="fu">filter</span>(ibd, time_pair <span class="sc">==</span> <span class="st">"present-day:present-day"</span>)</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>ibd_today</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 671,001 × 12
   sample1 sample2 geo_pair   time_pair      chrom start   end length longitude1
   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;      &lt;dbl&gt;
 1 HG00096 HG02230 UK:Spain   present-day:p…     1  2.23  4.73   2.51         NA
 2 HG00097 HG00277 UK:Finland present-day:p…     1  2.23  4.73   2.51         NA
 3 HG00097 HG01685 UK:Spain   present-day:p…     1  2.23  4.73   2.51         NA
 4 HG00097 NA20531 UK:Italy   present-day:p…     1  2.23  4.73   2.51         NA
 5 HG00099 HG00330 UK:Finland present-day:p…     1  2.23  4.73   2.51         NA
 6 HG00099 NA11830 UK:Europe  present-day:p…     1  2.23  4.73   2.51         NA
 7 HG00099 NA11840 UK:Europe  present-day:p…     1  2.23  4.73   2.51         NA
 8 HG00100 HG01786 UK:Spain   present-day:p…     1  2.23  4.73   2.51         NA
 9 HG00101 NA20522 UK:Italy   present-day:p…     1  2.23  4.73   2.51         NA
10 HG00102 NA12273 UK:Europe  present-day:p…     1  2.23  4.73   2.51         NA
# ℹ 670,991 more rows
# ℹ 3 more variables: latitude1 &lt;dbl&gt;, longitude2 &lt;dbl&gt;, latitude2 &lt;dbl&gt;</code></pre>
</div>
</div>
<hr>
<p><strong>Plot the distribution of lengths of IBD segments overall.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>ibd_today <span class="sc">%&gt;%</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>ibd_today <span class="sc">%&gt;%</span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(length)) <span class="sc">+</span></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="tidy-viz_files/figure-html/unnamed-chunk-77-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><strong>Use <code>group_by()</code> and <code>summarize()</code> to compute the total length of IBD sequence for each pair of <code>sample1</code> and <code>sample2</code> by summing it up into a new column <code>total</code>. Save the result into a variable <code>ibd_indiv</code>. Sort the result based on this computed <code>total</code> using the <code>arrange()</code> function.</strong></p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>ibd_indiv <span class="ot">&lt;-</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>  ibd <span class="sc">%&gt;%</span></span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample1, sample2) <span class="sc">%&gt;%</span></span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total =</span> <span class="fu">sum</span>(length)) <span class="sc">%&gt;%</span></span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'sample1'. You can override using the
`.groups` argument.</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr>
<p><strong>Although interesting, the previous result only shows the <code>total</code> for each combination of <code>sample1</code> and <code>sample2</code>, because <code>group_by()</code> lost other metadata (we only told it to group by the sample names). Modify your solution to the previous exercise by grouping not just based on samples, but also based on <code>geo_pair</code> and <code>time_pair</code>. This will allow us to investigate IBD patterns geographically and temporally.</strong></p>
<p><strong>Hint:</strong> Just as you did <code>group_by(sample1, sample2)</code> in the previous solution, add the additional pair columns you want to group over after <code>sample1, sample2</code>.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to see the solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>ibd_indiv <span class="ot">&lt;-</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>  ibd <span class="sc">%&gt;%</span></span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sample1, sample2, geo_pair, time_pair) <span class="sc">%&gt;%</span></span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total =</span> <span class="fu">sum</span>(length)) <span class="sc">%&gt;%</span></span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'sample1', 'sample2', 'geo_pair'. You can
override using the `.groups` argument.</code></pre>
</div>
</div>
</div>
</div>
</div>
<!-- --- -->
<!-- ```{r} -->
<!-- rel1 <- filter(metadata, grepl("1d_rel_", flag)) %>% print(n=Inf) %>% .$sampleId -->
<!-- rel2 <- filter(metadata, sampleId %in% rel1)$flag %>% gsub("1d_rel_", "", .) %>% gsub(";.*", "", .) -->
<!-- rels_df <- tibble(sample1 = c(rel1, rel2), sample2 = c(rel2, rel1)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- right_join(total_ibd, rels_df) %>% print(n=Inf) -->
<!-- ``` -->
<!-- 44 VK110      VK230       85.9 -->
<!-- 45 Vt779      Vt808       57.3 -->
<!-- 46 RISE487    RISE489     50.1 -->
<!-- 47 RISE486    RISE489     45.8 -->
<!-- Figure out which (if any) pairs of samples are either resulting from -->
<!-- duplicates or from first degree relatives? In other words, whether some samples -->
<!-- might, in reality, be the same person, or share an IBD spanning an entire -->
<!-- chromosome 1?* -->
<!-- **Generally speaking, very short IBD segments are not very trustworthy. -->
<!-- Filter only to those segments which are longer than 1 Mb. -->
<!-- Save the result back to the `ibd` variable.** -->
<!-- ::: {.callout-note collapse="true" icon=false} -->
<!-- #### Click to see the solution -->
<!-- ```{r} -->
<!-- ibd <- filter(ibd, length > 1) -->
<!-- ``` -->
<!-- --- -->
<!-- **What is the shorted IBD segment? What is the longest? How about mean and median?** -->
<!-- **Note:** Don't forget the `summary()` function! -->
<!-- ::: {.callout-note collapse="true" icon=false} -->
<!-- #### Click to see the solution -->
<!-- ```{r} -->
<!-- summary(ibd$length) -->
<!-- ``` -->
<!-- ::: -->
<!-- --- -->
<!-- **Visualize the histogram of lengths of all IBD segments across your entire -->
<!-- data set.** -->
<!-- ::: {.callout-note collapse="true" icon=false} -->
<!-- #### Click to see the solution -->
<!-- ```{r} -->
<!-- ggplot(ibd, aes(length)) + geom_histogram() -->
<!-- ``` -->
<!-- ::: -->
<!-- --- -->
<!-- ```{r} -->
<!-- ``` -->
<!-- **Let's start simple and ignore the temporal dimension of IBD sharing for now, -->
<!-- and focus only on present-day individuals. To do this, `filter()` only for IBD -->
<!-- segments shared by two present-day individuals (i.e., individuals with `age1` -->
<!-- or `age2` equal to 0). Similarly, `filter()` only on individuals from the -->
<!-- following geographic regions, also based on the additional metadata `country1` -->
<!-- and `country2` added to your IBD table earlier. Save your filtered IBD -->
<!-- table to a variable `ibd_eur`.** -->
<!-- ```{r} -->
<!-- europe <- c("SouthernEurope", "WesternEurope", "NorthernEurope", "CentralEasternEurope") -->
<!-- ``` -->
<!-- ::: {.callout-note collapse="true" icon=false} -->
<!-- #### Click to see the solution -->
<!-- ```{r} -->
<!-- ibd_eur <- filter( -->
<!--   ibd, -->
<!--   age1 == 0 & age2 == 0 & -->
<!--   region1 %in% europe & region2 %in% europe -->
<!-- ) -->
<!-- ``` -->
<!-- ::: -->
<!-- ```{r} -->
<!-- ibd_eur %>% -->
<!-- ggplot() + -->
<!--   geom_jitter(aes(length, pair)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- summary_df <- -->
<!--   ibd_eur %>% -->
<!--   group_by(pair) %>% -->
<!--   summarize(mean_length = mean(length), total_length = sum(length)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- summary_df %>% -->
<!--   ggplot() + -->
<!--   geom_point(aes(mean_length, pair)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- library(forcats) -->
<!-- ibd_eur %>% -->
<!--   group_by(pair) %>% -->
<!--   summarize(mean_length = mean(length)) %>% -->
<!--   ggplot() + -->
<!--   geom_point(aes(mean_length, fct_reorder(pair, mean_length))) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- summary_df <- -->
<!--   ibd_eur %>% -->
<!--   group_by(pair) %>% -->
<!--   summarize(mean_length = mean(length), total_length = sum(length)) %>% -->
<!--   pivot_longer( -->
<!--     cols = c(mean_length, total_length), -->
<!--     names_to = "statistic", -->
<!--     values_to = "value" -->
<!--   ) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ggplot(summary_df) + -->
<!--   geom_point(aes(value, pair)) + -->
<!--   facet_wrap(~ statistic) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- ggplot(summary_df) + -->
<!--   geom_point(aes(value, fct_reorder(pair, value))) + -->
<!--   facet_wrap(~ statistic, scales = "free_x") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- library(tidytext) -->
<!-- ggplot(summary_df) + -->
<!--   geom_point(aes(value, reorder_within(pair, value, within = statistic))) + -->
<!--   facet_wrap(~ statistic, scales = "free") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- p_mean <- -->
<!--   filter(summary_df, statistic == "mean_length") %>% -->
<!--   ggplot + -->
<!--   geom_point(aes(value, fct_reorder(pair, value))) + -->
<!--   facet_wrap(~ statistic, scales = "free_x") -->
<!-- p_mean -->
<!-- ``` -->
<!-- ```{r} -->
<!-- p_sum <- -->
<!--   filter(summary_df, statistic == "total_length") %>% -->
<!--   ggplot + -->
<!--   geom_point(aes(value, fct_reorder(pair, value))) + -->
<!--   facet_wrap(~ statistic, scales = "free_x") -->
<!-- p_sum -->
<!-- ``` -->
<!-- ```{r} -->
<!-- library(cowplot) -->
<!-- plot_grid(p_mean, p_sum) -->
<!-- ``` -->
<!-- ## Further practice -->
<!-- I'm sure you have done some form of table munging and data visualization in your -->
<!-- projects. If you're done with the exercises above, here are some ideas for you: -->
<!-- 1. If you haven't used R for processing of tabular data in your work -- maybe -->
<!-- you have done similar table munging in Excel or Google Sheets in a manual -->
<!-- (i.e., non-reproducible) way? -- write your processing steps as a self-contained -->
<!-- R script: -->
<!-- - If you used Excel, you can use the R package [_readxl_](https://readxl.tidyverse.org), -->
<!-- - If you used Google Sheets, you can use the R package [_googlesheets4_](https://googlesheets4.tidyverse.org) -->
<!-- 2. If you used Excel or base R for plotting figures, use what you learned -->
<!-- about _ggplot2_ to make beautiful, publication-ready plots as we did in our -->
<!-- exercises. -->
<!-- Yes, this will put you in an uncharted territory. This is scary and totally -->
<!-- normal when doing research. I'm happy to help you out with questions -->
<!-- and issues! -->


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./tidy-tables.html" class="pagination-link" aria-label="Basics of data science">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-title">Basics of data science</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./repro-proj.html" class="pagination-link" aria-label="Structuring projects">
        <span class="nav-page-text"><span class="chapter-title">Structuring projects</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>